{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNrYESYftJb2b3MjsARIlmN",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mbjallow6/Algorithms-python/blob/main/Boinformatic_Problems_Rosalind.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# import data from the computer\n",
        "from google.colab import files\n",
        "uploaded = files.upload()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 73
        },
        "id": "gvUovf3CR0Q5",
        "outputId": "9d763e03-860b-4fce-e135-1c66189843e3"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-9cbeaa63-172e-4c11-bc01-d2f3baa9e43a\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-9cbeaa63-172e-4c11-bc01-d2f3baa9e43a\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving rosalind_sign.txt to rosalind_sign.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Enumerating Oriented Gene Orderings Problem"
      ],
      "metadata": {
        "id": "rNKC2vmsn_XP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Enumerating Oriented Gene Orderings Problem Solution\n",
        "\n",
        "This module generates all signed permutations of length n, where each element\n",
        "can have either positive or negative orientation. This models synteny blocks\n",
        "in chromosomes that can be oriented on either DNA strand.\n",
        "\n",
        "A signed permutation allows each integer to have a + or - sign, representing\n",
        "the orientation of gene blocks on chromosomes.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Tuple\n",
        "from itertools import permutations, product\n",
        "import math\n",
        "\n",
        "\n",
        "class SignedPermutationGenerator:\n",
        "    \"\"\"\n",
        "    A class to generate signed permutations representing oriented gene orderings.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the signed permutation generator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_input(n: int) -> bool:\n",
        "        \"\"\"\n",
        "        Validate input parameter according to problem constraints.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            bool: True if input is valid, False otherwise\n",
        "        \"\"\"\n",
        "        return isinstance(n, int) and 1 <= n <= 6\n",
        "\n",
        "    @staticmethod\n",
        "    def calculate_signed_permutation_count(n: int) -> int:\n",
        "        \"\"\"\n",
        "        Calculate total number of signed permutations.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            int: Total count = n! × 2^n\n",
        "        \"\"\"\n",
        "        return math.factorial(n) * (2 ** n)\n",
        "\n",
        "    def generate_all_signed_permutations(self, n: int) -> List[List[int]]:\n",
        "        \"\"\"\n",
        "        Generate all signed permutations of length n.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            List[List[int]]: List of all signed permutations\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If n is invalid\n",
        "        \"\"\"\n",
        "        if not self.validate_input(n):\n",
        "            raise ValueError(f\"Invalid input: n={n}. Must be 1 ≤ n ≤ 6\")\n",
        "\n",
        "        signed_permutations = []\n",
        "\n",
        "        # Generate all base permutations of [1, 2, ..., n]\n",
        "        base_permutations = list(permutations(range(1, n + 1)))\n",
        "\n",
        "        # For each base permutation, generate all sign combinations\n",
        "        for base_perm in base_permutations:\n",
        "            # Generate all possible sign combinations (each element can be +1 or -1)\n",
        "            for signs in product([-1, 1], repeat=n):\n",
        "                # Apply signs to the permutation\n",
        "                signed_perm = [sign * value for sign, value in zip(signs, base_perm)]\n",
        "                signed_permutations.append(signed_perm)\n",
        "\n",
        "        return signed_permutations\n",
        "\n",
        "    def solve_signed_permutation_problem(self, n: int) -> Tuple[int, List[List[int]]]:\n",
        "        \"\"\"\n",
        "        Solve the complete signed permutation problem.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            Tuple[int, List[List[int]]]: (total_count, all_signed_permutations)\n",
        "        \"\"\"\n",
        "        # Generate all signed permutations\n",
        "        signed_perms = self.generate_all_signed_permutations(n)\n",
        "\n",
        "        # Count should match theoretical calculation\n",
        "        expected_count = self.calculate_signed_permutation_count(n)\n",
        "        actual_count = len(signed_perms)\n",
        "\n",
        "        if actual_count != expected_count:\n",
        "            raise ValueError(f\"Count mismatch: expected {expected_count}, got {actual_count}\")\n",
        "\n",
        "        return actual_count, signed_perms\n",
        "\n",
        "    def get_analysis(self, n: int) -> dict:\n",
        "        \"\"\"\n",
        "        Get detailed analysis of signed permutation generation.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            dict: Analysis including counts, examples, and mathematical breakdown\n",
        "        \"\"\"\n",
        "        if not self.validate_input(n):\n",
        "            raise ValueError(f\"Invalid input: n={n}\")\n",
        "\n",
        "        count, signed_perms = self.solve_signed_permutation_problem(n)\n",
        "\n",
        "        # Group by base permutation for analysis\n",
        "        base_perms = list(permutations(range(1, n + 1)))\n",
        "\n",
        "        analysis = {\n",
        "            'n': n,\n",
        "            'base_permutation_count': len(base_perms),\n",
        "            'sign_combinations_per_perm': 2 ** n,\n",
        "            'total_signed_permutations': count,\n",
        "            'calculation': f\"{len(base_perms)} × {2**n} = {count}\",\n",
        "            'factorial_calculation': f\"{n}! × 2^{n} = {math.factorial(n)} × {2**n} = {count}\",\n",
        "            'signed_permutations': signed_perms,\n",
        "            'first_few_examples': signed_perms[:min(8, len(signed_perms))],\n",
        "            'base_permutations': [list(perm) for perm in base_perms]\n",
        "        }\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> int:\n",
        "    \"\"\"\n",
        "    Parse input file to extract n.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        int: The integer n\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file content is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        try:\n",
        "            n = int(content)\n",
        "            return n\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Invalid input: '{content}' is not a valid integer\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, count: int, signed_permutations: List[List[int]]) -> None:\n",
        "    \"\"\"\n",
        "    Write signed permutations to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        count (int): Total number of signed permutations\n",
        "        signed_permutations (List[List[int]]): List of signed permutations\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            # Write count first\n",
        "            file.write(f\"{count}\\n\")\n",
        "\n",
        "            # Write all signed permutations\n",
        "            for perm in signed_permutations:\n",
        "                file.write(' '.join(map(str, perm)) + '\\n')\n",
        "\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_signed_permutation_problem(input_file_path: str) -> Tuple[int, List[List[int]]]:\n",
        "    \"\"\"\n",
        "    Solve the Signed Permutation problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file containing integer n\n",
        "\n",
        "    Returns:\n",
        "        Tuple[int, List[List[int]]]: (count, list_of_signed_permutations)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        n = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize generator\n",
        "        generator = SignedPermutationGenerator()\n",
        "\n",
        "        # Solve the problem\n",
        "        count, signed_permutations = generator.solve_signed_permutation_problem(n)\n",
        "\n",
        "        return count, signed_permutations\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Signed Permutation problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_sign.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_sign.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Signed Permutation Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        count, signed_permutations = solve_signed_permutation_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResults:\")\n",
        "        print(f\"Total number of signed permutations: {count}\")\n",
        "        print(f\"All signed permutations:\")\n",
        "\n",
        "        for perm in signed_permutations:\n",
        "            print(' '.join(map(str, perm)))\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, count, signed_permutations)\n",
        "        print(f\"\\nResults written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data from search results.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input from the problem and search results\n",
        "    n = 2\n",
        "\n",
        "    print(f\"Input: n = {n}\")\n",
        "\n",
        "    generator = SignedPermutationGenerator()\n",
        "    analysis = generator.get_analysis(n)\n",
        "\n",
        "    print(f\"\\nDetailed Analysis:\")\n",
        "    print(f\"Base permutations: {analysis['base_permutation_count']}\")\n",
        "    print(f\"Sign combinations per permutation: {analysis['sign_combinations_per_perm']}\")\n",
        "    print(f\"Calculation: {analysis['calculation']}\")\n",
        "    print(f\"Mathematical formula: {analysis['factorial_calculation']}\")\n",
        "    print(f\"Total signed permutations: {analysis['total_signed_permutations']}\")\n",
        "\n",
        "    print(f\"\\nBase permutations of [1, 2, ..., {n}]:\")\n",
        "    for base_perm in analysis['base_permutations']:\n",
        "        print(f\"  {base_perm}\")\n",
        "\n",
        "    print(f\"\\nAll signed permutations:\")\n",
        "    for perm in analysis['signed_permutations']:\n",
        "        print(' '.join(map(str, perm)))\n",
        "\n",
        "    # Expected output verification (from search results)\n",
        "    expected_count = 8\n",
        "    expected_perms = [\n",
        "        [-1, -2], [-1, 2], [1, -2], [1, 2],\n",
        "        [-2, -1], [-2, 1], [2, -1], [2, 1]\n",
        "    ]\n",
        "\n",
        "    print(f\"\\nVerification:\")\n",
        "    print(f\"Expected count: {expected_count}\")\n",
        "    print(f\"Our count: {analysis['total_signed_permutations']}\")\n",
        "    print(f\"Count match: {'✓' if analysis['total_signed_permutations'] == expected_count else '✗'}\")\n",
        "\n",
        "    # Check if our permutations match (order may be different)\n",
        "    our_perms_set = set(tuple(perm) for perm in analysis['signed_permutations'])\n",
        "    expected_perms_set = set(tuple(perm) for perm in expected_perms)\n",
        "\n",
        "    print(f\"Permutations match: {'✓' if our_perms_set == expected_perms_set else '✗'}\")\n",
        "\n",
        "\n",
        "def demonstrate_mathematical_concept():\n",
        "    \"\"\"\n",
        "    Demonstrate the mathematical concept behind signed permutations.\n",
        "    \"\"\"\n",
        "    print(\"=== Mathematical Concept Demonstration ===\")\n",
        "\n",
        "    print(\"Signed permutations combine:\")\n",
        "    print(\"1. Regular permutations: arrangements of {1, 2, ..., n}\")\n",
        "    print(\"2. Sign assignments: each element can be positive or negative\")\n",
        "    print()\n",
        "    print(\"Formula: Total signed permutations = n! × 2^n\")\n",
        "    print(\"  • n! regular permutations\")\n",
        "    print(\"  • 2^n sign combinations for each permutation\")\n",
        "    print()\n",
        "\n",
        "    generator = SignedPermutationGenerator()\n",
        "\n",
        "    for test_n in range(1, 5):\n",
        "        try:\n",
        "            analysis = generator.get_analysis(test_n)\n",
        "            print(f\"n = {test_n}:\")\n",
        "            print(f\"  Regular permutations: {test_n}! = {math.factorial(test_n)}\")\n",
        "            print(f\"  Sign combinations: 2^{test_n} = {2**test_n}\")\n",
        "            print(f\"  Total signed permutations: {analysis['factorial_calculation']}\")\n",
        "            print(f\"  Examples: {analysis['first_few_examples']}\")\n",
        "            print()\n",
        "        except Exception as e:\n",
        "            print(f\"n = {test_n}: Error - {e}\")\n",
        "\n",
        "\n",
        "def demonstrate_biological_relevance():\n",
        "    \"\"\"\n",
        "    Explain the biological relevance of signed permutations.\n",
        "    \"\"\"\n",
        "    print(\"=== Biological Relevance ===\")\n",
        "    print(\"Signed permutations in genomics:\")\n",
        "    print(\"• Model synteny blocks with strand orientation\")\n",
        "    print(\"• Each block can be on forward (+) or reverse (-) strand\")\n",
        "    print(\"• Essential for understanding chromosomal rearrangements\")\n",
        "    print(\"• Help track evolutionary changes between species\")\n",
        "    print()\n",
        "    print(\"Example: Block arrangement [+1, -2, +3] means:\")\n",
        "    print(\"  - Block 1: forward orientation\")\n",
        "    print(\"  - Block 2: reverse orientation\")\n",
        "    print(\"  - Block 3: forward orientation\")\n",
        "    print()\n",
        "    print(\"Chromosomal inversions change + to - or vice versa\")\n",
        "    print(\"Translocations change the order of blocks\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate mathematical concept\n",
        "    demonstrate_mathematical_concept()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate biological relevance\n",
        "    demonstrate_biological_relevance()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test edge cases\n",
        "    print(\"=== Testing Edge Cases ===\")\n",
        "    generator = SignedPermutationGenerator()\n",
        "\n",
        "    for test_n in [1, 3, 4]:\n",
        "        try:\n",
        "            count, signed_perms = generator.solve_signed_permutation_problem(test_n)\n",
        "            print(f\"n = {test_n}: {count} signed permutations\")\n",
        "            if test_n <= 3:  # Show examples for small cases\n",
        "                print(f\"  Examples: {signed_perms[:min(6, len(signed_perms))]}\")\n",
        "        except Exception as e:\n",
        "            print(f\"n = {test_n}: Error - {e}\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "lTpAlETYoA7u",
        "outputId": "2f2cdcc1-3d38-4743-d7d8-8fca92ec39a2",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input: n = 2\n",
            "\n",
            "Detailed Analysis:\n",
            "Base permutations: 2\n",
            "Sign combinations per permutation: 4\n",
            "Calculation: 2 × 4 = 8\n",
            "Mathematical formula: 2! × 2^2 = 2 × 4 = 8\n",
            "Total signed permutations: 8\n",
            "\n",
            "Base permutations of [1, 2, ..., 2]:\n",
            "  [1, 2]\n",
            "  [2, 1]\n",
            "\n",
            "All signed permutations:\n",
            "-1 -2\n",
            "-1 2\n",
            "1 -2\n",
            "1 2\n",
            "-2 -1\n",
            "-2 1\n",
            "2 -1\n",
            "2 1\n",
            "\n",
            "Verification:\n",
            "Expected count: 8\n",
            "Our count: 8\n",
            "Count match: ✓\n",
            "Permutations match: ✓\n",
            "\n",
            "============================================================\n",
            "=== Mathematical Concept Demonstration ===\n",
            "Signed permutations combine:\n",
            "1. Regular permutations: arrangements of {1, 2, ..., n}\n",
            "2. Sign assignments: each element can be positive or negative\n",
            "\n",
            "Formula: Total signed permutations = n! × 2^n\n",
            "  • n! regular permutations\n",
            "  • 2^n sign combinations for each permutation\n",
            "\n",
            "n = 1:\n",
            "  Regular permutations: 1! = 1\n",
            "  Sign combinations: 2^1 = 2\n",
            "  Total signed permutations: 1! × 2^1 = 1 × 2 = 2\n",
            "  Examples: [[-1], [1]]\n",
            "\n",
            "n = 2:\n",
            "  Regular permutations: 2! = 2\n",
            "  Sign combinations: 2^2 = 4\n",
            "  Total signed permutations: 2! × 2^2 = 2 × 4 = 8\n",
            "  Examples: [[-1, -2], [-1, 2], [1, -2], [1, 2], [-2, -1], [-2, 1], [2, -1], [2, 1]]\n",
            "\n",
            "n = 3:\n",
            "  Regular permutations: 3! = 6\n",
            "  Sign combinations: 2^3 = 8\n",
            "  Total signed permutations: 3! × 2^3 = 6 × 8 = 48\n",
            "  Examples: [[-1, -2, -3], [-1, -2, 3], [-1, 2, -3], [-1, 2, 3], [1, -2, -3], [1, -2, 3], [1, 2, -3], [1, 2, 3]]\n",
            "\n",
            "n = 4:\n",
            "  Regular permutations: 4! = 24\n",
            "  Sign combinations: 2^4 = 16\n",
            "  Total signed permutations: 4! × 2^4 = 24 × 16 = 384\n",
            "  Examples: [[-1, -2, -3, -4], [-1, -2, -3, 4], [-1, -2, 3, -4], [-1, -2, 3, 4], [-1, 2, -3, -4], [-1, 2, -3, 4], [-1, 2, 3, -4], [-1, 2, 3, 4]]\n",
            "\n",
            "\n",
            "============================================================\n",
            "=== Biological Relevance ===\n",
            "Signed permutations in genomics:\n",
            "• Model synteny blocks with strand orientation\n",
            "• Each block can be on forward (+) or reverse (-) strand\n",
            "• Essential for understanding chromosomal rearrangements\n",
            "• Help track evolutionary changes between species\n",
            "\n",
            "Example: Block arrangement [+1, -2, +3] means:\n",
            "  - Block 1: forward orientation\n",
            "  - Block 2: reverse orientation\n",
            "  - Block 3: forward orientation\n",
            "\n",
            "Chromosomal inversions change + to - or vice versa\n",
            "Translocations change the order of blocks\n",
            "\n",
            "============================================================\n",
            "=== Testing Edge Cases ===\n",
            "n = 1: 2 signed permutations\n",
            "  Examples: [[-1], [1]]\n",
            "n = 3: 48 signed permutations\n",
            "  Examples: [[-1, -2, -3], [-1, -2, 3], [-1, 2, -3], [-1, 2, 3], [1, -2, -3], [1, -2, 3]]\n",
            "n = 4: 384 signed permutations\n",
            "\n",
            "============================================================\n",
            "Solving Signed Permutation Problem...\n",
            "\n",
            "Results:\n",
            "Total number of signed permutations: 384\n",
            "All signed permutations:\n",
            "-1 -2 -3 -4\n",
            "-1 -2 -3 4\n",
            "-1 -2 3 -4\n",
            "-1 -2 3 4\n",
            "-1 2 -3 -4\n",
            "-1 2 -3 4\n",
            "-1 2 3 -4\n",
            "-1 2 3 4\n",
            "1 -2 -3 -4\n",
            "1 -2 -3 4\n",
            "1 -2 3 -4\n",
            "1 -2 3 4\n",
            "1 2 -3 -4\n",
            "1 2 -3 4\n",
            "1 2 3 -4\n",
            "1 2 3 4\n",
            "-1 -2 -4 -3\n",
            "-1 -2 -4 3\n",
            "-1 -2 4 -3\n",
            "-1 -2 4 3\n",
            "-1 2 -4 -3\n",
            "-1 2 -4 3\n",
            "-1 2 4 -3\n",
            "-1 2 4 3\n",
            "1 -2 -4 -3\n",
            "1 -2 -4 3\n",
            "1 -2 4 -3\n",
            "1 -2 4 3\n",
            "1 2 -4 -3\n",
            "1 2 -4 3\n",
            "1 2 4 -3\n",
            "1 2 4 3\n",
            "-1 -3 -2 -4\n",
            "-1 -3 -2 4\n",
            "-1 -3 2 -4\n",
            "-1 -3 2 4\n",
            "-1 3 -2 -4\n",
            "-1 3 -2 4\n",
            "-1 3 2 -4\n",
            "-1 3 2 4\n",
            "1 -3 -2 -4\n",
            "1 -3 -2 4\n",
            "1 -3 2 -4\n",
            "1 -3 2 4\n",
            "1 3 -2 -4\n",
            "1 3 -2 4\n",
            "1 3 2 -4\n",
            "1 3 2 4\n",
            "-1 -3 -4 -2\n",
            "-1 -3 -4 2\n",
            "-1 -3 4 -2\n",
            "-1 -3 4 2\n",
            "-1 3 -4 -2\n",
            "-1 3 -4 2\n",
            "-1 3 4 -2\n",
            "-1 3 4 2\n",
            "1 -3 -4 -2\n",
            "1 -3 -4 2\n",
            "1 -3 4 -2\n",
            "1 -3 4 2\n",
            "1 3 -4 -2\n",
            "1 3 -4 2\n",
            "1 3 4 -2\n",
            "1 3 4 2\n",
            "-1 -4 -2 -3\n",
            "-1 -4 -2 3\n",
            "-1 -4 2 -3\n",
            "-1 -4 2 3\n",
            "-1 4 -2 -3\n",
            "-1 4 -2 3\n",
            "-1 4 2 -3\n",
            "-1 4 2 3\n",
            "1 -4 -2 -3\n",
            "1 -4 -2 3\n",
            "1 -4 2 -3\n",
            "1 -4 2 3\n",
            "1 4 -2 -3\n",
            "1 4 -2 3\n",
            "1 4 2 -3\n",
            "1 4 2 3\n",
            "-1 -4 -3 -2\n",
            "-1 -4 -3 2\n",
            "-1 -4 3 -2\n",
            "-1 -4 3 2\n",
            "-1 4 -3 -2\n",
            "-1 4 -3 2\n",
            "-1 4 3 -2\n",
            "-1 4 3 2\n",
            "1 -4 -3 -2\n",
            "1 -4 -3 2\n",
            "1 -4 3 -2\n",
            "1 -4 3 2\n",
            "1 4 -3 -2\n",
            "1 4 -3 2\n",
            "1 4 3 -2\n",
            "1 4 3 2\n",
            "-2 -1 -3 -4\n",
            "-2 -1 -3 4\n",
            "-2 -1 3 -4\n",
            "-2 -1 3 4\n",
            "-2 1 -3 -4\n",
            "-2 1 -3 4\n",
            "-2 1 3 -4\n",
            "-2 1 3 4\n",
            "2 -1 -3 -4\n",
            "2 -1 -3 4\n",
            "2 -1 3 -4\n",
            "2 -1 3 4\n",
            "2 1 -3 -4\n",
            "2 1 -3 4\n",
            "2 1 3 -4\n",
            "2 1 3 4\n",
            "-2 -1 -4 -3\n",
            "-2 -1 -4 3\n",
            "-2 -1 4 -3\n",
            "-2 -1 4 3\n",
            "-2 1 -4 -3\n",
            "-2 1 -4 3\n",
            "-2 1 4 -3\n",
            "-2 1 4 3\n",
            "2 -1 -4 -3\n",
            "2 -1 -4 3\n",
            "2 -1 4 -3\n",
            "2 -1 4 3\n",
            "2 1 -4 -3\n",
            "2 1 -4 3\n",
            "2 1 4 -3\n",
            "2 1 4 3\n",
            "-2 -3 -1 -4\n",
            "-2 -3 -1 4\n",
            "-2 -3 1 -4\n",
            "-2 -3 1 4\n",
            "-2 3 -1 -4\n",
            "-2 3 -1 4\n",
            "-2 3 1 -4\n",
            "-2 3 1 4\n",
            "2 -3 -1 -4\n",
            "2 -3 -1 4\n",
            "2 -3 1 -4\n",
            "2 -3 1 4\n",
            "2 3 -1 -4\n",
            "2 3 -1 4\n",
            "2 3 1 -4\n",
            "2 3 1 4\n",
            "-2 -3 -4 -1\n",
            "-2 -3 -4 1\n",
            "-2 -3 4 -1\n",
            "-2 -3 4 1\n",
            "-2 3 -4 -1\n",
            "-2 3 -4 1\n",
            "-2 3 4 -1\n",
            "-2 3 4 1\n",
            "2 -3 -4 -1\n",
            "2 -3 -4 1\n",
            "2 -3 4 -1\n",
            "2 -3 4 1\n",
            "2 3 -4 -1\n",
            "2 3 -4 1\n",
            "2 3 4 -1\n",
            "2 3 4 1\n",
            "-2 -4 -1 -3\n",
            "-2 -4 -1 3\n",
            "-2 -4 1 -3\n",
            "-2 -4 1 3\n",
            "-2 4 -1 -3\n",
            "-2 4 -1 3\n",
            "-2 4 1 -3\n",
            "-2 4 1 3\n",
            "2 -4 -1 -3\n",
            "2 -4 -1 3\n",
            "2 -4 1 -3\n",
            "2 -4 1 3\n",
            "2 4 -1 -3\n",
            "2 4 -1 3\n",
            "2 4 1 -3\n",
            "2 4 1 3\n",
            "-2 -4 -3 -1\n",
            "-2 -4 -3 1\n",
            "-2 -4 3 -1\n",
            "-2 -4 3 1\n",
            "-2 4 -3 -1\n",
            "-2 4 -3 1\n",
            "-2 4 3 -1\n",
            "-2 4 3 1\n",
            "2 -4 -3 -1\n",
            "2 -4 -3 1\n",
            "2 -4 3 -1\n",
            "2 -4 3 1\n",
            "2 4 -3 -1\n",
            "2 4 -3 1\n",
            "2 4 3 -1\n",
            "2 4 3 1\n",
            "-3 -1 -2 -4\n",
            "-3 -1 -2 4\n",
            "-3 -1 2 -4\n",
            "-3 -1 2 4\n",
            "-3 1 -2 -4\n",
            "-3 1 -2 4\n",
            "-3 1 2 -4\n",
            "-3 1 2 4\n",
            "3 -1 -2 -4\n",
            "3 -1 -2 4\n",
            "3 -1 2 -4\n",
            "3 -1 2 4\n",
            "3 1 -2 -4\n",
            "3 1 -2 4\n",
            "3 1 2 -4\n",
            "3 1 2 4\n",
            "-3 -1 -4 -2\n",
            "-3 -1 -4 2\n",
            "-3 -1 4 -2\n",
            "-3 -1 4 2\n",
            "-3 1 -4 -2\n",
            "-3 1 -4 2\n",
            "-3 1 4 -2\n",
            "-3 1 4 2\n",
            "3 -1 -4 -2\n",
            "3 -1 -4 2\n",
            "3 -1 4 -2\n",
            "3 -1 4 2\n",
            "3 1 -4 -2\n",
            "3 1 -4 2\n",
            "3 1 4 -2\n",
            "3 1 4 2\n",
            "-3 -2 -1 -4\n",
            "-3 -2 -1 4\n",
            "-3 -2 1 -4\n",
            "-3 -2 1 4\n",
            "-3 2 -1 -4\n",
            "-3 2 -1 4\n",
            "-3 2 1 -4\n",
            "-3 2 1 4\n",
            "3 -2 -1 -4\n",
            "3 -2 -1 4\n",
            "3 -2 1 -4\n",
            "3 -2 1 4\n",
            "3 2 -1 -4\n",
            "3 2 -1 4\n",
            "3 2 1 -4\n",
            "3 2 1 4\n",
            "-3 -2 -4 -1\n",
            "-3 -2 -4 1\n",
            "-3 -2 4 -1\n",
            "-3 -2 4 1\n",
            "-3 2 -4 -1\n",
            "-3 2 -4 1\n",
            "-3 2 4 -1\n",
            "-3 2 4 1\n",
            "3 -2 -4 -1\n",
            "3 -2 -4 1\n",
            "3 -2 4 -1\n",
            "3 -2 4 1\n",
            "3 2 -4 -1\n",
            "3 2 -4 1\n",
            "3 2 4 -1\n",
            "3 2 4 1\n",
            "-3 -4 -1 -2\n",
            "-3 -4 -1 2\n",
            "-3 -4 1 -2\n",
            "-3 -4 1 2\n",
            "-3 4 -1 -2\n",
            "-3 4 -1 2\n",
            "-3 4 1 -2\n",
            "-3 4 1 2\n",
            "3 -4 -1 -2\n",
            "3 -4 -1 2\n",
            "3 -4 1 -2\n",
            "3 -4 1 2\n",
            "3 4 -1 -2\n",
            "3 4 -1 2\n",
            "3 4 1 -2\n",
            "3 4 1 2\n",
            "-3 -4 -2 -1\n",
            "-3 -4 -2 1\n",
            "-3 -4 2 -1\n",
            "-3 -4 2 1\n",
            "-3 4 -2 -1\n",
            "-3 4 -2 1\n",
            "-3 4 2 -1\n",
            "-3 4 2 1\n",
            "3 -4 -2 -1\n",
            "3 -4 -2 1\n",
            "3 -4 2 -1\n",
            "3 -4 2 1\n",
            "3 4 -2 -1\n",
            "3 4 -2 1\n",
            "3 4 2 -1\n",
            "3 4 2 1\n",
            "-4 -1 -2 -3\n",
            "-4 -1 -2 3\n",
            "-4 -1 2 -3\n",
            "-4 -1 2 3\n",
            "-4 1 -2 -3\n",
            "-4 1 -2 3\n",
            "-4 1 2 -3\n",
            "-4 1 2 3\n",
            "4 -1 -2 -3\n",
            "4 -1 -2 3\n",
            "4 -1 2 -3\n",
            "4 -1 2 3\n",
            "4 1 -2 -3\n",
            "4 1 -2 3\n",
            "4 1 2 -3\n",
            "4 1 2 3\n",
            "-4 -1 -3 -2\n",
            "-4 -1 -3 2\n",
            "-4 -1 3 -2\n",
            "-4 -1 3 2\n",
            "-4 1 -3 -2\n",
            "-4 1 -3 2\n",
            "-4 1 3 -2\n",
            "-4 1 3 2\n",
            "4 -1 -3 -2\n",
            "4 -1 -3 2\n",
            "4 -1 3 -2\n",
            "4 -1 3 2\n",
            "4 1 -3 -2\n",
            "4 1 -3 2\n",
            "4 1 3 -2\n",
            "4 1 3 2\n",
            "-4 -2 -1 -3\n",
            "-4 -2 -1 3\n",
            "-4 -2 1 -3\n",
            "-4 -2 1 3\n",
            "-4 2 -1 -3\n",
            "-4 2 -1 3\n",
            "-4 2 1 -3\n",
            "-4 2 1 3\n",
            "4 -2 -1 -3\n",
            "4 -2 -1 3\n",
            "4 -2 1 -3\n",
            "4 -2 1 3\n",
            "4 2 -1 -3\n",
            "4 2 -1 3\n",
            "4 2 1 -3\n",
            "4 2 1 3\n",
            "-4 -2 -3 -1\n",
            "-4 -2 -3 1\n",
            "-4 -2 3 -1\n",
            "-4 -2 3 1\n",
            "-4 2 -3 -1\n",
            "-4 2 -3 1\n",
            "-4 2 3 -1\n",
            "-4 2 3 1\n",
            "4 -2 -3 -1\n",
            "4 -2 -3 1\n",
            "4 -2 3 -1\n",
            "4 -2 3 1\n",
            "4 2 -3 -1\n",
            "4 2 -3 1\n",
            "4 2 3 -1\n",
            "4 2 3 1\n",
            "-4 -3 -1 -2\n",
            "-4 -3 -1 2\n",
            "-4 -3 1 -2\n",
            "-4 -3 1 2\n",
            "-4 3 -1 -2\n",
            "-4 3 -1 2\n",
            "-4 3 1 -2\n",
            "-4 3 1 2\n",
            "4 -3 -1 -2\n",
            "4 -3 -1 2\n",
            "4 -3 1 -2\n",
            "4 -3 1 2\n",
            "4 3 -1 -2\n",
            "4 3 -1 2\n",
            "4 3 1 -2\n",
            "4 3 1 2\n",
            "-4 -3 -2 -1\n",
            "-4 -3 -2 1\n",
            "-4 -3 2 -1\n",
            "-4 -3 2 1\n",
            "-4 3 -2 -1\n",
            "-4 3 -2 1\n",
            "-4 3 2 -1\n",
            "-4 3 2 1\n",
            "4 -3 -2 -1\n",
            "4 -3 -2 1\n",
            "4 -3 2 -1\n",
            "4 -3 2 1\n",
            "4 3 -2 -1\n",
            "4 3 -2 1\n",
            "4 3 2 -1\n",
            "4 3 2 1\n",
            "\n",
            "Results written to: output_sign.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Random Strings Problem"
      ],
      "metadata": {
        "id": "rNP1k8hIisU4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Introduction to Random Strings Problem Solution\n",
        "\n",
        "This module calculates the probability that random DNA strings with specified\n",
        "GC-content will exactly match a given target string. This helps determine\n",
        "whether DNA motifs occur due to functional significance or random chance.\n",
        "\n",
        "The problem models random genome generation and uses logarithmic probability\n",
        "calculations to handle very small probability values.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Dict, Tuple\n",
        "import math\n",
        "import re\n",
        "\n",
        "\n",
        "class RandomStringProbabilityCalculator:\n",
        "    \"\"\"\n",
        "    A class to calculate probabilities of DNA strings occurring randomly\n",
        "    given different GC-content values.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the random string probability calculator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_dna_sequence(sequence: str) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that sequence contains only valid DNA bases.\n",
        "\n",
        "        Args:\n",
        "            sequence (str): DNA sequence to validate\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid, False otherwise\n",
        "        \"\"\"\n",
        "        return bool(re.match(r'^[ATGC]*$', sequence.upper()))\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_gc_content(gc_content: float) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that GC-content is between 0 and 1.\n",
        "\n",
        "        Args:\n",
        "            gc_content (float): GC-content value to validate\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid, False otherwise\n",
        "        \"\"\"\n",
        "        return 0 <= gc_content <= 1\n",
        "\n",
        "    def count_nucleotides(self, dna_string: str) -> Dict[str, int]:\n",
        "        \"\"\"\n",
        "        Count occurrences of each nucleotide in the DNA string.\n",
        "\n",
        "        Args:\n",
        "            dna_string (str): DNA sequence\n",
        "\n",
        "        Returns:\n",
        "            Dict[str, int]: Dictionary with counts for A, T, G, C\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence contains invalid bases\n",
        "        \"\"\"\n",
        "        if not self.validate_dna_sequence(dna_string):\n",
        "            raise ValueError(\"DNA sequence contains invalid bases\")\n",
        "\n",
        "        sequence = dna_string.upper()\n",
        "        counts = {\n",
        "            'A': sequence.count('A'),\n",
        "            'T': sequence.count('T'),\n",
        "            'G': sequence.count('G'),\n",
        "            'C': sequence.count('C')\n",
        "        }\n",
        "\n",
        "        return counts\n",
        "\n",
        "    def calculate_nucleotide_probabilities(self, gc_content: float) -> Dict[str, float]:\n",
        "        \"\"\"\n",
        "        Calculate individual nucleotide probabilities given GC-content.\n",
        "\n",
        "        Args:\n",
        "            gc_content (float): GC-content (proportion between 0 and 1)\n",
        "\n",
        "        Returns:\n",
        "            Dict[str, float]: Probabilities for each nucleotide\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If GC-content is invalid\n",
        "        \"\"\"\n",
        "        if not self.validate_gc_content(gc_content):\n",
        "            raise ValueError(f\"Invalid GC-content: {gc_content}. Must be between 0 and 1\")\n",
        "\n",
        "        # GC-content is split equally between G and C\n",
        "        prob_g = gc_content / 2\n",
        "        prob_c = gc_content / 2\n",
        "\n",
        "        # Remaining probability is split equally between A and T\n",
        "        prob_a = (1 - gc_content) / 2\n",
        "        prob_t = (1 - gc_content) / 2\n",
        "\n",
        "        return {\n",
        "            'A': prob_a,\n",
        "            'T': prob_t,\n",
        "            'G': prob_g,\n",
        "            'C': prob_c\n",
        "        }\n",
        "\n",
        "    def calculate_string_probability(self, dna_string: str, gc_content: float) -> float:\n",
        "        \"\"\"\n",
        "        Calculate the probability of a DNA string occurring randomly\n",
        "        given a specific GC-content.\n",
        "\n",
        "        Args:\n",
        "            dna_string (str): Target DNA string\n",
        "            gc_content (float): GC-content for random model\n",
        "\n",
        "        Returns:\n",
        "            float: Probability of the string occurring randomly\n",
        "        \"\"\"\n",
        "        # Count nucleotides\n",
        "        counts = self.count_nucleotides(dna_string)\n",
        "\n",
        "        # Get nucleotide probabilities\n",
        "        probs = self.calculate_nucleotide_probabilities(gc_content)\n",
        "\n",
        "        # Calculate overall probability as product of individual probabilities\n",
        "        # P(string) = P(A)^count_A × P(T)^count_T × P(G)^count_G × P(C)^count_C\n",
        "        probability = (\n",
        "            (probs['A'] ** counts['A']) *\n",
        "            (probs['T'] ** counts['T']) *\n",
        "            (probs['G'] ** counts['G']) *\n",
        "            (probs['C'] ** counts['C'])\n",
        "        )\n",
        "\n",
        "        return probability\n",
        "\n",
        "    def calculate_log_probability(self, dna_string: str, gc_contents: List[float]) -> List[float]:\n",
        "        \"\"\"\n",
        "        Calculate log10 probabilities for multiple GC-content values.\n",
        "\n",
        "        Args:\n",
        "            dna_string (str): Target DNA string\n",
        "            gc_contents (List[float]): List of GC-content values\n",
        "\n",
        "        Returns:\n",
        "            List[float]: List of log10 probabilities\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If inputs are invalid\n",
        "        \"\"\"\n",
        "        if not self.validate_dna_sequence(dna_string):\n",
        "            raise ValueError(\"Invalid DNA sequence\")\n",
        "\n",
        "        if len(dna_string) > 100:\n",
        "            raise ValueError(f\"DNA string length {len(dna_string)} exceeds maximum of 100 bp\")\n",
        "\n",
        "        if len(gc_contents) > 20:\n",
        "            raise ValueError(f\"Number of GC-content values {len(gc_contents)} exceeds maximum of 20\")\n",
        "\n",
        "        log_probabilities = []\n",
        "\n",
        "        for gc in gc_contents:\n",
        "            if not self.validate_gc_content(gc):\n",
        "                raise ValueError(f\"Invalid GC-content: {gc}\")\n",
        "\n",
        "            # Calculate probability\n",
        "            probability = self.calculate_string_probability(dna_string, gc)\n",
        "\n",
        "            # Calculate log10 probability\n",
        "            if probability > 0:\n",
        "                log_prob = math.log10(probability)\n",
        "            else:\n",
        "                log_prob = float('-inf')  # Handle edge case of zero probability\n",
        "\n",
        "            log_probabilities.append(log_prob)\n",
        "\n",
        "        return log_probabilities\n",
        "\n",
        "    def get_detailed_analysis(self, dna_string: str, gc_contents: List[float]) -> Dict:\n",
        "        \"\"\"\n",
        "        Get detailed analysis of probability calculations.\n",
        "\n",
        "        Args:\n",
        "            dna_string (str): Target DNA string\n",
        "            gc_contents (List[float]): List of GC-content values\n",
        "\n",
        "        Returns:\n",
        "            Dict: Detailed analysis including counts, probabilities, and calculations\n",
        "        \"\"\"\n",
        "        sequence = dna_string.upper()\n",
        "        counts = self.count_nucleotides(sequence)\n",
        "        log_probs = self.calculate_log_probability(sequence, gc_contents)\n",
        "\n",
        "        analysis = {\n",
        "            'dna_string': sequence,\n",
        "            'length': len(sequence),\n",
        "            'nucleotide_counts': counts,\n",
        "            'gc_contents': gc_contents,\n",
        "            'num_gc_values': len(gc_contents),\n",
        "            'log_probabilities': log_probs,\n",
        "            'calculations': []\n",
        "        }\n",
        "\n",
        "        # Add detailed calculations for each GC-content\n",
        "        for i, gc in enumerate(gc_contents):\n",
        "            probs = self.calculate_nucleotide_probabilities(gc)\n",
        "            raw_prob = self.calculate_string_probability(sequence, gc)\n",
        "\n",
        "            calc_detail = {\n",
        "                'gc_content': gc,\n",
        "                'nucleotide_probs': probs,\n",
        "                'calculation': f\"({probs['A']:.6f})^{counts['A']} × ({probs['T']:.6f})^{counts['T']} × ({probs['G']:.6f})^{counts['G']} × ({probs['C']:.6f})^{counts['C']}\",\n",
        "                'raw_probability': raw_prob,\n",
        "                'log10_probability': log_probs[i]\n",
        "            }\n",
        "            analysis['calculations'].append(calc_detail)\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> Tuple[str, List[float]]:\n",
        "    \"\"\"\n",
        "    Parse input file to extract DNA string and GC-content array.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[str, List[float]]: (dna_string, gc_contents)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            lines = file.read().strip().split('\\n')\n",
        "\n",
        "        if len(lines) < 2:\n",
        "            raise ValueError(\"Input file must contain at least 2 lines\")\n",
        "\n",
        "        # First line: DNA string\n",
        "        dna_string = lines[0].strip()\n",
        "\n",
        "        # Second line: GC-content values\n",
        "        try:\n",
        "            gc_contents = list(map(float, lines[1].strip().split()))\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Second line must contain valid numbers, got: '{lines[1]}'\")\n",
        "\n",
        "        return dna_string, gc_contents\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, log_probabilities: List[float]) -> None:\n",
        "    \"\"\"\n",
        "    Write log probabilities to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        log_probabilities (List[float]): List of log10 probabilities\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            # Format to 3 decimal places\n",
        "            formatted_probs = [f\"{prob:.3f}\" for prob in log_probabilities]\n",
        "            file.write(' '.join(formatted_probs) + '\\n')\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_random_strings_problem(input_file_path: str) -> List[float]:\n",
        "    \"\"\"\n",
        "    Solve the Random Strings problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        List[float]: List of log10 probabilities\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        dna_string, gc_contents = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize calculator\n",
        "        calculator = RandomStringProbabilityCalculator()\n",
        "\n",
        "        # Calculate log probabilities\n",
        "        log_probabilities = calculator.calculate_log_probability(dna_string, gc_contents)\n",
        "\n",
        "        return log_probabilities\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Random Strings problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_prob.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_prob.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Random Strings Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        log_probabilities = solve_random_strings_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResults:\")\n",
        "        formatted_results = [f\"{prob:.3f}\" for prob in log_probabilities]\n",
        "        print(' '.join(formatted_results))\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, log_probabilities)\n",
        "        print(f\"\\nResults written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data from search results.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample data from the problem and search results\n",
        "    sample_dna = \"ACGATACAA\"\n",
        "    sample_gc_contents = [0.129, 0.287, 0.423, 0.476, 0.641, 0.742, 0.783]\n",
        "\n",
        "    print(f\"Input DNA string: {sample_dna}\")\n",
        "    print(f\"Length: {len(sample_dna)} bp\")\n",
        "    print(f\"GC-content values: {sample_gc_contents}\")\n",
        "\n",
        "    calculator = RandomStringProbabilityCalculator()\n",
        "\n",
        "    # Get detailed analysis\n",
        "    analysis = calculator.get_detailed_analysis(sample_dna, sample_gc_contents)\n",
        "\n",
        "    print(f\"\\nNucleotide counts: {analysis['nucleotide_counts']}\")\n",
        "    print(f\"Number of GC-content values: {analysis['num_gc_values']}\")\n",
        "\n",
        "    # Show a few detailed calculations\n",
        "    print(f\"\\nDetailed calculations (first 3):\")\n",
        "    for i in range(min(3, len(analysis['calculations']))):\n",
        "        calc = analysis['calculations'][i]\n",
        "        print(f\"GC-content {calc['gc_content']:.3f}:\")\n",
        "        print(f\"  Nucleotide probabilities: {calc['nucleotide_probs']}\")\n",
        "        print(f\"  Raw probability: {calc['raw_probability']:.2e}\")\n",
        "        print(f\"  Log10 probability: {calc['log10_probability']:.3f}\")\n",
        "\n",
        "    # Final results\n",
        "    result = calculator.calculate_log_probability(sample_dna, sample_gc_contents)\n",
        "    formatted_result = ' '.join(f\"{x:.3f}\" for x in result)\n",
        "\n",
        "    print(f\"\\nFinal results:\")\n",
        "    print(formatted_result)\n",
        "\n",
        "    # Expected output verification (from search results)\n",
        "    expected = \"-5.737 -5.217 -5.263 -5.360 -5.958 -6.628 -7.009\"\n",
        "    print(f\"\\nVerification:\")\n",
        "    print(f\"Expected: {expected}\")\n",
        "    print(f\"Our result: {formatted_result}\")\n",
        "    print(f\"Match: {'✓' if formatted_result == expected else '✗'}\")\n",
        "\n",
        "\n",
        "def demonstrate_probability_concepts():\n",
        "    \"\"\"\n",
        "    Demonstrate the probability concepts used in this problem.\n",
        "    \"\"\"\n",
        "    print(\"=== Probability Concepts Demonstration ===\")\n",
        "\n",
        "    calculator = RandomStringProbabilityCalculator()\n",
        "\n",
        "    print(\"Key concepts:\")\n",
        "    print(\"1. For GC-content x: P(G) = P(C) = x/2, P(A) = P(T) = (1-x)/2\")\n",
        "    print(\"2. String probability = ∏(P(nucleotide)^count)\")\n",
        "    print(\"3. Log probabilities help handle very small values\")\n",
        "    print()\n",
        "\n",
        "    # Example with simple string\n",
        "    test_string = \"ATGC\"\n",
        "    test_gc = 0.4\n",
        "\n",
        "    print(f\"Example: String '{test_string}' with GC-content {test_gc}\")\n",
        "\n",
        "    counts = calculator.count_nucleotides(test_string)\n",
        "    probs = calculator.calculate_nucleotide_probabilities(test_gc)\n",
        "    raw_prob = calculator.calculate_string_probability(test_string, test_gc)\n",
        "    log_prob = math.log10(raw_prob)\n",
        "\n",
        "    print(f\"Nucleotide counts: {counts}\")\n",
        "    print(f\"Nucleotide probabilities: {probs}\")\n",
        "    print(f\"Calculation: {probs['A']:.3f}^{counts['A']} × {probs['T']:.3f}^{counts['T']} × {probs['G']:.3f}^{counts['G']} × {probs['C']:.3f}^{counts['C']}\")\n",
        "    print(f\"Raw probability: {raw_prob:.6e}\")\n",
        "    print(f\"Log10 probability: {log_prob:.3f}\")\n",
        "\n",
        "\n",
        "def test_edge_cases():\n",
        "    \"\"\"\n",
        "    Test various edge cases and validation.\n",
        "    \"\"\"\n",
        "    print(\"=== Testing Edge Cases ===\")\n",
        "\n",
        "    calculator = RandomStringProbabilityCalculator()\n",
        "\n",
        "    test_cases = [\n",
        "        (\"A\", [0.5], \"Single nucleotide\"),\n",
        "        (\"AT\", [0.0], \"Zero GC-content\"),\n",
        "        (\"GC\", [1.0], \"100% GC-content\"),\n",
        "        (\"AAAA\", [0.25, 0.5, 0.75], \"Homopolymer\"),\n",
        "        (\"ATGCATGC\", [0.5], \"Balanced composition\"),\n",
        "    ]\n",
        "\n",
        "    for dna_string, gc_contents, description in test_cases:\n",
        "        try:\n",
        "            result = calculator.calculate_log_probability(dna_string, gc_contents)\n",
        "            formatted_result = ' '.join(f\"{x:.3f}\" for x in result)\n",
        "            print(f\"{description}: '{dna_string}' -> {formatted_result}\")\n",
        "        except Exception as e:\n",
        "            print(f\"{description}: '{dna_string}' -> Error: {e}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate probability concepts\n",
        "    demonstrate_probability_concepts()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test edge cases\n",
        "    test_edge_cases()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "wJgMsvy_itCu",
        "outputId": "a0d8490c-8b33-473a-9be8-ba3b136a5825",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input DNA string: ACGATACAA\n",
            "Length: 9 bp\n",
            "GC-content values: [0.129, 0.287, 0.423, 0.476, 0.641, 0.742, 0.783]\n",
            "\n",
            "Nucleotide counts: {'A': 5, 'T': 1, 'G': 1, 'C': 2}\n",
            "Number of GC-content values: 7\n",
            "\n",
            "Detailed calculations (first 3):\n",
            "GC-content 0.129:\n",
            "  Nucleotide probabilities: {'A': 0.4355, 'T': 0.4355, 'G': 0.0645, 'C': 0.0645}\n",
            "  Raw probability: 1.83e-06\n",
            "  Log10 probability: -5.737\n",
            "GC-content 0.287:\n",
            "  Nucleotide probabilities: {'A': 0.35650000000000004, 'T': 0.35650000000000004, 'G': 0.1435, 'C': 0.1435}\n",
            "  Raw probability: 6.07e-06\n",
            "  Log10 probability: -5.217\n",
            "GC-content 0.423:\n",
            "  Nucleotide probabilities: {'A': 0.2885, 'T': 0.2885, 'G': 0.2115, 'C': 0.2115}\n",
            "  Raw probability: 5.46e-06\n",
            "  Log10 probability: -5.263\n",
            "\n",
            "Final results:\n",
            "-5.737 -5.217 -5.263 -5.360 -5.958 -6.628 -7.009\n",
            "\n",
            "Verification:\n",
            "Expected: -5.737 -5.217 -5.263 -5.360 -5.958 -6.628 -7.009\n",
            "Our result: -5.737 -5.217 -5.263 -5.360 -5.958 -6.628 -7.009\n",
            "Match: ✓\n",
            "\n",
            "============================================================\n",
            "=== Probability Concepts Demonstration ===\n",
            "Key concepts:\n",
            "1. For GC-content x: P(G) = P(C) = x/2, P(A) = P(T) = (1-x)/2\n",
            "2. String probability = ∏(P(nucleotide)^count)\n",
            "3. Log probabilities help handle very small values\n",
            "\n",
            "Example: String 'ATGC' with GC-content 0.4\n",
            "Nucleotide counts: {'A': 1, 'T': 1, 'G': 1, 'C': 1}\n",
            "Nucleotide probabilities: {'A': 0.3, 'T': 0.3, 'G': 0.2, 'C': 0.2}\n",
            "Calculation: 0.300^1 × 0.300^1 × 0.200^1 × 0.200^1\n",
            "Raw probability: 3.600000e-03\n",
            "Log10 probability: -2.444\n",
            "\n",
            "============================================================\n",
            "=== Testing Edge Cases ===\n",
            "Single nucleotide: 'A' -> -0.602\n",
            "Zero GC-content: 'AT' -> -0.602\n",
            "100% GC-content: 'GC' -> -0.602\n",
            "Homopolymer: 'AAAA' -> -1.704 -2.408 -3.612\n",
            "Balanced composition: 'ATGCATGC' -> -4.816\n",
            "\n",
            "============================================================\n",
            "Solving Random Strings Problem...\n",
            "\n",
            "Results:\n",
            "-91.165 -72.725 -70.123 -67.276 -62.263 -62.032 -59.951 -59.646 -59.594 -59.866 -61.402 -62.129 -65.635 -66.264 -72.725 -79.110 -84.661\n",
            "\n",
            "Results written to: output_prob.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Partial Permutations Problem"
      ],
      "metadata": {
        "id": "vYO57_zDeRkn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Partial Permutations Problem Solution\n",
        "\n",
        "This module calculates the number of partial permutations P(n,k), which represents\n",
        "the number of ways to arrange k objects from a collection of n objects where order\n",
        "matters. This is useful for comparing partial gene orderings between species.\n",
        "\n",
        "The formula is P(n,k) = n!/(n-k)! = n × (n-1) × ... × (n-k+1)\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import Tuple\n",
        "import math\n",
        "\n",
        "\n",
        "class PartialPermutationCalculator:\n",
        "    \"\"\"\n",
        "    A class to calculate partial permutations with modular arithmetic.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, modulo: int = 1000000):\n",
        "        \"\"\"\n",
        "        Initialize the partial permutation calculator.\n",
        "\n",
        "        Args:\n",
        "            modulo (int): Modulo value for calculations (default: 1,000,000)\n",
        "        \"\"\"\n",
        "        self.modulo = modulo\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_input(n: int, k: int) -> bool:\n",
        "        \"\"\"\n",
        "        Validate input parameters according to problem constraints.\n",
        "\n",
        "        Args:\n",
        "            n (int): Total number of objects\n",
        "            k (int): Number of objects to arrange\n",
        "\n",
        "        Returns:\n",
        "            bool: True if input is valid, False otherwise\n",
        "        \"\"\"\n",
        "        # Check constraints: 100 ≥ n > 0 and 10 ≥ k > 0 and k ≤ n\n",
        "        if not (0 < n <= 100):\n",
        "            return False\n",
        "        if not (0 < k <= 10):\n",
        "            return False\n",
        "        if k > n:\n",
        "            return False\n",
        "        return True\n",
        "\n",
        "    def calculate_partial_permutations_factorial(self, n: int, k: int) -> int:\n",
        "        \"\"\"\n",
        "        Calculate P(n,k) using factorial formula: n!/(n-k)!\n",
        "\n",
        "        Args:\n",
        "            n (int): Total number of objects\n",
        "            k (int): Number of objects to arrange\n",
        "\n",
        "        Returns:\n",
        "            int: P(n,k) modulo self.modulo\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If input parameters are invalid\n",
        "        \"\"\"\n",
        "        if not self.validate_input(n, k):\n",
        "            raise ValueError(f\"Invalid input: n={n}, k={k}. \"\n",
        "                           f\"Required: 0 < n ≤ 100, 0 < k ≤ 10, k ≤ n\")\n",
        "\n",
        "        # Calculate using factorial formula\n",
        "        numerator = math.factorial(n)\n",
        "        denominator = math.factorial(n - k)\n",
        "        result = (numerator // denominator) % self.modulo\n",
        "\n",
        "        return result\n",
        "\n",
        "    def calculate_partial_permutations_iterative(self, n: int, k: int) -> int:\n",
        "        \"\"\"\n",
        "        Calculate P(n,k) using iterative multiplication: n × (n-1) × ... × (n-k+1)\n",
        "        This approach is more efficient and avoids large intermediate values.\n",
        "\n",
        "        Args:\n",
        "            n (int): Total number of objects\n",
        "            k (int): Number of objects to arrange\n",
        "\n",
        "        Returns:\n",
        "            int: P(n,k) modulo self.modulo\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If input parameters are invalid\n",
        "        \"\"\"\n",
        "        if not self.validate_input(n, k):\n",
        "            raise ValueError(f\"Invalid input: n={n}, k={k}. \"\n",
        "                           f\"Required: 0 < n ≤ 100, 0 < k ≤ 10, k ≤ n\")\n",
        "\n",
        "        # Calculate P(n,k) = n × (n-1) × (n-2) × ... × (n-k+1)\n",
        "        result = 1\n",
        "        for i in range(n, n - k, -1):\n",
        "            result = (result * i) % self.modulo\n",
        "\n",
        "        return result\n",
        "\n",
        "    def calculate_partial_permutations(self, n: int, k: int, method: str = \"iterative\") -> int:\n",
        "        \"\"\"\n",
        "        Calculate P(n,k) using specified method.\n",
        "\n",
        "        Args:\n",
        "            n (int): Total number of objects\n",
        "            k (int): Number of objects to arrange\n",
        "            method (str): Calculation method - \"iterative\" or \"factorial\"\n",
        "\n",
        "        Returns:\n",
        "            int: P(n,k) modulo self.modulo\n",
        "        \"\"\"\n",
        "        if method == \"iterative\":\n",
        "            return self.calculate_partial_permutations_iterative(n, k)\n",
        "        elif method == \"factorial\":\n",
        "            return self.calculate_partial_permutations_factorial(n, k)\n",
        "        else:\n",
        "            raise ValueError(\"Method must be 'iterative' or 'factorial'\")\n",
        "\n",
        "    def get_calculation_details(self, n: int, k: int) -> dict:\n",
        "        \"\"\"\n",
        "        Get detailed information about the partial permutation calculation.\n",
        "\n",
        "        Args:\n",
        "            n (int): Total number of objects\n",
        "            k (int): Number of objects to arrange\n",
        "\n",
        "        Returns:\n",
        "            dict: Detailed calculation information\n",
        "        \"\"\"\n",
        "        if not self.validate_input(n, k):\n",
        "            raise ValueError(f\"Invalid input: n={n}, k={k}\")\n",
        "\n",
        "        # Calculate using both methods for comparison\n",
        "        result_iterative = self.calculate_partial_permutations_iterative(n, k)\n",
        "        result_factorial = self.calculate_partial_permutations_factorial(n, k)\n",
        "\n",
        "        # Build step-by-step calculation for iterative method\n",
        "        steps = []\n",
        "        temp_result = 1\n",
        "        for i in range(n, n - k, -1):\n",
        "            temp_result = (temp_result * i) % self.modulo\n",
        "            steps.append(f\"× {i} = {temp_result}\")\n",
        "\n",
        "        details = {\n",
        "            'n': n,\n",
        "            'k': k,\n",
        "            'formula': f\"P({n},{k}) = {n}!/{n-k}!\",\n",
        "            'iterative_formula': f\"P({n},{k}) = {' × '.join(str(i) for i in range(n, n-k, -1))}\",\n",
        "            'calculation_steps': steps,\n",
        "            'result_iterative': result_iterative,\n",
        "            'result_factorial': result_factorial,\n",
        "            'methods_agree': result_iterative == result_factorial,\n",
        "            'modulo': self.modulo,\n",
        "            'final_result': result_iterative\n",
        "        }\n",
        "\n",
        "        return details\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> Tuple[int, int]:\n",
        "    \"\"\"\n",
        "    Parse input file to extract n and k.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[int, int]: (n, k)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        # Parse n and k from single line\n",
        "        values = content.split()\n",
        "        if len(values) != 2:\n",
        "            raise ValueError(f\"Input file must contain exactly 2 integers, found {len(values)}\")\n",
        "\n",
        "        try:\n",
        "            n = int(values[0])\n",
        "            k = int(values[1])\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Input must contain integers, got: '{content}'\")\n",
        "\n",
        "        return n, k\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, result: int) -> None:\n",
        "    \"\"\"\n",
        "    Write result to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        result (int): Partial permutation count\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            file.write(f\"{result}\\n\")\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_partial_permutations_problem(input_file_path: str) -> int:\n",
        "    \"\"\"\n",
        "    Solve the Partial Permutations problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        int: Number of partial permutations modulo 1,000,000\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        n, k = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize calculator\n",
        "        calculator = PartialPermutationCalculator()\n",
        "\n",
        "        # Calculate partial permutations\n",
        "        result = calculator.calculate_partial_permutations(n, k)\n",
        "\n",
        "        return result\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Partial Permutations problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_pper.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_pper.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Partial Permutations Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        result = solve_partial_permutations_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResult:\")\n",
        "        print(f\"P(n,k) modulo 1,000,000 = {result:,}\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, result)\n",
        "        print(f\"\\nResult written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input from the problem\n",
        "    sample_n = 21\n",
        "    sample_k = 7\n",
        "\n",
        "    print(f\"Input: n = {sample_n}, k = {sample_k}\")\n",
        "\n",
        "    calculator = PartialPermutationCalculator()\n",
        "\n",
        "    # Get detailed calculation\n",
        "    details = calculator.get_calculation_details(sample_n, sample_k)\n",
        "\n",
        "    print(f\"\\nDetailed Calculation:\")\n",
        "    print(f\"Formula: {details['formula']}\")\n",
        "    print(f\"Iterative: {details['iterative_formula']}\")\n",
        "    print(f\"Result: {details['final_result']:,}\")\n",
        "\n",
        "    # Expected output verification\n",
        "    expected = 51200\n",
        "    print(f\"\\nVerification:\")\n",
        "    print(f\"Expected: {expected:,}\")\n",
        "    print(f\"Our result: {details['final_result']:,}\")\n",
        "    print(f\"Match: {'✓' if details['final_result'] == expected else '✗'}\")\n",
        "\n",
        "    # Show both calculation methods agree\n",
        "    print(f\"Methods agree: {'✓' if details['methods_agree'] else '✗'}\")\n",
        "\n",
        "\n",
        "def demonstrate_calculation_methods():\n",
        "    \"\"\"\n",
        "    Demonstrate different calculation methods and their efficiency.\n",
        "    \"\"\"\n",
        "    print(\"=== Calculation Methods Demonstration ===\")\n",
        "\n",
        "    calculator = PartialPermutationCalculator()\n",
        "\n",
        "    # Test case\n",
        "    n, k = 21, 7\n",
        "\n",
        "    print(f\"Calculating P({n},{k}):\")\n",
        "\n",
        "    # Method 1: Factorial approach\n",
        "    print(f\"\\nMethod 1 - Factorial: P({n},{k}) = {n}!/({n-k})! = {n}!/{n-k}!\")\n",
        "    result1 = calculator.calculate_partial_permutations(n, k, \"factorial\")\n",
        "    print(f\"Result: {result1:,}\")\n",
        "\n",
        "    # Method 2: Iterative approach\n",
        "    print(f\"\\nMethod 2 - Iterative: P({n},{k}) = {n} × {n-1} × ... × {n-k+1}\")\n",
        "    result2 = calculator.calculate_partial_permutations(n, k, \"iterative\")\n",
        "\n",
        "    # Show step-by-step for iterative\n",
        "    steps = []\n",
        "    temp = 1\n",
        "    for i in range(n, n - k, -1):\n",
        "        temp = (temp * i) % 1000000\n",
        "        steps.append(f\"{i}\")\n",
        "\n",
        "    print(f\"Calculation: {' × '.join(steps)}\")\n",
        "    print(f\"Result: {result2:,}\")\n",
        "\n",
        "    print(f\"\\nBoth methods give same result: {'✓' if result1 == result2 else '✗'}\")\n",
        "\n",
        "\n",
        "def test_various_cases():\n",
        "    \"\"\"\n",
        "    Test the solution with various input cases.\n",
        "    \"\"\"\n",
        "    print(\"=== Testing Various Cases ===\")\n",
        "\n",
        "    calculator = PartialPermutationCalculator()\n",
        "\n",
        "    test_cases = [\n",
        "        (1, 1, \"Minimal case\"),\n",
        "        (5, 2, \"Small case\"),\n",
        "        (10, 3, \"Medium case\"),\n",
        "        (21, 7, \"Sample case\"),\n",
        "        (100, 10, \"Maximum case\"),\n",
        "        (50, 5, \"Moderate case\"),\n",
        "    ]\n",
        "\n",
        "    for n, k, description in test_cases:\n",
        "        try:\n",
        "            result = calculator.calculate_partial_permutations(n, k)\n",
        "            theoretical = math.factorial(n) // math.factorial(n - k)\n",
        "            theoretical_mod = theoretical % 1000000\n",
        "\n",
        "            print(f\"{description}: P({n},{k}) = {result:,}\")\n",
        "            print(f\"  Verification: {theoretical_mod:,} {'✓' if result == theoretical_mod else '✗'}\")\n",
        "        except Exception as e:\n",
        "            print(f\"{description}: P({n},{k}) -> Error: {e}\")\n",
        "\n",
        "\n",
        "def explain_biological_relevance():\n",
        "    \"\"\"\n",
        "    Explain the biological relevance of partial permutations.\n",
        "    \"\"\"\n",
        "    print(\"=== Biological Relevance ===\")\n",
        "    print(\"Partial permutations in genomics:\")\n",
        "    print(\"• Model gene arrangements when species share only some genes\")\n",
        "    print(\"• Compare evolutionary relationships between organisms\")\n",
        "    print(\"• Analyze chromosomal rearrangements over time\")\n",
        "    print(\"• Study synteny blocks in comparative genomics\")\n",
        "    print()\n",
        "    print(\"Example: If two species share 21 genes, and we want to analyze\")\n",
        "    print(\"arrangements of any 7 genes, there are P(21,7) = 51,200 possible\")\n",
        "    print(\"partial orderings to consider in evolutionary comparisons.\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate calculation methods\n",
        "    demonstrate_calculation_methods()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test various cases\n",
        "    test_various_cases()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Explain biological relevance\n",
        "    explain_biological_relevance()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "TGExzRomeYad",
        "outputId": "02fddece-8e1e-44f1-987b-cafca3fe47aa",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input: n = 21, k = 7\n",
            "\n",
            "Detailed Calculation:\n",
            "Formula: P(21,7) = 21!/14!\n",
            "Iterative: P(21,7) = 21 × 20 × 19 × 18 × 17 × 16 × 15\n",
            "Result: 51,200\n",
            "\n",
            "Verification:\n",
            "Expected: 51,200\n",
            "Our result: 51,200\n",
            "Match: ✓\n",
            "Methods agree: ✓\n",
            "\n",
            "============================================================\n",
            "=== Calculation Methods Demonstration ===\n",
            "Calculating P(21,7):\n",
            "\n",
            "Method 1 - Factorial: P(21,7) = 21!/(14)! = 21!/14!\n",
            "Result: 51,200\n",
            "\n",
            "Method 2 - Iterative: P(21,7) = 21 × 20 × ... × 15\n",
            "Calculation: 21 × 20 × 19 × 18 × 17 × 16 × 15\n",
            "Result: 51,200\n",
            "\n",
            "Both methods give same result: ✓\n",
            "\n",
            "============================================================\n",
            "=== Testing Various Cases ===\n",
            "Minimal case: P(1,1) = 1\n",
            "  Verification: 1 ✓\n",
            "Small case: P(5,2) = 20\n",
            "  Verification: 20 ✓\n",
            "Medium case: P(10,3) = 720\n",
            "  Verification: 720 ✓\n",
            "Sample case: P(21,7) = 51,200\n",
            "  Verification: 51,200 ✓\n",
            "Maximum case: P(100,10) = 472,000\n",
            "  Verification: 472,000 ✓\n",
            "Moderate case: P(50,5) = 251,200\n",
            "  Verification: 251,200 ✓\n",
            "\n",
            "============================================================\n",
            "=== Biological Relevance ===\n",
            "Partial permutations in genomics:\n",
            "• Model gene arrangements when species share only some genes\n",
            "• Compare evolutionary relationships between organisms\n",
            "• Analyze chromosomal rearrangements over time\n",
            "• Study synteny blocks in comparative genomics\n",
            "\n",
            "Example: If two species share 21 genes, and we want to analyze\n",
            "arrangements of any 7 genes, there are P(21,7) = 51,200 possible\n",
            "partial orderings to consider in evolutionary comparisons.\n",
            "\n",
            "============================================================\n",
            "Solving Partial Permutations Problem...\n",
            "\n",
            "Result:\n",
            "P(n,k) modulo 1,000,000 = 22,400\n",
            "\n",
            "Result written to: output_pper.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Perfect Matchings and RNA Secondary Structures"
      ],
      "metadata": {
        "id": "tltOWIpNadxD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Perfect Matchings and RNA Secondary Structures Problem Solution\n",
        "\n",
        "This module calculates the number of perfect matchings of basepair edges in an RNA\n",
        "bonding graph. This represents the total number of possible secondary structures\n",
        "where every nucleotide forms a base pair (A-U and C-G).\n",
        "\n",
        "The problem reduces to finding perfect matchings in complete bipartite graphs\n",
        "for A-U pairs and C-G pairs independently.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import Dict, Tuple, Optional\n",
        "import math\n",
        "import re\n",
        "\n",
        "\n",
        "class RNAMatchingCalculator:\n",
        "    \"\"\"\n",
        "    A class to calculate perfect matchings in RNA secondary structures.\n",
        "    \"\"\"\n",
        "\n",
        "    # Valid RNA bases\n",
        "    VALID_BASES = {'A', 'U', 'G', 'C'}\n",
        "\n",
        "    # Base pairing rules\n",
        "    BASE_PAIRS = {\n",
        "        ('A', 'U'), ('U', 'A'),\n",
        "        ('G', 'C'), ('C', 'G')\n",
        "    }\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the RNA matching calculator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_rna_sequence(sequence: str) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that sequence contains only valid RNA bases.\n",
        "\n",
        "        Args:\n",
        "            sequence (str): RNA sequence to validate\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid, False otherwise\n",
        "        \"\"\"\n",
        "        return bool(re.match(r'^[AUGC]*$', sequence.upper()))\n",
        "\n",
        "    def count_nucleotides(self, rna_sequence: str) -> Dict[str, int]:\n",
        "        \"\"\"\n",
        "        Count occurrences of each nucleotide in the RNA sequence.\n",
        "\n",
        "        Args:\n",
        "            rna_sequence (str): RNA sequence\n",
        "\n",
        "        Returns:\n",
        "            Dict[str, int]: Dictionary with counts for each nucleotide\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence contains invalid bases\n",
        "        \"\"\"\n",
        "        if not self.validate_rna_sequence(rna_sequence):\n",
        "            raise ValueError(\"RNA sequence contains invalid bases\")\n",
        "\n",
        "        sequence = rna_sequence.upper()\n",
        "        counts = {'A': 0, 'U': 0, 'G': 0, 'C': 0}\n",
        "\n",
        "        for base in sequence:\n",
        "            if base in counts:\n",
        "                counts[base] += 1\n",
        "\n",
        "        return counts\n",
        "\n",
        "    def validate_perfect_matching_possible(self, counts: Dict[str, int]) -> bool:\n",
        "        \"\"\"\n",
        "        Check if perfect matching is possible given nucleotide counts.\n",
        "\n",
        "        Args:\n",
        "            counts (Dict[str, int]): Nucleotide count dictionary\n",
        "\n",
        "        Returns:\n",
        "            bool: True if perfect matching is possible, False otherwise\n",
        "        \"\"\"\n",
        "        # For perfect matching: #A must equal #U, and #G must equal #C\n",
        "        return counts['A'] == counts['U'] and counts['G'] == counts['C']\n",
        "\n",
        "    def calculate_perfect_matchings(self, rna_sequence: str) -> int:\n",
        "        \"\"\"\n",
        "        Calculate the total number of perfect matchings for an RNA sequence.\n",
        "\n",
        "        Args:\n",
        "            rna_sequence (str): RNA sequence\n",
        "\n",
        "        Returns:\n",
        "            int: Number of perfect matchings\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence is invalid or perfect matching impossible\n",
        "        \"\"\"\n",
        "        # Count nucleotides\n",
        "        counts = self.count_nucleotides(rna_sequence)\n",
        "\n",
        "        # Validate that perfect matching is possible\n",
        "        if not self.validate_perfect_matching_possible(counts):\n",
        "            raise ValueError(\n",
        "                f\"Perfect matching not possible. A:{counts['A']}, U:{counts['U']}, \"\n",
        "                f\"G:{counts['G']}, C:{counts['C']}. Need A=U and G=C.\"\n",
        "            )\n",
        "\n",
        "        # Calculate perfect matchings\n",
        "        # Number of ways to match A's with U's: (# of A's)!\n",
        "        # Number of ways to match G's with C's: (# of G's)!\n",
        "        # Total combinations: (# of A's)! × (# of G's)!\n",
        "\n",
        "        num_au_pairs = counts['A']  # equals counts['U']\n",
        "        num_gc_pairs = counts['G']  # equals counts['C']\n",
        "\n",
        "        au_matchings = math.factorial(num_au_pairs)\n",
        "        gc_matchings = math.factorial(num_gc_pairs)\n",
        "\n",
        "        total_matchings = au_matchings * gc_matchings\n",
        "\n",
        "        return total_matchings\n",
        "\n",
        "    def get_detailed_analysis(self, rna_sequence: str) -> Dict:\n",
        "        \"\"\"\n",
        "        Get detailed analysis of the RNA sequence and its matching possibilities.\n",
        "\n",
        "        Args:\n",
        "            rna_sequence (str): RNA sequence\n",
        "\n",
        "        Returns:\n",
        "            Dict: Detailed analysis including counts, matchings, and statistics\n",
        "        \"\"\"\n",
        "        sequence = rna_sequence.upper()\n",
        "        counts = self.count_nucleotides(sequence)\n",
        "\n",
        "        analysis = {\n",
        "            'sequence': sequence,\n",
        "            'length': len(sequence),\n",
        "            'nucleotide_counts': counts,\n",
        "            'au_pairs_possible': counts['A'],\n",
        "            'gc_pairs_possible': counts['G'],\n",
        "            'total_pairs': counts['A'] + counts['G'],\n",
        "            'is_perfect_matching_possible': self.validate_perfect_matching_possible(counts)\n",
        "        }\n",
        "\n",
        "        if analysis['is_perfect_matching_possible']:\n",
        "            au_matchings = math.factorial(counts['A'])\n",
        "            gc_matchings = math.factorial(counts['G'])\n",
        "            total_matchings = au_matchings * gc_matchings\n",
        "\n",
        "            analysis.update({\n",
        "                'au_matchings': au_matchings,\n",
        "                'gc_matchings': gc_matchings,\n",
        "                'total_perfect_matchings': total_matchings,\n",
        "                'au_factorial': f\"{counts['A']}! = {au_matchings}\",\n",
        "                'gc_factorial': f\"{counts['G']}! = {gc_matchings}\",\n",
        "                'calculation': f\"{counts['A']}! × {counts['G']}! = {au_matchings} × {gc_matchings} = {total_matchings}\"\n",
        "            })\n",
        "        else:\n",
        "            analysis.update({\n",
        "                'error': 'Perfect matching not possible with given nucleotide counts'\n",
        "            })\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_fasta_file(file_path: str) -> Tuple[str, str]:\n",
        "    \"\"\"\n",
        "    Parse FASTA file to extract header and RNA sequence.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to FASTA file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[str, str]: (header, rna_sequence)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        lines = content.split('\\n')\n",
        "        if not lines or not lines[0].startswith('>'):\n",
        "            raise ValueError(\"Invalid FASTA format: missing header\")\n",
        "\n",
        "        header = lines[0][1:]  # Remove '>' character\n",
        "        sequence = ''.join(line.strip() for line in lines[1:] if not line.startswith('>'))\n",
        "\n",
        "        if not sequence:\n",
        "            raise ValueError(\"Invalid FASTA format: no sequence found\")\n",
        "\n",
        "        # Validate sequence length constraint\n",
        "        if len(sequence) > 80:\n",
        "            raise ValueError(f\"Sequence length {len(sequence)} exceeds maximum of 80 bp\")\n",
        "\n",
        "        return header, sequence\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, num_matchings: int) -> None:\n",
        "    \"\"\"\n",
        "    Write number of perfect matchings to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        num_matchings (int): Number of perfect matchings\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            file.write(f\"{num_matchings}\\n\")\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_rna_matching_problem(input_file_path: str) -> int:\n",
        "    \"\"\"\n",
        "    Solve the RNA Perfect Matching problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input FASTA file\n",
        "\n",
        "    Returns:\n",
        "        int: Number of perfect matchings\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse FASTA file\n",
        "        header, rna_sequence = parse_fasta_file(input_file_path)\n",
        "\n",
        "        # Initialize calculator\n",
        "        calculator = RNAMatchingCalculator()\n",
        "\n",
        "        # Calculate perfect matchings\n",
        "        num_matchings = calculator.calculate_perfect_matchings(rna_sequence)\n",
        "\n",
        "        return num_matchings\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the RNA Perfect Matching problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_pmch.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_pmch.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving RNA Perfect Matching Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        num_matchings = solve_rna_matching_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResult:\")\n",
        "        print(f\"Number of perfect matchings: {num_matchings:,}\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, num_matchings)\n",
        "        print(f\"\\nResult written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample data from the problem\n",
        "    sample_header = \"Rosalind_23\"\n",
        "    sample_rna = \"AGCUAGUCAU\"\n",
        "\n",
        "    print(f\"Header: {sample_header}\")\n",
        "    print(f\"RNA Sequence: {sample_rna}\")\n",
        "    print(f\"Length: {len(sample_rna)} bp\")\n",
        "\n",
        "    # Get detailed analysis\n",
        "    calculator = RNAMatchingCalculator()\n",
        "    analysis = calculator.get_detailed_analysis(sample_rna)\n",
        "\n",
        "    print(f\"\\nDetailed Analysis:\")\n",
        "    print(f\"Nucleotide counts: {analysis['nucleotide_counts']}\")\n",
        "    print(f\"A-U pairs possible: {analysis['au_pairs_possible']}\")\n",
        "    print(f\"G-C pairs possible: {analysis['gc_pairs_possible']}\")\n",
        "    print(f\"Perfect matching possible: {analysis['is_perfect_matching_possible']}\")\n",
        "\n",
        "    if analysis['is_perfect_matching_possible']:\n",
        "        print(f\"\\nMatching calculations:\")\n",
        "        print(f\"A-U matchings: {analysis['au_factorial']}\")\n",
        "        print(f\"G-C matchings: {analysis['gc_factorial']}\")\n",
        "        print(f\"Total calculation: {analysis['calculation']}\")\n",
        "        print(f\"Final result: {analysis['total_perfect_matchings']:,}\")\n",
        "\n",
        "        # Expected output verification\n",
        "        expected = 12\n",
        "        print(f\"\\nVerification:\")\n",
        "        print(f\"Expected: {expected}\")\n",
        "        print(f\"Our result: {analysis['total_perfect_matchings']}\")\n",
        "        print(f\"Match: {'✓' if analysis['total_perfect_matchings'] == expected else '✗'}\")\n",
        "    else:\n",
        "        print(f\"Error: {analysis['error']}\")\n",
        "\n",
        "\n",
        "def demonstrate_mathematical_concept():\n",
        "    \"\"\"\n",
        "    Demonstrate the mathematical concept behind perfect matchings.\n",
        "    \"\"\"\n",
        "    print(\"=== Mathematical Concept Demonstration ===\")\n",
        "\n",
        "    print(\"Perfect matchings in RNA secondary structures:\")\n",
        "    print(\"- Each A must pair with exactly one U\")\n",
        "    print(\"- Each G must pair with exactly one C\")\n",
        "    print(\"- Total matchings = (# of A's)! × (# of G's)!\")\n",
        "    print()\n",
        "\n",
        "    # Example with simple cases\n",
        "    test_cases = [\n",
        "        (\"AU\", \"1 A, 1 U\"),\n",
        "        (\"AUAU\", \"2 A's, 2 U's\"),\n",
        "        (\"GC\", \"1 G, 1 C\"),\n",
        "        (\"GCGC\", \"2 G's, 2 C's\"),\n",
        "        (\"AUGC\", \"1 A, 1 U, 1 G, 1 C\"),\n",
        "        (\"AGCUAGUCAU\", \"Sample case\"),\n",
        "    ]\n",
        "\n",
        "    calculator = RNAMatchingCalculator()\n",
        "\n",
        "    for rna_seq, description in test_cases:\n",
        "        try:\n",
        "            analysis = calculator.get_detailed_analysis(rna_seq)\n",
        "            if analysis['is_perfect_matching_possible']:\n",
        "                print(f\"{description}: {rna_seq}\")\n",
        "                print(f\"  Counts: A={analysis['nucleotide_counts']['A']}, \"\n",
        "                      f\"U={analysis['nucleotide_counts']['U']}, \"\n",
        "                      f\"G={analysis['nucleotide_counts']['G']}, \"\n",
        "                      f\"C={analysis['nucleotide_counts']['C']}\")\n",
        "                print(f\"  Calculation: {analysis['calculation']}\")\n",
        "                print(f\"  Result: {analysis['total_perfect_matchings']}\")\n",
        "            else:\n",
        "                print(f\"{description}: {rna_seq} - No perfect matching possible\")\n",
        "            print()\n",
        "        except Exception as e:\n",
        "            print(f\"{description}: {rna_seq} - Error: {e}\")\n",
        "            print()\n",
        "\n",
        "\n",
        "def test_edge_cases():\n",
        "    \"\"\"\n",
        "    Test various edge cases and validation.\n",
        "    \"\"\"\n",
        "    print(\"=== Testing Edge Cases ===\")\n",
        "\n",
        "    calculator = RNAMatchingCalculator()\n",
        "\n",
        "    test_cases = [\n",
        "        (\"A\", \"Single nucleotide\"),\n",
        "        (\"AU\", \"Minimal valid case\"),\n",
        "        (\"AAUU\", \"Multiple A-U pairs\"),\n",
        "        (\"GGCC\", \"Multiple G-C pairs\"),\n",
        "        (\"AUGU\", \"Imbalanced A-U\"),\n",
        "        (\"AUGCT\", \"Contains invalid T\"),\n",
        "        (\"AUGCAUGCAUGC\", \"Larger valid case\"),\n",
        "    ]\n",
        "\n",
        "    for rna_seq, description in test_cases:\n",
        "        try:\n",
        "            result = calculator.calculate_perfect_matchings(rna_seq)\n",
        "            print(f\"{description}: {rna_seq} -> {result:,} matchings\")\n",
        "        except Exception as e:\n",
        "            print(f\"{description}: {rna_seq} -> Error: {e}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate mathematical concept\n",
        "    demonstrate_mathematical_concept()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test edge cases\n",
        "    test_edge_cases()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "1GSOUZuXafqO",
        "outputId": "ee4877f8-055b-4fdb-ac58-f40ec16c9beb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Header: Rosalind_23\n",
            "RNA Sequence: AGCUAGUCAU\n",
            "Length: 10 bp\n",
            "\n",
            "Detailed Analysis:\n",
            "Nucleotide counts: {'A': 3, 'U': 3, 'G': 2, 'C': 2}\n",
            "A-U pairs possible: 3\n",
            "G-C pairs possible: 2\n",
            "Perfect matching possible: True\n",
            "\n",
            "Matching calculations:\n",
            "A-U matchings: 3! = 6\n",
            "G-C matchings: 2! = 2\n",
            "Total calculation: 3! × 2! = 6 × 2 = 12\n",
            "Final result: 12\n",
            "\n",
            "Verification:\n",
            "Expected: 12\n",
            "Our result: 12\n",
            "Match: ✓\n",
            "\n",
            "============================================================\n",
            "=== Mathematical Concept Demonstration ===\n",
            "Perfect matchings in RNA secondary structures:\n",
            "- Each A must pair with exactly one U\n",
            "- Each G must pair with exactly one C\n",
            "- Total matchings = (# of A's)! × (# of G's)!\n",
            "\n",
            "1 A, 1 U: AU\n",
            "  Counts: A=1, U=1, G=0, C=0\n",
            "  Calculation: 1! × 0! = 1 × 1 = 1\n",
            "  Result: 1\n",
            "\n",
            "2 A's, 2 U's: AUAU\n",
            "  Counts: A=2, U=2, G=0, C=0\n",
            "  Calculation: 2! × 0! = 2 × 1 = 2\n",
            "  Result: 2\n",
            "\n",
            "1 G, 1 C: GC\n",
            "  Counts: A=0, U=0, G=1, C=1\n",
            "  Calculation: 0! × 1! = 1 × 1 = 1\n",
            "  Result: 1\n",
            "\n",
            "2 G's, 2 C's: GCGC\n",
            "  Counts: A=0, U=0, G=2, C=2\n",
            "  Calculation: 0! × 2! = 1 × 2 = 2\n",
            "  Result: 2\n",
            "\n",
            "1 A, 1 U, 1 G, 1 C: AUGC\n",
            "  Counts: A=1, U=1, G=1, C=1\n",
            "  Calculation: 1! × 1! = 1 × 1 = 1\n",
            "  Result: 1\n",
            "\n",
            "Sample case: AGCUAGUCAU\n",
            "  Counts: A=3, U=3, G=2, C=2\n",
            "  Calculation: 3! × 2! = 6 × 2 = 12\n",
            "  Result: 12\n",
            "\n",
            "\n",
            "============================================================\n",
            "=== Testing Edge Cases ===\n",
            "Single nucleotide: A -> Error: Perfect matching not possible. A:1, U:0, G:0, C:0. Need A=U and G=C.\n",
            "Minimal valid case: AU -> 1 matchings\n",
            "Multiple A-U pairs: AAUU -> 2 matchings\n",
            "Multiple G-C pairs: GGCC -> 2 matchings\n",
            "Imbalanced A-U: AUGU -> Error: Perfect matching not possible. A:1, U:2, G:1, C:0. Need A=U and G=C.\n",
            "Contains invalid T: AUGCT -> Error: RNA sequence contains invalid bases\n",
            "Larger valid case: AUGCAUGCAUGC -> 36 matchings\n",
            "\n",
            "============================================================\n",
            "Solving RNA Perfect Matching Problem...\n",
            "\n",
            "Result:\n",
            "Number of perfect matchings: 778,817,392,288,148,379,660,189,696,000,000\n",
            "\n",
            "Result written to: output_pmch.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Shortest Superstring Problem"
      ],
      "metadata": {
        "id": "xTtQBv0tS9lv"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Genome Assembly as Shortest Superstring Problem Solution\n",
        "\n",
        "This module reconstructs a chromosome from DNA reads by finding the shortest\n",
        "superstring that contains all reads as substrings. This simulates the process\n",
        "of genome assembly where DNA fragments are merged based on their overlaps.\n",
        "\n",
        "The problem uses a greedy approach to iteratively merge reads with maximal\n",
        "overlaps (> half read length) until a single superstring is obtained.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Tuple, Optional, Set\n",
        "import re\n",
        "\n",
        "\n",
        "class GenomeAssembler:\n",
        "    \"\"\"\n",
        "    A class to handle genome assembly using shortest superstring approach.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the genome assembler.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_dna_sequence(sequence: str) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that a sequence contains only valid DNA bases.\n",
        "\n",
        "        Args:\n",
        "            sequence (str): DNA sequence to validate\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid, False otherwise\n",
        "        \"\"\"\n",
        "        return bool(re.match(r'^[ATGC]*$', sequence.upper()))\n",
        "\n",
        "    def calculate_overlap(self, read_a: str, read_b: str, min_length: int) -> int:\n",
        "        \"\"\"\n",
        "        Calculate the length of longest suffix of read_a matching prefix of read_b.\n",
        "\n",
        "        Args:\n",
        "            read_a (str): First DNA read\n",
        "            read_b (str): Second DNA read\n",
        "            min_length (int): Minimum required overlap length\n",
        "\n",
        "        Returns:\n",
        "            int: Length of overlap, or 0 if no sufficient overlap exists\n",
        "        \"\"\"\n",
        "        if read_a == read_b:\n",
        "            return 0\n",
        "\n",
        "        start = 0\n",
        "        while True:\n",
        "            # Look for read_b's prefix in read_a\n",
        "            start = read_a.find(read_b[:min_length], start)\n",
        "            if start == -1:\n",
        "                return 0\n",
        "\n",
        "            # Check if read_b starts with suffix of read_a from 'start'\n",
        "            if read_b.startswith(read_a[start:]):\n",
        "                return len(read_a) - start\n",
        "\n",
        "            start += 1\n",
        "\n",
        "    def find_all_overlaps(self, reads: List[str], min_overlap: int) -> dict:\n",
        "        \"\"\"\n",
        "        Find all pairwise overlaps between reads that meet minimum length requirement.\n",
        "\n",
        "        Args:\n",
        "            reads (List[str]): List of DNA reads\n",
        "            min_overlap (int): Minimum overlap length required\n",
        "\n",
        "        Returns:\n",
        "            dict: Dictionary mapping (read_a, read_b) -> overlap_length\n",
        "        \"\"\"\n",
        "        overlaps = {}\n",
        "\n",
        "        for i, read_a in enumerate(reads):\n",
        "            for j, read_b in enumerate(reads):\n",
        "                if i != j:  # Don't compare read with itself\n",
        "                    overlap_len = self.calculate_overlap(read_a, read_b, min_overlap)\n",
        "                    if overlap_len > 0:\n",
        "                        overlaps[(read_a, read_b)] = overlap_len\n",
        "\n",
        "        return overlaps\n",
        "\n",
        "    def find_maximal_overlap(self, reads: List[str], min_overlap: int) -> Tuple[Optional[str], Optional[str], int]:\n",
        "        \"\"\"\n",
        "        Find the pair of reads with maximal overlap.\n",
        "\n",
        "        Args:\n",
        "            reads (List[str]): List of DNA reads\n",
        "            min_overlap (int): Minimum overlap length required\n",
        "\n",
        "        Returns:\n",
        "            Tuple[Optional[str], Optional[str], int]: (read_a, read_b, overlap_length)\n",
        "        \"\"\"\n",
        "        best_read_a, best_read_b = None, None\n",
        "        best_overlap = 0\n",
        "\n",
        "        for read_a in reads:\n",
        "            for read_b in reads:\n",
        "                if read_a != read_b:\n",
        "                    overlap_len = self.calculate_overlap(read_a, read_b, min_overlap)\n",
        "                    if overlap_len > best_overlap:\n",
        "                        best_read_a, best_read_b = read_a, read_b\n",
        "                        best_overlap = overlap_len\n",
        "\n",
        "        return best_read_a, best_read_b, best_overlap\n",
        "\n",
        "    def merge_reads(self, read_a: str, read_b: str, overlap_length: int) -> str:\n",
        "        \"\"\"\n",
        "        Merge two reads given their overlap length.\n",
        "\n",
        "        Args:\n",
        "            read_a (str): First read (prefix read)\n",
        "            read_b (str): Second read (suffix read)\n",
        "            overlap_length (int): Length of overlap between reads\n",
        "\n",
        "        Returns:\n",
        "            str: Merged read\n",
        "        \"\"\"\n",
        "        return read_a + read_b[overlap_length:]\n",
        "\n",
        "    def assemble_genome_greedy(self, reads: List[str]) -> str:\n",
        "        \"\"\"\n",
        "        Assemble genome using greedy shortest superstring approach.\n",
        "\n",
        "        Args:\n",
        "            reads (List[str]): List of DNA reads\n",
        "\n",
        "        Returns:\n",
        "            str: Assembled genome (shortest superstring)\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If reads are invalid or assembly fails\n",
        "        \"\"\"\n",
        "        if not reads:\n",
        "            raise ValueError(\"No reads provided\")\n",
        "\n",
        "        # Validate all reads\n",
        "        for i, read in enumerate(reads):\n",
        "            if not self.validate_dna_sequence(read):\n",
        "                raise ValueError(f\"Invalid DNA sequence in read {i+1}: {read}\")\n",
        "\n",
        "        # Calculate minimum overlap (more than half the read length)\n",
        "        if not reads:\n",
        "            return \"\"\n",
        "\n",
        "        avg_read_length = sum(len(read) for read in reads) // len(reads)\n",
        "        min_overlap = avg_read_length // 2 + 1\n",
        "\n",
        "        # Work with a copy of reads list\n",
        "        remaining_reads = reads.copy()\n",
        "\n",
        "        # Greedy assembly: repeatedly merge reads with maximal overlap\n",
        "        while len(remaining_reads) > 1:\n",
        "            read_a, read_b, overlap_len = self.find_maximal_overlap(remaining_reads, min_overlap)\n",
        "\n",
        "            if overlap_len == 0:\n",
        "                # No overlaps found - this shouldn't happen with valid input\n",
        "                # but we handle it by concatenating remaining reads\n",
        "                break\n",
        "\n",
        "            # Remove the two reads and add their merge\n",
        "            remaining_reads.remove(read_a)\n",
        "            remaining_reads.remove(read_b)\n",
        "            merged_read = self.merge_reads(read_a, read_b, overlap_len)\n",
        "            remaining_reads.append(merged_read)\n",
        "\n",
        "        # If we still have multiple reads, concatenate them\n",
        "        # (this shouldn't happen with well-formed input)\n",
        "        if len(remaining_reads) > 1:\n",
        "            return ''.join(remaining_reads)\n",
        "\n",
        "        return remaining_reads[0] if remaining_reads else \"\"\n",
        "\n",
        "    def get_assembly_analysis(self, reads: List[str]) -> dict:\n",
        "        \"\"\"\n",
        "        Get detailed analysis of the assembly process.\n",
        "\n",
        "        Args:\n",
        "            reads (List[str]): List of DNA reads\n",
        "\n",
        "        Returns:\n",
        "            dict: Analysis including statistics and intermediate steps\n",
        "        \"\"\"\n",
        "        if not reads:\n",
        "            return {}\n",
        "\n",
        "        # Basic statistics\n",
        "        read_lengths = [len(read) for read in reads]\n",
        "        avg_length = sum(read_lengths) // len(read_lengths)\n",
        "        min_overlap = avg_length // 2 + 1\n",
        "\n",
        "        # Find all overlaps\n",
        "        overlaps = self.find_all_overlaps(reads, min_overlap)\n",
        "\n",
        "        # Assemble genome\n",
        "        superstring = self.assemble_genome_greedy(reads)\n",
        "\n",
        "        analysis = {\n",
        "            'num_reads': len(reads),\n",
        "            'read_lengths': read_lengths,\n",
        "            'avg_read_length': avg_length,\n",
        "            'min_overlap_required': min_overlap,\n",
        "            'total_overlaps_found': len(overlaps),\n",
        "            'overlaps': overlaps,\n",
        "            'assembled_genome': superstring,\n",
        "            'genome_length': len(superstring),\n",
        "            'compression_ratio': sum(read_lengths) / len(superstring) if superstring else 0\n",
        "        }\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_fasta_file(file_path: str) -> List[Tuple[str, str]]:\n",
        "    \"\"\"\n",
        "    Parse FASTA file to extract sequences.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to FASTA file\n",
        "\n",
        "    Returns:\n",
        "        List[Tuple[str, str]]: List of (header, sequence) tuples\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "        return parse_fasta_string(content)\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def parse_fasta_string(fasta_content: str) -> List[Tuple[str, str]]:\n",
        "    \"\"\"\n",
        "    Parse FASTA format string to extract sequences.\n",
        "\n",
        "    Args:\n",
        "        fasta_content (str): FASTA format content\n",
        "\n",
        "    Returns:\n",
        "        List[Tuple[str, str]]: List of (header, sequence) tuples\n",
        "    \"\"\"\n",
        "    sequences = []\n",
        "    header = None\n",
        "    seq_lines = []\n",
        "\n",
        "    for line in fasta_content.strip().split('\\n'):\n",
        "        line = line.strip()\n",
        "        if line.startswith('>'):\n",
        "            # Save previous sequence if exists\n",
        "            if header is not None:\n",
        "                sequences.append((header, ''.join(seq_lines)))\n",
        "            # Start new sequence\n",
        "            header = line[1:]\n",
        "            seq_lines = []\n",
        "        else:\n",
        "            seq_lines.append(line)\n",
        "\n",
        "    # Add the last sequence\n",
        "    if header is not None:\n",
        "        sequences.append((header, ''.join(seq_lines)))\n",
        "\n",
        "    if not sequences:\n",
        "        raise ValueError(\"No valid FASTA sequences found\")\n",
        "\n",
        "    return sequences\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, superstring: str) -> None:\n",
        "    \"\"\"\n",
        "    Write assembled genome to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        superstring (str): Assembled genome sequence\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            file.write(superstring + '\\n')\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_genome_assembly_problem(input_file_path: str) -> str:\n",
        "    \"\"\"\n",
        "    Solve the Genome Assembly problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input FASTA file\n",
        "\n",
        "    Returns:\n",
        "        str: Assembled genome sequence\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse FASTA file\n",
        "        sequences = parse_fasta_file(input_file_path)\n",
        "\n",
        "        # Extract just the DNA sequences (ignore headers)\n",
        "        reads = [seq for _, seq in sequences]\n",
        "\n",
        "        if not reads:\n",
        "            raise ValueError(\"No DNA sequences found in input file\")\n",
        "\n",
        "        if len(reads) > 50:\n",
        "            raise ValueError(f\"Too many reads: {len(reads)} (maximum 50 allowed)\")\n",
        "\n",
        "        # Check length constraint\n",
        "        for i, read in enumerate(reads):\n",
        "            if len(read) > 1000:\n",
        "                raise ValueError(f\"Read {i+1} exceeds 1000bp limit: {len(read)}bp\")\n",
        "\n",
        "        # Initialize assembler\n",
        "        assembler = GenomeAssembler()\n",
        "\n",
        "        # Assemble genome\n",
        "        superstring = assembler.assemble_genome_greedy(reads)\n",
        "\n",
        "        return superstring\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Genome Assembly problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_long.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_long.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Genome Assembly Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        assembled_genome = solve_genome_assembly_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResults:\")\n",
        "        print(f\"Assembled genome length: {len(assembled_genome)} bp\")\n",
        "        print(f\"Assembled genome: {assembled_genome}\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, assembled_genome)\n",
        "        print(f\"\\nResult written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data from search results.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample FASTA from the problem and search results\n",
        "    sample_fasta = \"\"\">Rosalind_56\n",
        "ATTAGACCTG\n",
        ">Rosalind_57\n",
        "CCTGCCGGAA\n",
        ">Rosalind_58\n",
        "AGACCTGCCG\n",
        ">Rosalind_59\n",
        "GCCGGAATAC\"\"\"\n",
        "\n",
        "    # Parse sequences\n",
        "    sequences = parse_fasta_string(sample_fasta)\n",
        "    reads = [seq for _, seq in sequences]\n",
        "\n",
        "    print(f\"Input reads ({len(reads)} total):\")\n",
        "    for i, read in enumerate(reads, 1):\n",
        "        print(f\"  {i}. {read} ({len(read)} bp)\")\n",
        "\n",
        "    # Initialize assembler and get detailed analysis\n",
        "    assembler = GenomeAssembler()\n",
        "    analysis = assembler.get_assembly_analysis(reads)\n",
        "\n",
        "    print(f\"\\nAssembly Analysis:\")\n",
        "    print(f\"Average read length: {analysis['avg_read_length']} bp\")\n",
        "    print(f\"Minimum overlap required: {analysis['min_overlap_required']} bp\")\n",
        "    print(f\"Overlaps found: {analysis['total_overlaps_found']}\")\n",
        "\n",
        "    print(f\"\\nDetailed overlaps:\")\n",
        "    for (read_a, read_b), overlap_len in analysis['overlaps'].items():\n",
        "        print(f\"  {read_a} -> {read_b}: {overlap_len} bp\")\n",
        "\n",
        "    print(f\"\\nAssembled genome: {analysis['assembled_genome']}\")\n",
        "    print(f\"Genome length: {analysis['genome_length']} bp\")\n",
        "    print(f\"Compression ratio: {analysis['compression_ratio']:.2f}x\")\n",
        "\n",
        "    # Expected output verification\n",
        "    expected = \"ATTAGACCTGCCGGAATAC\"\n",
        "    print(f\"\\nVerification:\")\n",
        "    print(f\"Expected: {expected}\")\n",
        "    print(f\"Our result: {analysis['assembled_genome']}\")\n",
        "    print(f\"Match: {'✓' if analysis['assembled_genome'] == expected else '✗'}\")\n",
        "\n",
        "\n",
        "def demonstrate_assembly_steps():\n",
        "    \"\"\"\n",
        "    Show step-by-step assembly process.\n",
        "    \"\"\"\n",
        "    print(\"=== Step-by-Step Assembly Demonstration ===\")\n",
        "\n",
        "    reads = ['ATTAGACCTG', 'CCTGCCGGAA', 'AGACCTGCCG', 'GCCGGAATAC']\n",
        "    assembler = GenomeAssembler()\n",
        "\n",
        "    print(f\"Starting reads: {reads}\")\n",
        "\n",
        "    # Calculate minimum overlap\n",
        "    avg_length = sum(len(read) for read in reads) // len(reads)\n",
        "    min_overlap = avg_length // 2 + 1\n",
        "    print(f\"Minimum overlap required: {min_overlap} bp\")\n",
        "\n",
        "    remaining = reads.copy()\n",
        "    step = 1\n",
        "\n",
        "    while len(remaining) > 1:\n",
        "        print(f\"\\nStep {step}:\")\n",
        "        print(f\"Current reads: {remaining}\")\n",
        "\n",
        "        # Find best overlap\n",
        "        read_a, read_b, overlap_len = assembler.find_maximal_overlap(remaining, min_overlap)\n",
        "        print(f\"Best overlap: {read_a} -> {read_b} ({overlap_len} bp)\")\n",
        "\n",
        "        if overlap_len == 0:\n",
        "            print(\"No more overlaps found\")\n",
        "            break\n",
        "\n",
        "        # Merge reads\n",
        "        merged = assembler.merge_reads(read_a, read_b, overlap_len)\n",
        "        remaining.remove(read_a)\n",
        "        remaining.remove(read_b)\n",
        "        remaining.append(merged)\n",
        "\n",
        "        print(f\"Merged result: {merged}\")\n",
        "        step += 1\n",
        "\n",
        "    print(f\"\\nFinal assembled genome: {remaining[0] if remaining else 'None'}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate step-by-step assembly\n",
        "    demonstrate_assembly_steps()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test edge cases\n",
        "    print(\"=== Testing Edge Cases ===\")\n",
        "    assembler = GenomeAssembler()\n",
        "\n",
        "    # Test simple cases\n",
        "    test_cases = [\n",
        "        ([\"ATCG\", \"TCGA\", \"CGAT\"], \"Simple 3-read case\"),\n",
        "        ([\"AAAA\", \"AAAB\"], \"Low complexity case\"),\n",
        "        ([\"ABCD\"], \"Single read\"),\n",
        "    ]\n",
        "\n",
        "    for reads, description in test_cases:\n",
        "        try:\n",
        "            result = assembler.assemble_genome_greedy(reads)\n",
        "            print(f\"{description}: {reads} -> {result}\")\n",
        "        except Exception as e:\n",
        "            print(f\"{description}: Error - {e}\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "027zPVkITAiv",
        "outputId": "bcd7284d-d434-4f12-9bef-e98c02a50b89",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input reads (4 total):\n",
            "  1. ATTAGACCTG (10 bp)\n",
            "  2. CCTGCCGGAA (10 bp)\n",
            "  3. AGACCTGCCG (10 bp)\n",
            "  4. GCCGGAATAC (10 bp)\n",
            "\n",
            "Assembly Analysis:\n",
            "Average read length: 10 bp\n",
            "Minimum overlap required: 6 bp\n",
            "Overlaps found: 3\n",
            "\n",
            "Detailed overlaps:\n",
            "  ATTAGACCTG -> AGACCTGCCG: 7 bp\n",
            "  CCTGCCGGAA -> GCCGGAATAC: 7 bp\n",
            "  AGACCTGCCG -> CCTGCCGGAA: 7 bp\n",
            "\n",
            "Assembled genome: ATTAGACCTGCCGGAATAC\n",
            "Genome length: 19 bp\n",
            "Compression ratio: 2.11x\n",
            "\n",
            "Verification:\n",
            "Expected: ATTAGACCTGCCGGAATAC\n",
            "Our result: ATTAGACCTGCCGGAATAC\n",
            "Match: ✓\n",
            "\n",
            "============================================================\n",
            "=== Step-by-Step Assembly Demonstration ===\n",
            "Starting reads: ['ATTAGACCTG', 'CCTGCCGGAA', 'AGACCTGCCG', 'GCCGGAATAC']\n",
            "Minimum overlap required: 6 bp\n",
            "\n",
            "Step 1:\n",
            "Current reads: ['ATTAGACCTG', 'CCTGCCGGAA', 'AGACCTGCCG', 'GCCGGAATAC']\n",
            "Best overlap: ATTAGACCTG -> AGACCTGCCG (7 bp)\n",
            "Merged result: ATTAGACCTGCCG\n",
            "\n",
            "Step 2:\n",
            "Current reads: ['CCTGCCGGAA', 'GCCGGAATAC', 'ATTAGACCTGCCG']\n",
            "Best overlap: CCTGCCGGAA -> GCCGGAATAC (7 bp)\n",
            "Merged result: CCTGCCGGAATAC\n",
            "\n",
            "Step 3:\n",
            "Current reads: ['ATTAGACCTGCCG', 'CCTGCCGGAATAC']\n",
            "Best overlap: ATTAGACCTGCCG -> CCTGCCGGAATAC (7 bp)\n",
            "Merged result: ATTAGACCTGCCGGAATAC\n",
            "\n",
            "Final assembled genome: ATTAGACCTGCCGGAATAC\n",
            "\n",
            "============================================================\n",
            "=== Testing Edge Cases ===\n",
            "Simple 3-read case: ['ATCG', 'TCGA', 'CGAT'] -> ATCGAT\n",
            "Low complexity case: Error - Invalid DNA sequence in read 2: AAAB\n",
            "Single read: Error - Invalid DNA sequence in read 1: ABCD\n",
            "\n",
            "============================================================\n",
            "Solving Genome Assembly Problem...\n",
            "\n",
            "Results:\n",
            "Assembled genome length: 19046 bp\n",
            "Assembled genome: CCATAGGTAACTTCTGGGCTCAATCTGGTGAGTCGGACCTCCTACGACCTGTCGCCTGGATTGTAACACGCATGGGTGACCTAATTACTTAATAAGCAAGCACTCACGACGAGCTCATGGTATTGCGCGCCTCTAGCATAAATGCCTACAACTTCAGTGCGGGATTAGCGCGACTCCCCCCCAAATAGAACGATATACGGAGGAATCAGTGGACACCCTAAACCTACCAATTTAGCGTCTTGGTATGGTAAGCCCTCCACACTCCAGCGCCATGTCCTTTTATTAAGGCCCCTATGGCTGTGTCAGCCGCTCGCTACGCGGGCAACCCCAGAATGGCCCGATAACCGTGGACGCATAGCGGTAGGCACGGCTGTTCTCTGTTTGCGCCTTGGTAATCTTGGTAGTCTGCCGGATGGCCAACGGGGCCGTCTCTTGCCAGACGGCCTGACCCGTCCTCTCCCGTTCTTCGATACGTATCTAGTATAGACGTTGGTTTTTGAATATTGTTAACTAACGCCACGCGAGTGCTGTTATTATATGACCGACGGCTCTCATACGGTGGCGCCCCTGCCTGAATCACCTTCGAAAAGGACGGTCCCCCACGGGTCTCGGAAAGACCTACACGTTGCCCGTTACTCCGAATTATGGAGTCGAAGAGGCTAGGTCACAGAAATGAAATGCGAGGGCGACAAGCTCCTCCCTCCGCCGGGCACGTCCGGGTACGAAGCCCCTCAATCGCCTAGAAGGATTCAACGACGTAGACTTACAAAGAACAACAGGTTTTTGTACCGTTATTGGAATTAAACCCTCGCTCCTGAATGAGAGACGGCGCGATATCTGAACGGTAAGCATTTTATTACGCGTAGGTGAGATTGCCTCATTATGTGTCCTCAAAGAATCTCTAGTATCGCACTATACTGGATTAGTGCCTAATTTTTATCCCTGTCCGGTTTTTAAGACAGGAGTTGCGCGTGCCTGGGCCCTTTCAACCGGCGAACTAAAACAGCCTCCGGCATCTCCATCTTGCTTTCGCGATGTCAGCAGTTACCGGTCATTTCAACGCTGTGCGTGCTCACATCATGTTCGCTTGACGTCGCCTTTCCTGCATTCCATTCTCGGTTTGATACACTCATCAGGCCGATACCCGTATTGTCACGAAAAGTCCGCGCCAATTGAACCGAGACCATGGTTCAGACGCTAGTAAATGGATCGTCTATCATGCACGGCAGACGCTCGCTGCGAGCTCCGGGGGTGTTTCTTAACCCGCCCGTATAGAATGCCGCGGATGTGCGCCACTAACACTATAGATTATCCCAACGGGGACTTTGCGCAGTGCACTGTTTCTAGGACCTCACGCGTTAATTGCGTGGCTCCAAATGACGTCAATGGGTATCCTGCGGCCGGCAGTGGCGTGCCCTCGATCGCAGACCATCGACGTTCGATGGGCGAAACAGCGGCGAGCCATATTCCATAAGTAGGAACAAAACGACGTCGGGATTACCCACATCCTCGCTAAATGTGTAATCAATATGCAGACTGACGAACAACGGATAGAGATCAGTGCGTCAGTAACCCAGCACGGAGCCGTGGTTGCGGTCGAGACTACGACGACAGCTTCTCTATTTTTGTGATGACGGACGTATTTCAAGAGATATAGTTGGGGGCGCAAGCTTTATTAGTAGTTGTGCTTGGTAGGCTCAACGAGCATTGAATAAGCCCATAGTTATAGGTGGCTCGAGTTCCTGGTTAATTCTTAAAAGAACTCGACATAATTCTCAGCTCCCTACTAGCGAGAACACGGCCTACAACACAAGAGCGGGTCTGAGAAATGATAATGGGAACAATTGATGAAGGTTGGTACTAGGATGGACGTGCGTGAGCGTTAGAAGGTCTCCCCCTGGCCAAGAGCTTCTCTCTGCGATAGCTCCTCCAGCACATACCTATCTGTTCCATGAGTAATGATCCATGCAGCAGGGCAGATGCTCCGGAGCCCTTAGTCAATCTGGGCACTACTATCGAGACGAGACGGCCACATTTTGGGTCGTGTATCATTGGCACTTAAGGTATCGGAATGTGAGCGTCTTGGCACGTATGCCACTCACGGCCCGTTCTACTATTCTATTACTGATCTCATATGAATAGTGCAAAACACGTCCGCCCAGGCCGTTTGGTAGACCAGTAAAGTCAGGCAACGTTCACACCGTCTAAAACTCGCCGTCCTGTGCCCATACCGCAAGAAAGTCTCCAGGTTTCCGCGTACTCATGTCTGTTAAAGGGGCTATTAAACTCTGGAAGTTGGACTAAAAGTTAGTGTGAATACACTCTTAATTACTTCTTACGTTGCAGATGGCTGGAATGGAGAATAACCAAGGAGCATATTGCGAATGCACGTAACCTATTCATCGTAGGCTTCTGGCAAAGGTGAAGATGACATCCACTGCTGAGTTCGAAGCGTGACGATCTATTCCTGAGGAAGACCAGCGATGTACTACCCATCTACTGTCTATTCAGGATAGTCTTTGACACGATCACCCGCTATACTCACCTCACGTTCTGATAGAACGCGGTACGTTAGTGCATGTGGTGCGATATGTACCACGAGCAGCCAGAGGAGGGTCGGACGGTGTCGTGGACGCCTGTGATAGACAAAGGGATAGGCCTTGTTCACAACGTCGACAGCTTATACCCGCGAGGTGAACATGCTCGAATATGGAGGAAACGTATCAAACTGGGTGAGAAGTCGATGATCGTGTTAGTGACGGCGTGAACCGTGTGGCAGCATAGACCATGAGCGAAAGTGCTTCCCGAAAGAAGGGGTTTACACTCGTAAACTACTCTCGTCAAATCAAAGGACGGCATTCAACACCGCAAAGTAGGATACGAAATTTAGGCACGGGGGCCCACGTCCATGAGCCTTGGAACCCATGAGGCTGACGCGCTCCACATAGGATCACTATCCTATTTACCCACAAGAGTACTTGCTAGCCAGGCTAAACATGAGGACCATTCCTTGTGCGAGTACATTACACTCAACGATGAATGAATTGAGTTTGAAGACAATGCAATACATAGACCTACAATATGCGACTGAGCATTCGGGCCTCCAGGAATCCCGCTCAGTGTTGGATAACGAACCCCCCACCTAAAGATTTCGGAGGCTCATTCCATTATGGTCCCACTGTCAACAAAACTTCTAGTGCAATAGCGGACTGGAGGGGTAAAGAAGGTGAGCTGCCAAGAACTACCCTTTGCGCTACTTAGTTGTACAGTTAGCGGACGCACTGGTAATGTTACAGTGGTGATGATATTCGCACATTTTGGAAATACGAGCAGAAACGAACAGCGGCTAGGGCCTGCAGACAGATCTAAAAGAGAGTCAGTTATCGTAACTCAGGCGCGGACGATTATTTACTTCCGCGTGACGAGTCATCATGTATGAATTTGCCCCGCATTTAATTCCAGCGCATGACAGAACCCGCCTATGGAGGCCCCATCATTTCGCTTTACCCCACGGCGACTTAACGGAAAACCGTGGGAGCGAGATGGGTGCGGTTTCCGTACAGTGGTAGGAATAGACGATTCATCTCACCACCCGCGGTCCGCAGACCCCTTAGCGGACATAGGATCTACAAGAGGTCGGGGAGAAAATAGATGGATCCCGCCTCATTACTTTTGAAGCGTGCGGGGCTCGACGCGATCTGAGGCCAACAGTTGAGCGTGTGAAAGCTCCTCGGGGCCTCACCGGTTAAGAACCGAGTACTGCTTCTGGGTATCATCCCCTACCCAGTCGCCGAGACCTCGCCGGAAGGTAGGAGAGCGACCCTCGCATGGATTCTTTTCACACCTCCGACCCCAAAATGACGTATCGGACTTACGAGGCATGGACTCGTTCATACTTCAGTCAAGAATGGGGCACCAGCAAAGCTGAGCCGAGCTCGCCTCCACTGAAGAAAGTCAATCTTTCCTCGGCATAGGAGCTTATCACTTTTTCTACGGCCACATGGGCCCACAAACCCGTCAATCTGGTTGGCCAATTTAAGCACGAAATCAGCTGTTAAGTACCTGGAGTGCGCCCGCTCAGGAATGTCCCCTCGTACCGCCCGAATCGCACCGCCAGATCACCCTTCGGTCCCGATGGTGTTCCGCCGTCGCTTGTCGCTCTAGTCATGCTGTACACCATTTACCGGCTGAGAGCTCTAAGTGAAACACCTCTTTACTTCATCACGCCTAATAAGCATATACCCGTCCCATCATAACTCTCTGCGGGTCAAGATTTATAGACATTATCAGTAGATAGCGGCTTGTATCCCTGACGGCCTCACAGGACTACGAAGTCGCGCCTACGCCTTATCCAGCACAAATCGTATTGCTATTCTTGTTTACCACTACGTAAGCTTCTGCAGAGTGTGTGCATCACGCGGAGGTACCATTCAGGCGGAGTTGAGTCCCCTGTAGCAGCTCACTGTCCTATGAATGACGTTAGTCACACGATAGCTAGACGGTTTGTAATATTCACTTTGACTTTCAAGCTTCGATAAAAAAAGGGCCCATATATAACGTTCTCCTCCAAACGACTAGAAACAGGTGGAGCTCAGGACTGGTAGCGACGATGAAAAGGCCGAGTTCACAAGAGCGTCCGTGGCCGGAGGATATTCCTAAATCTAAGATATTTGTATCGAGCGTTCGTTCCTCTCCTCATAATACGAAGGCCCGATGTTCTCCAAGACGAGGAATCATACGGAAGTCTCCGTCCGTAGTAGGAGACCAGGCAAAACCGTAAGCCATGTTCCATGCCCGGACCGGTACTATATACAGGTGGAAGAAGTTCTTGTAGAGCGCCGCAAGGGACCACGCGGGCACCTTACTATCGCTCGGCAAAGGACCCGCTGAATGACTGTTGGCATTTTGCACGGGACCGCTCGACCCCCACTAATCGAAGTAGCGCTTCCATTTTGGACGAAAGCATTTGGAGAGCTATAGAACTGACATAGTTTCGCCGGCACCTAAAAAGAGAAATCCTGGGTAAAGGTAATCACAAGGAGCCTAATGTGCCTGATTTTGTGTGACAAGTAGGGGTTCTCCAAGCGCATGAACGAGCAACTCGTAGTGGCGTAGAGACTCATTAGGTGACGTCGTACTTTCAGCTCTGGAGACCCCTAAAGAGAACGAAGACACTACCCCCACTTCCTATACAGGGCCCTCACCCGGTTATAGACGAGTAGTGTCTAACTCTGGTCCAGGTTAGTGACGAACGAGGATTCTAAGGTAAGCCAGAAAATACTCATTGGCCGAATCGCTAGCAGCCTGTGGGACTTTATCTGTGCTGATTAAATTGAGCGGCGGGTGCTCATGTCTTCACGGCTCACGGTTCATCATTAGACATGTTTAGAATACAATAAGCCAGTCGCTTTTGGCAGTCTCTTATCTGGAGCATTGCGGCCAACGCGGCACACTCTTGCTTGTGTTTTAGTGGGAGGTTTGAGAGGGGTGAGAAAGGGCATGTGCCCAGCTCTAAGAAGCCCTTCCTCGACAGCTAACGAGCGGAAGCCGGTGGATGGTATATGCTATGGTTCCGGCAAAGTAAGCAAGGGCGTCCAGCGACAATGGACGGAAAGTCCTTTATCATAGATCTAAACTAGTGTATCCTCAAGCGAAGTTACGGTTGATCATTACGTCGTTTATGATGCCCGTGCTTTCGCCTCTACCTGCCGTCTTATCCGTGCTGCGCTCCTATTCCACGAACGTAGAAGGCAACTCGCAAGCCGATGTACACCCAGATAATATATGCGTCTTAGGTGATGATTCAGCGTCAGTGAGAGCGATAGTGGTAGTGAGCTCGACGACGTTACCTTTTTCTTCTACTACCATAATACAATATCACGAATGGGGTGTCCACTCTTCAAGCGGGTTTCAGACGCCGACGTTTCCACATGCATGTGTGCCTACGAAAGATCAAGCATCGCATTAACCGAATCTACGGCCAGCGTATTCACCAGGTGTCTAAACCGAGGCCAACAGTGATTAGGTTGTTTGTCACCGAAGTCAGTCCCTGCGAGAAAAGACTCGTTGAGCAGAAGCCAATAATCAGGAAGATCGCCACCTAGGCAACGAACTCAGGGTCAATCTTGACATTGTCCGACACGTGACGAAGAGTCGCCGGGGAGGCTCCTCTGTCATTATAGTACTTCTTTTTAGTCTGGATTTACTATATTAGAACGCGAGATCCATACTTGTGAGCCTCCTTCCTGTCGAGATACCCTTATGGGCCTTTCCGAAAGTTCCGTGCGACGGAGAGGTAGTATCACATCGGAAGCGCTTAGGTATTTTCTGTAGCTGATACTGAGCCCTGCTTGTCGACGTCGCATTCTTCTGAGCCGTGTGGCGGTTTCACATCGCGACTGCTAGGGCGCGTTTAATTGATTGCCATGTAGTGTCCCATATTAGTGATCGCTCCGGCACATTACAGATATCCTAATTCAGGTCAACTGGCCGGGTCGTTAAACCTTCACCAAGTGCTATTTATTACATGAGGTAATTGTAACGAGCAACGTGCGTCTTTTAATACAGGTGTCGATTCAGCTACCACATCGCCCTGAAGGCCCGAGGAAAGGCCCCGCGGAACGCCACGCCTCACTGTTTATATAGCCAGCGCCAGTCGGGTTTGGTGATAGGACAAAGTCGCAGTAGCACATCTACACTGGAAAAGAGGTCGTGAATCATACTGCCGCAACATCCTATAGTGCGCAAACCTTCAATGCCCAAGTCGATGTACGAGTTCCCATTCTGACAGATGTATCCGTCAACGCGGAGATATGTGCTTAGAAATGACATTTCCACTCTTCTCGTAAGGGTATTGTTTAGTGAGGGATTATGTGAGTGGACCTCAGGAAAGCCGAAGCACTCTGCGCATGTCATATGATTCACAAATACTATTGGGCGGCACAAGCGCGCGAATCCGAGTTTAGATTTCGATTTCTATCCACCGAAATATTTCCAACAGAGTAAGCTGGAGATACCCTGCGTTAGAGCCAGAGGTCAGGGCGCCAACTTTTAATGCGGGGGGAGGCTCACAGTGTCCTCGTACCATAGAGTGATGGTTGTATCTTACCCCGGCCAAATGTGTGCGTATCCGGTATCCGTCCGAGGCCCCCAGCGTGATGCACGTTGCTCCCTATCACGGTTTTTCTGTACGCCAATCCAGGGTATGGTTGGTGGTACCCCTGTCTGCAAACGTAGGAATGAGAGAATACACGGCGTCTGCTCCGGCTACTATGGCACATTCAGCAGTATCATTCGTCTCTCCACCCATTCTCTTCCATGAGCTGCATAACACGAAGGTGGTCATAAGGGTTGTATAGCTTTTTCACGGGTGCTCTGCCCCTGTCATACGTCCAAGCGGGTACCAAACTGTGCCGCGGTCACCGCACTGGTTTACAATATGTAGTTTGCACACTATCGTAAATTGGCGGCGACGCATCCAGTGATTACGGCCGTCCCCCCTTCTTAGATCTCTGGGAACACATCGCTATGGTGTCGCTTATGCTGGCCGTCTATAGGAAGTTTGGAGCTTTGTCGGCGATGGCTTGCAAAGACGAGGGAAATCAAATTGCTCTCGGACCCTTCCGTCTGGATCACTGATTGCCTAGCTCTGGCCCTTAGGTCCGTAGTTTGGTTACGGATAGTTGCGTTTTATCATTCTTCTACCATCGGTAAAGTGTAGATGGACCCGGTGTGGATCAGTTTGCGTCGCTGCTGGATAGCATGCGAATCAGCACTGGGGGATGGTGTGCCATCGATGTGGTGGGGGTGCGGACAGCGGCTTGTTCTCCTTCCATAACCCGCATGTTCCGCCCGTATTCGATGTCTTACTCGTAGAACATAACCCTGCTTATAAAGGCGCGCGATCGAATCATGATGGCCAGTAGACTCCCTGCAAAGTTCAGTTTATCCCATAGATTCGCAATAATCCCTTCAAATTAGATCAGCCTGCGGTACCGGGCTTGCTACCACCACACGTACGATTATACCGTCCATGGCGAGTGACTGGCGGTTATCTGTAACCAGCGAGGCTTCCGGATTGCTTCTAACAGTTATTTGTCTCGAGTGTTCGATCGATACCAAATGAACACTTCGGCAGAACATCATAATCGTTTGCAGATTTTAAACACGTACCATCCCCTAACGAGTACTATACGGTGTGATTGGGAGCTGCAGAGCAGGCAGGTCTGGTGTTAGAATGCACAATCCGCAAAACGAATTTATACATGAAACCATGTCCGGTGGAAGTTCTGGTGTCGTAGGGTTTTTGAATACCCGCGGAGGGGGTAGGTGAAACTACCCCCTTCTGAAGACCGTGGCCTCGGAGTACCTACCTTTAAAACGATTTGCGTTTGGTGAGAACCTTGTATTGCATCTCGAGTTGATTGAAGTCTTCCCAACCGAGTGACAGATGACCGGTTCGGTAATAGCACGGTACATAACCTATGGCCTCCGAGATATGAGGCATGCCTTTGATTGCTTAGCAGTTCCATTCTCAGACCCAGGTTCGAGGCAACTATAAGTGTACTGATGCGCCGTACCCGAGCTCCAGACTGTCGTTACGTCGTCGACCCCAGAAGGAGCGCAATGGCACTCCCATGAATTAAACGAAGTTCACGCCCAGGGGACCTTGGCGGAGTCCACAAAAATGCAGGTGAAACTCGAGCGTTACACGTCCCGGTACCTAATTAAACGTGAGGTTATACCGTGAAACCCCGCCGCCGCGGCAGGTCACCTGTATCCACACTAAGCACCCATTCGTGCGAGACCAAATTCACCAACTCGATAATCATTAACAGCTCACTATTACCGTGCAGGAGCCATGAGGGCAGTACTGAAGTAATTAATGTGGCGAGGTAATCCCCTTCCATCCTGGTCAAGACAAGATGAGTTTTCTTGAGAAGGGGGTCTCGACTAGAGCCTTTCGTGAGCACCACAGGATTGCTTCTCAATGAACATCCAATTGATCTAGGTATGGATCACGCTTCTGGTTCTCCGTCTTGAGAAGGAAAGGATTAGCGAAGTGGGGTGGGCCGTTAGTTTTGCGGTGTACACCCGCTGCGATTTTTACTTGATAGCATAAATGCTACCCAGCCGCGGACATGTGCAAGGCCCTATCATTTCCTCGCGAATATAATGGATATGTGATAGAAGAGCCTAGCCTTACCGTTATACCAAGCCTGCCCGAAGTTCCCTGATGTTCTTTCGGCATTTCGAGATAGGTAAATGGGATGAGCGTATTGCTTCATGATGGGCTTGAGGGCATTCGGGTCTATCGCCAAGGACGTCGACTACGCAGGTTGTCTATACGCGGGAGCACCATCACAAGACGTGAGCTTGAATGAGTAACTCAAAAAATATCTCACCCTTACTCTGCCACAAGATTGATAGGTTTGAAGCGAACCCAAATAGCTGGCGGGTTATTCTGCTGGGCGCCGACGATTGCCATAGAGTCAGCTCGCACAGACGTGTAGTGCCTATAGTTTAGAGGATGATCCCCTACCGATATACAGTGGGGTGTCATAAACCAACAACCGAACGAGCTATGTTTCCAGCGTCAGCAAGGCGTCGGGTTGCAGCTTCTGACGGAGAAACGTTTGTATGTCGCCTATTGCTTGGGGATCGCCCAAGTGAGGGTTTAGCATATGGTACTGTGTCGAGGTCCATTAGATCAGGACTGTCGTGTCGCTCTACGCGCTTATACATGGCCGGACAGCTGTCGCCCTCTAGGATTCCTCCAAACCTCACAACAACTTGCGACCACAGGATGTTTAGTTCCTATGGCAGTCCTCCTTGAGAACTGCGGTATAGTTGTTACAGGGGGCTGCATGTTGGGCAAACACAGTCCTAGAAGACTCAGCGGCTAGAAAGTGGTGGCATTTAATCGTAACGTATCGGGGGTGTCACTAGACAATTTCCCTACGGTCGGCGGTCTGCGACTTCACTACTTGCCGAAGGGTAGCGATCCCTGCGTAACTTATAGCTTACGTCAAGTAGCTTCTACGCTCAATTCACGGGTGTCTACCCCTTAATTTCCAACCGCTGAGCCAGTTTGTTGACGGCGGCTTCCATGGCCCACGGGTCATCAAGATGAACTCCGCCCTTGTGTCGTCATGGGTCAGGCTGGTAGTACGACAGCACTATAGCTTGTGCATGCGAAAGCAATGCTTCAAGCACGAATTACGACCGGTTCAGTTATAGGATGCTTCCCGTCCCATTTTCACATATAGGTTTGAGGATCTCCCGGTCATCGAACCTAGGCTTGTGAGATAGAGAATTTGGCAGCATCAGGCTCCACAGGACCTCGGTTAATCGTAAGGAAGCCGAGGTCGCATACCATAAAAGCGCGGCAATTTCGAACAGCCCTTCTATAGTCTACGGAACGGACCTACCTATTCGTAACTATATCGGTGGGCCCCGTGTACCGATCCGTCCAGAGCTGATGGCCAGGGGAGAAGTCCCTTATCGTGTACACCACGAAATTGCAGAGCCTCGAACATTGCACGTCTACTGTTCAATTGTGAATAGACACGCGTAACTACCCAAGGGCACTATTGTTTGGTGAAGAGCCATATAGGAACAACTGTTCCTTGTTGTGCGTTCGCGTTGATACAACCCTCGGGAGCTCTTTGACCGGCGTGACCCATCCCAAGTGAGACTCCCTCCGCGCCATTTAGGTCCGTGCCTTACCATCTGCGCGCGATGCTTTTGCAAGTGCGTGTGTCGAGGCCACGTGTACTAGCACAACCTGTCTCCGAGGTCCAGGGGAGCGATTAGTGGTGCGAGTTGAACCGTTGCACTCGACACGTAATTAGCAAGATGCCTGTAAGATGTAAAAACCCCCAGGTAGGCCCGATTTGGGGTTATGGTCCAGCAATACTTTGCCCTCCGCATTGAACATGAGGAGTTAAGCGTCCGTGGCGACTTATGGCTGTCCTAAGCACCTGCAGTCGAGCGACCCCGCTTGACCTGAGTCCGTTGGATTGTAATTCATTGTGGCTAATCGTCGTGTAGGCTGATGGATGGATTCTTTCGTCGCTCACCCCGCTTGCCATTGCGAACACTGGTTTCCCCCAGGTATAAGGGGGAATACAGAGAGAACAATTTCCCTCTTACAGGTGTATAGATCAGTGGTACCCACTCGCCCACCGGGCTGCTTAGGAGAGGGAGTCTCCTATCAATACTTAACGAAGCGCTTGGGGTATATGTAGTTCTCCGGCCTTAGGCACACAGTTACAGTATCGAGCAGAGTAACGAATCGTAGTGCCCGCATCATTTGCGTTGTAAAAGTTCGCCAGCACTGGCAAGTACCTTCATTTTCGCAGTCTCCGACTCGGGAAAATCCAAAGTCCTTTGCAGATTATATCACTCCCTAACATCTCCTCCGGAGTTGTCCTGAGGTCAGGCGTCGGTTCACAGTCAACGTTGGAAATAAAAAGGCGGCCGACCGAGTTGTACATCTAAAGTCGTTAGTGACCCGAAACGTGAAACCAGAGCATTAAACTCTTCACGGGCGTAACTCGTTAATGGTGCTCTTACGGCCGAGATGACAACGGTCAGATGGTGACGAAAAAGACGCCGGATCACGGATAACGACGAGAACTTGTGAGCGAGTTAAAAGGCGCGTAAAACGGCTTTATCCTCGCTGGTGACCGAGGTATCTGACTGTTAACGTCTAACATTATGTGCAGCCTCACACAAATGTCCTCTCCACCGAATTTCAGGGTGAAAACTCTCGCTAGGTGAACCCGAGATCCCGCGGTCTCGACGTGTGTGGATGGATGTGGGTAAGAGCATATTATATCCATCCTGCCGGTTTCAGACGCAAGGCCCTTTGGGGCTAAGAAGACTCCTTCGTATATTACGGCGTGGGCGGTCCCTCACGGGTTTCGTGGGAGATAGGGTCATCCCGGCAGAAGAAGGGGTATGGTAGACCGGATTTCTAACTAGGTGAGTTAGGGATGCGGGTAGACGAGGTTAACCCCATGTCATCGCGAGATATATCGTCTAGCGCTGTAGACGGCTACACCTATAAGGTCTCATACCGACCAGGTGGCGCAACGCGTCACAAACCTATTTTTCATCCCATTGTCGTGCTAGCTGTGGTATCTACACTCCGTAGAGCCTAGTTTCGGCTATTACTAGGGTGCGCTAATATGAGTGAGTCTGGCGCACACTTAACCACAAATACGCTGTGACGGAGCACTTCGCCAGGATAGGACTCCCAGGGTTTAGTGCCACAACATCATCACGCAGGCATACTTTCGACAGTGTGTCTCATTGGAGATAAAGTATCTAGGGAGTTTAAACTACCTGACAGCTATCACAATTCTAGTCTCATATGGTAGTTCTCTTCACAACCGGACATGACCGCGAATGGAGGCGGCTCTGTGTCTCAACAGTACCAGCGTCCGATTGCGAAGAGGGTTGTTATCGAGCTGAGCAGAGCATTCCAATCCTTACTCTTATCAGAGCTGACACGGATATGCAGTGCCTGTGGTGTCTTCAATAGGTCCCAAGCTGTCGTAAAGACTGAGCGTAAATCACCTAACTACGAAGGTCCAGGGGAAGACCTTAAAACTTCCAGCCGTCAGATGATACTCCCCTGCCTCATTGGGCTCAGTACTCACAACGCCGGACCTTGGGCGCGCCCTATTTCACCCAACGCTTTTAAGCATACATCTGTACATGAAGGCCCCTAGGCTCAACTAGGCTGCCCGCCTCCTAAATCTAGGCGCCGCGTGATGCCGTGAGCTCAAGCCTATTTCCCTTAAGCGCGCCCAGGTTCAGTGTGTGTTTTTTTGCGTGCATCAATGTAAGGTGTCGTATAGAGTACATCAATAAGGGTGGTCAACGATAATTCACAAACAAGTGAACCAAAACGTCCCCGATGACTCGATAGACTACCAATCGGAGTCATTACCTAAGTTGGTCGTTTTCGGATGCATAAGAAGAGAGTGTTCGGATATTCTTGGATTCTGCCACGAGCTCTCAAAATACAACTCTGATCTGTGGTTATCACTTTGACTCCCGCCCGCATTTGCAATCGACAAGTGCGCAATGTATACCACCCAAAGGACTTTCCCGACGTAAGAGACGCATGCACCTTAATGACCCAACCCGGGTTGATAGTGCAAGCTATCAGGGAACATTCTGAGAACATAGGGTGATCGGTTAAGCCTATATGGCGCCCTAGCATATGGGTAATTTAGGCGTCCGTTCCATATTTTATTAGTGTTTAAGAAATTGACAACAAACGATTAGATATCCAAGAGCCCAGTTTATATAGAATGCCGATGTTCAGGGCTGAACGTTCACCTACATATTTTGAGAAGCCCGTGTCAAGTATGCTACACCTTATCTGGGGGATCTGTAATTTGGAAACTACCAAACGAGTCGTACCATCCACGATAAGGGTCTAAGGAGCTAGTCCATTAGGGCTTTACGCTCATAGCGGGCAGTTCGATTCTTTCCAGCTTTGCCCCTACTATATACCACGGACATGGATTTCCGTTACGAGAGCAGTCGGGATTGCGAAGATTGCCATCATACAAGTCCCTGTACATTAATGCGAGCTCAGTGCATCCACATGACTCTCTCACCAACGTGTCAAAATGGATGTCGTCATTCGAGCGGGGTACACGCTTAATTTGGCATCAGGTGCTCCACCGGAGACGTCCAGGGTTAAGGAATCCCATTGGCACCCGAGGTGATTTGTAGATCTCCCTATGGGGAGTGCATGAGGTCGGTGACTGTCTTATGGAGTGTATAGCTGGGTCGGGCTTACGTGAATGAAACTCTCGCTGCAACACCTCGAAAACGTTCGGAGTCAACCGCCTCCCTAAAGACCTCGCTGACGCGCTAACGGGAAATAGGTACCGCAGGCATTCTTAGCTTAGACGTGCTCAACTGCGGGCGGGGCCATTGAAGGGGCAGCTGGACATCTATGCTAATAAAGCCAAAATAGGCTATCCACACGTCACTGATGGGTGGGGTCCATCGTAGATATTCACGTCCAATCGCAGTCACCTGTAACTGTTTCCGCCTGAACAGACCGAGTAATAGTGCAATATCCCTCAAGTATAGTGGCTTAGACTGGCGAGGTGAAACAATTTCAGATTCCCATCTAGATGTGGTGCCGGTTGGCGAGGAACGGGGGATTGCGATGTCCCTACCTCTCCACGAGTATGAAGTACTAAGGCGATACAAAAATTCGCTCTGATGAGTATCGGCGCCACAACAATCTCAAGGAGCTTAGCCGCGCCAACAACCGTGACACAACTTTCATTGGGTTTGAATGAAACACAAATCGTTTTATTGGTGCTAGGCGGCATTCCGCCATTCTGGTGTTCAGTCTCATGCACGGTTTTATTTGCTTTTTCCTAGTAGTAGAATATAGCTCACAACCCTAATAGGTCCTAATTATTGATTCATGTACGTAGGAGATTCTTATAGGCTGGCGACCAATTGCCTGCATAGGTGCCGTAGGGAGACTTAAACTAGGATCGAATGAGCCAGCCGTACACACTACACTTGCTTTGTGATGGGCCTACAGCGGCGTTCGCCATCCTTTTGATCGGTATAGCTTCCTTACTTCATGTATGCCGATGGCGCCATTTCGAGGGGCTAAGCGGACGCCATTGATAAGGGAAGGATCGATGTCTCTTGACTCGGTCCGGAGAGCTCCGTGTTGCCCCCCGGCAGCCTAAGTCATCTTGGGTCCACATAACATCAGTCCACGCTAATCAAACCGTATGATATAACACATATGTCCGCGTCCCATTCATCGTTATCGTGGGAACGTGCACTGGAGAACACACTACGACGATTAATTTATATATATAGTGAGATCCTGCGCACCCGGACCAGCCTCATTTTCACTCTAGTGCTATCGCCGAGAAGCACCGAGGCGAAAAGCGGGATTAGCAATGGCGCCTGTCAATTGAAGTGTTTGAAGGTGCAGCGTAACATCGGCGATCGAAACCCAGCATCCCTCATAACTTATGTAACCGACCTTACTTCTTTCGGTCCTCGACGCTCTAAAAGAATGTCGCTGCCTCGGGGTCGCTCTTAGTCACGGACTATGTGCCGACGTCGCCGTTTTATTCTCAATTTCTGTGATAGGCGCGCAACGGACATACGGTGACAAAGGAGCACCAATCGATGGGGATTGTGACGCGTAACTCCTAAACCATGGCGGTCGATAGGGCGTATAATCAATATCTGGACCAGGGATGACGTCCGGTCCTGTGCCGATAACTCATCATCCCATAAAGCCACCGAGGTGCGTAAAACTCCTGCCACTGTATCTGAACGCGAATACACCGACCGAAACAGTTTTCAGCTAAGCGAACACTAGGTAAGACCCAACATCGCTAAATACTACTCTACCATGAGATAGGTACAGGCATCGTAGCCGTTTTTGACTGTGAGTATAGAGCGTCAGACGACCTGAAGTAGGTTGCTCCAGGGGCTCCCCTTGCGCTATGATTTTGGGTCGCTACCTGTAGGTTAACCTAGATGCGTCGGCCGAAGCTTCTGCTTCGTAACCCGGCAGCCCCCTGGAGTTTGCACAAGAAATCGCAAATGCTTGATATTGCATTGTAGGGCCTTATCGGGTCATCGGTATGGCAATGTAGTGGGCATTGAGGGAGTCTTTCGCCTGGGCACATGACAAGGCGCATTGGCAGGAGCTATCTCGCAAACATCTAACCACCCACGATAGCTGACCGCGGCCCTCAAAGAAGGTAGGGCCCGGTGCGATAAGTATTTCAGGGACCTTTCCACCATTTGGGTTTAGCGGTTTTGCTGGCGGGTGCTATAGATGATCTTCATGATTCGCAAATAATTTAATAGGTCTTGGGCGGCATCTTTTTCACCTTGTACGGGCACGCGGGGCCCTGGGTCCAGGTAACCGGTCCTCTACTTAGCGTCAGATAAGCAGGTCACGTATCGCCTGCAATAATATGGAATGAAACGCGCACCTGGGTGCACTCGCTAATAACGCGCCTTTGAAATCCTATCGCCTTCATTATTTCGCCGTGAGATGCTTCTATATATTCACTTATGTCCCACCTATAGTATCCGCATGGGAACAGTTGTGATCTGATAGTTGAGAATGGAGTTAGAGCATATACGTGTCCACCAGAAAAGCGAGACTCCTTAATAGTAGTTACAGACCCATCGAAATGAAAAATCTATGTTACTGTGCGTCTACTAAGTATCCTCCGTTTGTTCGCTAAAAACATGTTTAAATGTCTATGTGGGTTAAATGCCCCAACCATTATCACACGTTACCGCTTAGGCCCAGCATTATCCAAGGCGAATAACGCGAGCCATCCTAACTCGGACTGCATCGTGAGTCATGACCACGGGCTGCACCAGGAGGCCGACGGAGCCATCTCATACACTGGATGACCATATCAGTAGAGGTGATTAGGAGCGGAACTGCCAACGTCGGCTAGGGATCTGAGCTTACACACTATGCCATGGGAGGCAACTACCGGCAACCGACAAGCACGCCCTTACAGTCACGGAATATGCGTCCCTGTCTATTCACTGAGCCCAAAATTTAGTGGCCCATACGACACCTAGCCAGATAAATCTTTTGTTAACTACGTATATAGGCCTACATGAACCCAGTGGATTGGATAGTCCAGCAGTTCTTACAACTACCTTACTTTATGCCATTCAATCGACGTGAGTATGCTTAATACAAGGGTATTTACCGGCGAAGAAATGTAAGTGTGAGACCCTTCCCTGGCGAATCACTGCCTGGCACATAGGGGCAACGTTTCTACGGAGTCGAAGGAGTACAATCAACTGATGCGAGCTCCGAGGCACGATCTGGCTTCAGCACCCAGTGGCTCTGTTTGGCAGGAAGGTTCGAACGAAGAAGAATAGGCTCATCTTCTTAGTTGCTAAGGTCCCGCTTATTATAACAAGCCATGTGATTTCTTGGCAATCCACCTAGTGGACAATTTTGCGAGGGGCTCGATGAAAGCTTGGTATTCACAAGGGGTAGCAGTGGGTGCCTTCGTCCACCAGGCACGCGCCCCTTCGCTGAATATGGCCAATGGGTAGTACTTAGCGACAAATGATTTTGCCGTCCACGTGTGCCTGCCTTCAGCTCGTTGCGGGTACAGAGGCGGAGCTAAATAGAAGCATTTGAGGGAAGGTACCCGCATACACACGGAATCCATTGTCAGGAGGAGCGACCTCAGGGATCTATCTGTGTAGACCCGGCCGGACGGGGAATACTTCACAGTACCGAGTAGTACATACCGGTATGTCGGTAATGTCTCAGCAGGCCTACGACACCACCTAAGTGACCCTTTTCCTGTACAATGTTGTGCGTCCGCTCATTCGCGTCGCTAACAAGGTGATGTATGGTGAAAGCCTCCTGCCACGGTTCACTAAGTGTATCATAGTGTCAAGCCCATGCTGATGAGGGACTTGATCTTGTTCTTCGGGTATACATGTTCTTCTGGTTAGCCTTTGCGTAAGCCAGACCGATCTGCAACCCATTATTTTGCGCTCAGCCCGGTCCGGAGCTTCAACCTTCCCATATATACCTCCACAACTTTAACCCACGCCAGTCGAAGGCCTGACATTAGGTTGTCACCGGAAATAGACGTCACCCATATATTTTATGCTCAATGGAGCATTAAGTTACAGATAGAGCTTTTTGGGGCTCCATCGTCTTTAACAACAAAATCGTTCCGGATTTTAGGATCAGGCCAACGGTGATATACGCATGTCCATAGTCGTCTTTCGCTACACTCAGTCCTCAAACATTAGACATCCGGACATATGGCTATGGTGAACTTCAAGCCCCTGCACCTCCCCCACTCAAATTGACCGGACCCGTTAGCAATGATTAACAGCTTTAGGAGATGTACAGCGGCAAAGGTCAACTAAAAAACAAATATGGTGGTTTCAGTCGAAGCGGATGAGCCGTTGGAGCGCACTAATAACTGTCTCTATCCTTAAATAAAATTAGAAATGGACCGGGCCGATTACCAACCACCCGGTCCGGTGTTACAATGCAGACGCCTTCACTCTGTATCCTGAGGCCCTTATATAAGGTCCGCTGCGATCCTAGGGTCGCACTTCTCAATAATGCTCTCAGCGATTGAACAAGAGGGCGGCACTAAACTCAATGCACCCAACGATTCTGCGAACGATAACTAAGAACCCTTTGGCAAGATGGGTGACTTCTGCAAATAAGATCGGAAACTTAAGAGGGAGAATCCCCTATGATCCAAAGCCCAATACCGATTACGGTTATGCGAACAGTCCTCCCTGTGTTCCCAGAAAACGATTATTGAAAGTTAGTAACATCATCAACGGCGTTAACAATAGAATTGTGATCGTCGTAGTCGTCCCACTCTGTACTTATTTAATCGGACGGGTGTAAACGACAGTTACGAGATCGATGCCTGCCGGCGTGTTAACCTTATAGTAGTCTGTAACATCCTCAGAGACCTTAATATACCCAGAGCGATGACAGAACGGGAGCGCGATATATTAGTGCGCTTAAATTGAATCTTAAGAGGTTTTCTGCATGGATAGGATTCCAGGTCAAAACGGCATGTATAATAGGACGGTAAGTGGCGGATTGGACTTACGCTCTTTACTACTGAGCTAATTGCCACTAGTGCCAGTATAGCCGGGAACACACGAATGAAAGCCCCTTTGTCATGATTAGCTTAATTTCTTGGTCATCAAAGTAGCTTTGAAAGGCGGTCCCGGACTAGGTGTATTTTTCCACGCCCGGTTTCTTGTCCAGTTCCTGACTAGCCCCATCTGTATTGAGGATGCCATGCGCGCAGGCAGCCGGGTCTCTGATCTCGACGATAAGAGCTATTCTCTCCAGTCGCGGACTAGTCTATGCTCGCAGCTGGTATTTTAAAGTCATACCTCTCCCCGGTGGGAGGTATAGCTATTTGTGGGTGTTCAGGGCTTGCTTACCAACAGTTTAGCGTGCGAGTCCTATTTGCGCCTCA\n",
            "\n",
            "Result written to: output_long.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Longest Increasing Subsequence"
      ],
      "metadata": {
        "id": "i_kXMoNolUB5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Longest Increasing Subsequence Problem Solution\n",
        "\n",
        "This module finds the longest increasing subsequence (LIS) and longest decreasing\n",
        "subsequence (LDS) in a permutation. This is useful for comparing gene orders\n",
        "between chromosomes from different species to measure evolutionary similarity.\n",
        "\n",
        "The problem uses dynamic programming to efficiently find subsequences that\n",
        "represent the largest collections of genes appearing in the same relative order.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Tuple, Optional\n",
        "import sys\n",
        "\n",
        "\n",
        "class SubsequenceFinder:\n",
        "    \"\"\"\n",
        "    A class to find longest increasing and decreasing subsequences in permutations.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the subsequence finder.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_permutation(sequence: List[int], n: int) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that the sequence is a valid permutation of 1 to n.\n",
        "\n",
        "        Args:\n",
        "            sequence (List[int]): The sequence to validate\n",
        "            n (int): Expected length and maximum value\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid permutation, False otherwise\n",
        "        \"\"\"\n",
        "        if len(sequence) != n:\n",
        "            return False\n",
        "\n",
        "        # Check if it contains exactly the numbers 1 to n\n",
        "        expected_set = set(range(1, n + 1))\n",
        "        actual_set = set(sequence)\n",
        "\n",
        "        return expected_set == actual_set\n",
        "\n",
        "    def longest_increasing_subsequence(self, sequence: List[int]) -> List[int]:\n",
        "        \"\"\"\n",
        "        Find the longest increasing subsequence using dynamic programming.\n",
        "\n",
        "        Args:\n",
        "            sequence (List[int]): Input sequence\n",
        "\n",
        "        Returns:\n",
        "            List[int]: One of the longest increasing subsequences\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence is empty\n",
        "        \"\"\"\n",
        "        if not sequence:\n",
        "            raise ValueError(\"Sequence cannot be empty\")\n",
        "\n",
        "        n = len(sequence)\n",
        "\n",
        "        # dp[i] stores the length of LIS ending at index i\n",
        "        dp = [1] * n\n",
        "\n",
        "        # prev[i] stores the index of the previous element in the LIS ending at i\n",
        "        prev = [-1] * n\n",
        "\n",
        "        # Fill dp and prev arrays\n",
        "        for i in range(1, n):\n",
        "            for j in range(i):\n",
        "                if sequence[i] > sequence[j] and dp[i] < dp[j] + 1:\n",
        "                    dp[i] = dp[j] + 1\n",
        "                    prev[i] = j\n",
        "\n",
        "        # Find the ending position of the longest increasing subsequence\n",
        "        max_length = max(dp)\n",
        "        max_index = dp.index(max_length)\n",
        "\n",
        "        # Reconstruct the LIS by backtracking\n",
        "        lis = []\n",
        "        current = max_index\n",
        "        while current != -1:\n",
        "            lis.append(sequence[current])\n",
        "            current = prev[current]\n",
        "\n",
        "        # Reverse to get the correct order\n",
        "        lis.reverse()\n",
        "        return lis\n",
        "\n",
        "    def longest_decreasing_subsequence(self, sequence: List[int]) -> List[int]:\n",
        "        \"\"\"\n",
        "        Find the longest decreasing subsequence using dynamic programming.\n",
        "\n",
        "        Args:\n",
        "            sequence (List[int]): Input sequence\n",
        "\n",
        "        Returns:\n",
        "            List[int]: One of the longest decreasing subsequences\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence is empty\n",
        "        \"\"\"\n",
        "        if not sequence:\n",
        "            raise ValueError(\"Sequence cannot be empty\")\n",
        "\n",
        "        n = len(sequence)\n",
        "\n",
        "        # dp[i] stores the length of LDS ending at index i\n",
        "        dp = [1] * n\n",
        "\n",
        "        # prev[i] stores the index of the previous element in the LDS ending at i\n",
        "        prev = [-1] * n\n",
        "\n",
        "        # Fill dp and prev arrays\n",
        "        for i in range(1, n):\n",
        "            for j in range(i):\n",
        "                if sequence[i] < sequence[j] and dp[i] < dp[j] + 1:\n",
        "                    dp[i] = dp[j] + 1\n",
        "                    prev[i] = j\n",
        "\n",
        "        # Find the ending position of the longest decreasing subsequence\n",
        "        max_length = max(dp)\n",
        "        max_index = dp.index(max_length)\n",
        "\n",
        "        # Reconstruct the LDS by backtracking\n",
        "        lds = []\n",
        "        current = max_index\n",
        "        while current != -1:\n",
        "            lds.append(sequence[current])\n",
        "            current = prev[current]\n",
        "\n",
        "        # Reverse to get the correct order\n",
        "        lds.reverse()\n",
        "        return lds\n",
        "\n",
        "    def find_both_subsequences(self, sequence: List[int]) -> Tuple[List[int], List[int]]:\n",
        "        \"\"\"\n",
        "        Find both longest increasing and decreasing subsequences.\n",
        "\n",
        "        Args:\n",
        "            sequence (List[int]): Input sequence\n",
        "\n",
        "        Returns:\n",
        "            Tuple[List[int], List[int]]: (LIS, LDS)\n",
        "        \"\"\"\n",
        "        lis = self.longest_increasing_subsequence(sequence)\n",
        "        lds = self.longest_decreasing_subsequence(sequence)\n",
        "        return lis, lds\n",
        "\n",
        "    def get_subsequence_analysis(self, sequence: List[int]) -> dict:\n",
        "        \"\"\"\n",
        "        Get detailed analysis of the longest subsequences.\n",
        "\n",
        "        Args:\n",
        "            sequence (List[int]): Input sequence\n",
        "\n",
        "        Returns:\n",
        "            dict: Analysis including lengths, subsequences, and statistics\n",
        "        \"\"\"\n",
        "        lis, lds = self.find_both_subsequences(sequence)\n",
        "\n",
        "        analysis = {\n",
        "            'sequence': sequence,\n",
        "            'sequence_length': len(sequence),\n",
        "            'longest_increasing_subsequence': lis,\n",
        "            'lis_length': len(lis),\n",
        "            'longest_decreasing_subsequence': lds,\n",
        "            'lds_length': len(lds),\n",
        "            'lis_percentage': (len(lis) / len(sequence)) * 100,\n",
        "            'lds_percentage': (len(lds) / len(sequence)) * 100\n",
        "        }\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> Tuple[int, List[int]]:\n",
        "    \"\"\"\n",
        "    Parse input file to extract n and the permutation.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[int, List[int]]: (n, permutation)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            lines = file.read().strip().split('\\n')\n",
        "\n",
        "        if len(lines) < 2:\n",
        "            raise ValueError(\"Input file must contain at least 2 lines\")\n",
        "\n",
        "        # Parse n\n",
        "        try:\n",
        "            n = int(lines[0].strip())\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"First line must be an integer, got: '{lines[0]}'\")\n",
        "\n",
        "        # Parse permutation\n",
        "        try:\n",
        "            permutation = list(map(int, lines[1].strip().split()))\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Second line must contain integers, got: '{lines[1]}'\")\n",
        "\n",
        "        # Validate constraints\n",
        "        if n <= 0 or n > 10000:\n",
        "            raise ValueError(f\"n must be between 1 and 10000, got: {n}\")\n",
        "\n",
        "        if len(permutation) != n:\n",
        "            raise ValueError(f\"Permutation length {len(permutation)} doesn't match n={n}\")\n",
        "\n",
        "        return n, permutation\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, lis: List[int], lds: List[int]) -> None:\n",
        "    \"\"\"\n",
        "    Write longest increasing and decreasing subsequences to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        lis (List[int]): Longest increasing subsequence\n",
        "        lds (List[int]): Longest decreasing subsequence\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            file.write(' '.join(map(str, lis)) + '\\n')\n",
        "            file.write(' '.join(map(str, lds)) + '\\n')\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_longest_subsequence_problem(input_file_path: str) -> Tuple[List[int], List[int]]:\n",
        "    \"\"\"\n",
        "    Solve the Longest Subsequence problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[List[int], List[int]]: (LIS, LDS)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        n, permutation = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize finder\n",
        "        finder = SubsequenceFinder()\n",
        "\n",
        "        # Validate permutation (optional but good practice)\n",
        "        if not finder.validate_permutation(permutation, n):\n",
        "            print(f\"Warning: Input may not be a valid permutation of 1 to {n}\")\n",
        "\n",
        "        # Find longest subsequences\n",
        "        lis, lds = finder.find_both_subsequences(permutation)\n",
        "\n",
        "        return lis, lds\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Longest Subsequence problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_lgis.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_lgis.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Longest Increasing Subsequence Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        lis, lds = solve_longest_subsequence_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResults:\")\n",
        "        print(f\"Longest Increasing Subsequence: {' '.join(map(str, lis))}\")\n",
        "        print(f\"LIS Length: {len(lis)}\")\n",
        "        print(f\"Longest Decreasing Subsequence: {' '.join(map(str, lds))}\")\n",
        "        print(f\"LDS Length: {len(lds)}\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, lis, lds)\n",
        "        print(f\"\\nResults written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input from the problem\n",
        "    sample_n = 5\n",
        "    sample_permutation = [5, 1, 4, 2, 3]\n",
        "\n",
        "    print(f\"Input:\")\n",
        "    print(f\"n = {sample_n}\")\n",
        "    print(f\"Permutation = {sample_permutation}\")\n",
        "\n",
        "    finder = SubsequenceFinder()\n",
        "\n",
        "    # Get detailed analysis\n",
        "    analysis = finder.get_subsequence_analysis(sample_permutation)\n",
        "\n",
        "    print(f\"\\nDetailed Analysis:\")\n",
        "    print(f\"Sequence: {analysis['sequence']}\")\n",
        "    print(f\"Length: {analysis['sequence_length']}\")\n",
        "    print(f\"LIS: {analysis['longest_increasing_subsequence']} (length: {analysis['lis_length']})\")\n",
        "    print(f\"LDS: {analysis['longest_decreasing_subsequence']} (length: {analysis['lds_length']})\")\n",
        "    print(f\"LIS covers {analysis['lis_percentage']:.1f}% of sequence\")\n",
        "    print(f\"LDS covers {analysis['lds_percentage']:.1f}% of sequence\")\n",
        "\n",
        "    # Expected output verification\n",
        "    expected_lis = [1, 2, 3]\n",
        "    expected_lds = [5, 4, 2]\n",
        "\n",
        "    print(f\"\\nVerification:\")\n",
        "    print(f\"Expected LIS: {expected_lis}\")\n",
        "    print(f\"Our LIS:      {analysis['longest_increasing_subsequence']}\")\n",
        "    print(f\"LIS Match: {'✓' if analysis['longest_increasing_subsequence'] == expected_lis else '✗'}\")\n",
        "\n",
        "    print(f\"Expected LDS: {expected_lds}\")\n",
        "    print(f\"Our LDS:      {analysis['longest_decreasing_subsequence']}\")\n",
        "    print(f\"LDS Match: {'✓' if analysis['longest_decreasing_subsequence'] == expected_lds else '✗'}\")\n",
        "\n",
        "\n",
        "def demonstrate_algorithm_steps():\n",
        "    \"\"\"\n",
        "    Show step-by-step how the dynamic programming algorithm works.\n",
        "    \"\"\"\n",
        "    print(\"=== Algorithm Demonstration ===\")\n",
        "\n",
        "    sequence = [5, 1, 4, 2, 3]\n",
        "    print(f\"Finding LIS in sequence: {sequence}\")\n",
        "\n",
        "    n = len(sequence)\n",
        "    dp = [1] * n\n",
        "    prev = [-1] * n\n",
        "\n",
        "    print(f\"\\nStep-by-step LIS calculation:\")\n",
        "    print(f\"Initial: dp = {dp}, prev = {prev}\")\n",
        "\n",
        "    for i in range(1, n):\n",
        "        for j in range(i):\n",
        "            if sequence[i] > sequence[j] and dp[i] < dp[j] + 1:\n",
        "                old_dp_i = dp[i]\n",
        "                dp[i] = dp[j] + 1\n",
        "                prev[i] = j\n",
        "                print(f\"Step i={i}, j={j}: {sequence[i]} > {sequence[j]}, \"\n",
        "                      f\"dp[{i}] updated from {old_dp_i} to {dp[i]}, prev[{i}] = {j}\")\n",
        "        print(f\"After position {i}: dp = {dp}, prev = {prev}\")\n",
        "\n",
        "    # Reconstruct\n",
        "    max_length = max(dp)\n",
        "    max_index = dp.index(max_length)\n",
        "    print(f\"\\nReconstruction starting from index {max_index} (max length: {max_length})\")\n",
        "\n",
        "    lis = []\n",
        "    current = max_index\n",
        "    while current != -1:\n",
        "        lis.append(sequence[current])\n",
        "        print(f\"Add {sequence[current]} at index {current}\")\n",
        "        current = prev[current]\n",
        "\n",
        "    lis.reverse()\n",
        "    print(f\"Final LIS: {lis}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Demonstrate algorithm steps\n",
        "    demonstrate_algorithm_steps()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test with additional cases\n",
        "    print(\"=== Additional Test Cases ===\")\n",
        "    finder = SubsequenceFinder()\n",
        "\n",
        "    test_cases = [\n",
        "        ([1, 2, 3, 4, 5], \"Already sorted\"),\n",
        "        ([5, 4, 3, 2, 1], \"Reverse sorted\"),\n",
        "        ([8, 2, 1, 6, 5, 7, 4, 3, 9], \"Example from problem description\"),\n",
        "        ([3, 1, 4, 1, 5, 9, 2, 6], \"Random case\"),\n",
        "    ]\n",
        "\n",
        "    for sequence, description in test_cases:\n",
        "        print(f\"\\n{description}: {sequence}\")\n",
        "        try:\n",
        "            lis, lds = finder.find_both_subsequences(sequence)\n",
        "            print(f\"  LIS: {lis} (length: {len(lis)})\")\n",
        "            print(f\"  LDS: {lds} (length: {len(lds)})\")\n",
        "        except Exception as e:\n",
        "            print(f\"  Error: {e}\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "n3UY-05AlU6Z",
        "outputId": "664ab9c0-2ac5-4e9a-8955-15f06a2158b7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input:\n",
            "n = 5\n",
            "Permutation = [5, 1, 4, 2, 3]\n",
            "\n",
            "Detailed Analysis:\n",
            "Sequence: [5, 1, 4, 2, 3]\n",
            "Length: 5\n",
            "LIS: [1, 2, 3] (length: 3)\n",
            "LDS: [5, 4, 2] (length: 3)\n",
            "LIS covers 60.0% of sequence\n",
            "LDS covers 60.0% of sequence\n",
            "\n",
            "Verification:\n",
            "Expected LIS: [1, 2, 3]\n",
            "Our LIS:      [1, 2, 3]\n",
            "LIS Match: ✓\n",
            "Expected LDS: [5, 4, 2]\n",
            "Our LDS:      [5, 4, 2]\n",
            "LDS Match: ✓\n",
            "\n",
            "============================================================\n",
            "=== Algorithm Demonstration ===\n",
            "Finding LIS in sequence: [5, 1, 4, 2, 3]\n",
            "\n",
            "Step-by-step LIS calculation:\n",
            "Initial: dp = [1, 1, 1, 1, 1], prev = [-1, -1, -1, -1, -1]\n",
            "After position 1: dp = [1, 1, 1, 1, 1], prev = [-1, -1, -1, -1, -1]\n",
            "Step i=2, j=1: 4 > 1, dp[2] updated from 1 to 2, prev[2] = 1\n",
            "After position 2: dp = [1, 1, 2, 1, 1], prev = [-1, -1, 1, -1, -1]\n",
            "Step i=3, j=1: 2 > 1, dp[3] updated from 1 to 2, prev[3] = 1\n",
            "After position 3: dp = [1, 1, 2, 2, 1], prev = [-1, -1, 1, 1, -1]\n",
            "Step i=4, j=1: 3 > 1, dp[4] updated from 1 to 2, prev[4] = 1\n",
            "Step i=4, j=3: 3 > 2, dp[4] updated from 2 to 3, prev[4] = 3\n",
            "After position 4: dp = [1, 1, 2, 2, 3], prev = [-1, -1, 1, 1, 3]\n",
            "\n",
            "Reconstruction starting from index 4 (max length: 3)\n",
            "Add 3 at index 4\n",
            "Add 2 at index 3\n",
            "Add 1 at index 1\n",
            "Final LIS: [1, 2, 3]\n",
            "\n",
            "============================================================\n",
            "=== Additional Test Cases ===\n",
            "\n",
            "Already sorted: [1, 2, 3, 4, 5]\n",
            "  LIS: [1, 2, 3, 4, 5] (length: 5)\n",
            "  LDS: [1] (length: 1)\n",
            "\n",
            "Reverse sorted: [5, 4, 3, 2, 1]\n",
            "  LIS: [5] (length: 1)\n",
            "  LDS: [5, 4, 3, 2, 1] (length: 5)\n",
            "\n",
            "Example from problem description: [8, 2, 1, 6, 5, 7, 4, 3, 9]\n",
            "  LIS: [2, 6, 7, 9] (length: 4)\n",
            "  LDS: [8, 6, 5, 4, 3] (length: 5)\n",
            "\n",
            "Random case: [3, 1, 4, 1, 5, 9, 2, 6]\n",
            "  LIS: [3, 4, 5, 9] (length: 4)\n",
            "  LDS: [3, 1] (length: 2)\n",
            "\n",
            "============================================================\n",
            "Solving Longest Increasing Subsequence Problem...\n",
            "\n",
            "Results:\n",
            "Longest Increasing Subsequence: 140 143 151 174 194 277 368 383 598 688 690 825 832 962 1031 1047 1145 1384 1388 1402 1456 1512 1530 1543 1544 1601 1642 1646 1657 1666 1692 1700 1717 1799 1812 1955 1996 2012 2034 2080 2082 2100 2156 2238 2284 2352 2387 2425 2431 2454 2457 2460 2467 2482 2523 2538 2572 2673 2736 2818 2847 2980 3094 3118 3126 3184 3298 3406 3409 3447 3579 3693 3729 3751 3791 3863 3909 3940 3990 4033 4056 4101 4113 4133 4382 4383 4385 4401 4402 4468 4535 4574 4581 4648 4697 4712 4781 4810 4878 5028 5109 5130 5189 5288 5401 5402 5439 5486 5524 5531 5583 5613 5618 5624 5653 5713 5743 5780 5806 5830 5909 5925 5955 6084 6139 6157 6162 6174 6183 6213 6217 6230 6295 6377 6402 6417 6435 6473 6481 6543 6619 6625 6629 6653 6677 6691 6706 6713 6720 6869 6902 7015 7202 7264 7305 7341 7395 7421 7424 7456 7725 7793 7891 7933 8002 8010 8046 8145 8189 8272 8437 8510 8567 8607 8616 8707 8738 8767 8813 8865 8927 8935 9002 9007 9021\n",
            "LIS Length: 185\n",
            "Longest Decreasing Subsequence: 8914 8902 8803 8663 8605 8596 8574 8541 8294 8225 8054 8025 7805 7788 7774 7713 7557 7536 7505 7465 7457 7179 7107 7082 7066 7026 7012 6995 6956 6905 6881 6864 6747 6705 6644 6635 6617 6531 6453 6432 6394 6285 6223 6006 5998 5988 5953 5651 5647 5603 5599 5563 5552 5551 5484 5483 5479 5446 5337 5320 5302 5266 5165 5141 5070 5043 5026 4906 4842 4783 4743 4673 4653 4602 4527 4471 4459 4447 4423 4372 4323 4322 4268 4213 4177 4152 4148 4140 4135 4065 4028 3990 3986 3965 3915 3894 3761 3748 3746 3738 3669 3637 3600 3415 3408 3393 3347 3335 3332 3300 3253 3166 3145 3133 3124 3086 3076 3061 3048 3039 2984 2950 2923 2909 2880 2878 2760 2743 2723 2712 2631 2545 2416 2357 2336 2235 2147 2116 2088 2019 2002 1983 1967 1956 1947 1923 1915 1911 1910 1879 1832 1795 1698 1653 1571 1503 1483 1437 1343 1285 1249 1171 1142 1071 982 955 782 744 644 630 626 615 574 549 540 523 461 430 416 393 382 358 281 259 254 214 192 155 135\n",
            "LDS Length: 189\n",
            "\n",
            "Results written to: output_lgis.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Enumerating k-mers Lexicographically"
      ],
      "metadata": {
        "id": "bfq2ghLkdaDA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Enumerating k-mers Lexicographically Problem Solution\n",
        "\n",
        "This module generates all possible strings of length n from a given ordered alphabet\n",
        "in lexicographic order. This is useful for cataloguing genetic strings and creating\n",
        "systematic orderings of sequence data.\n",
        "\n",
        "The problem generates the Cartesian product of an alphabet with itself n times,\n",
        "producing alphabet^n total strings in lexicographic order.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Iterator, Tuple\n",
        "import itertools\n",
        "import math\n",
        "\n",
        "\n",
        "class LexicographicStringGenerator:\n",
        "    \"\"\"\n",
        "    A class to generate strings of specified length from an ordered alphabet\n",
        "    in lexicographic order.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the lexicographic string generator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_alphabet(alphabet: List[str]) -> bool:\n",
        "        \"\"\"\n",
        "        Validate the input alphabet.\n",
        "\n",
        "        Args:\n",
        "            alphabet (List[str]): List of alphabet symbols\n",
        "\n",
        "        Returns:\n",
        "            bool: True if alphabet is valid, False otherwise\n",
        "        \"\"\"\n",
        "        if not alphabet:\n",
        "            return False\n",
        "\n",
        "        # Check for duplicates\n",
        "        if len(alphabet) != len(set(alphabet)):\n",
        "            return False\n",
        "\n",
        "        # Check constraint: at most 10 symbols\n",
        "        if len(alphabet) > 10:\n",
        "            return False\n",
        "\n",
        "        # Check that all symbols are single characters (optional but good practice)\n",
        "        return all(len(symbol) == 1 for symbol in alphabet)\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_n(n: int, max_alphabet_size: int) -> bool:\n",
        "        \"\"\"\n",
        "        Validate the string length parameter.\n",
        "\n",
        "        Args:\n",
        "            n (int): Desired string length\n",
        "            max_alphabet_size (int): Size of alphabet for complexity estimation\n",
        "\n",
        "        Returns:\n",
        "            bool: True if n is valid, False otherwise\n",
        "        \"\"\"\n",
        "        if not isinstance(n, int) or n <= 0:\n",
        "            return False\n",
        "\n",
        "        # Check constraint: n ≤ 10\n",
        "        if n > 10:\n",
        "            return False\n",
        "\n",
        "        # Optional: Check if the total number of strings would be manageable\n",
        "        total_strings = max_alphabet_size ** n\n",
        "        if total_strings > 10**6:  # Reasonable limit for memory/time\n",
        "            return False\n",
        "\n",
        "        return True\n",
        "\n",
        "    def calculate_total_strings(self, alphabet_size: int, n: int) -> int:\n",
        "        \"\"\"\n",
        "        Calculate the total number of strings that will be generated.\n",
        "\n",
        "        Args:\n",
        "            alphabet_size (int): Size of the alphabet\n",
        "            n (int): Length of strings\n",
        "\n",
        "        Returns:\n",
        "            int: Total number of strings (alphabet_size^n)\n",
        "        \"\"\"\n",
        "        return alphabet_size ** n\n",
        "\n",
        "    def generate_strings_iterator(self, alphabet: List[str], n: int) -> Iterator[str]:\n",
        "        \"\"\"\n",
        "        Generate all strings of length n from alphabet in lexicographic order.\n",
        "        Uses iterator for memory efficiency.\n",
        "\n",
        "        Args:\n",
        "            alphabet (List[str]): Ordered alphabet symbols\n",
        "            n (int): Length of strings to generate\n",
        "\n",
        "        Yields:\n",
        "            str: Each generated string in lexicographic order\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If alphabet or n is invalid\n",
        "        \"\"\"\n",
        "        # Validate inputs\n",
        "        if not self.validate_alphabet(alphabet):\n",
        "            raise ValueError(\"Invalid alphabet: must be non-empty, unique symbols, ≤10 characters\")\n",
        "\n",
        "        if not self.validate_n(n, len(alphabet)):\n",
        "            raise ValueError(\"Invalid n: must be positive integer ≤10\")\n",
        "\n",
        "        # Generate all combinations using Cartesian product\n",
        "        for combination in itertools.product(alphabet, repeat=n):\n",
        "            yield ''.join(combination)\n",
        "\n",
        "    def generate_all_strings(self, alphabet: List[str], n: int) -> List[str]:\n",
        "        \"\"\"\n",
        "        Generate all strings and return as a list.\n",
        "\n",
        "        Args:\n",
        "            alphabet (List[str]): Ordered alphabet symbols\n",
        "            n (int): Length of strings to generate\n",
        "\n",
        "        Returns:\n",
        "            List[str]: All generated strings in lexicographic order\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If alphabet or n is invalid\n",
        "        \"\"\"\n",
        "        return list(self.generate_strings_iterator(alphabet, n))\n",
        "\n",
        "    def get_generation_info(self, alphabet: List[str], n: int) -> dict:\n",
        "        \"\"\"\n",
        "        Get information about the string generation process.\n",
        "\n",
        "        Args:\n",
        "            alphabet (List[str]): Ordered alphabet symbols\n",
        "            n (int): Length of strings to generate\n",
        "\n",
        "        Returns:\n",
        "            dict: Information about the generation process\n",
        "        \"\"\"\n",
        "        alphabet_size = len(alphabet)\n",
        "        total_strings = self.calculate_total_strings(alphabet_size, n)\n",
        "\n",
        "        return {\n",
        "            'alphabet': alphabet,\n",
        "            'alphabet_size': alphabet_size,\n",
        "            'string_length': n,\n",
        "            'total_strings': total_strings,\n",
        "            'memory_estimate_mb': (total_strings * n * 4) / (1024 * 1024),  # Rough estimate\n",
        "            'is_manageable': total_strings <= 10**5\n",
        "        }\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> Tuple[List[str], int]:\n",
        "    \"\"\"\n",
        "    Parse input file to extract alphabet and n.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[List[str], int]: (alphabet, n)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            lines = file.read().strip().split('\\n')\n",
        "\n",
        "        if len(lines) < 2:\n",
        "            raise ValueError(\"Input file must contain at least 2 lines\")\n",
        "\n",
        "        # Parse alphabet (first line)\n",
        "        alphabet = lines[0].strip().split()\n",
        "        if not alphabet:\n",
        "            raise ValueError(\"First line must contain alphabet symbols\")\n",
        "\n",
        "        # Parse n (second line)\n",
        "        try:\n",
        "            n = int(lines[1].strip())\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Second line must be an integer, got: '{lines[1]}'\")\n",
        "\n",
        "        return alphabet, n\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, strings: Iterator[str]) -> None:\n",
        "    \"\"\"\n",
        "    Write generated strings to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        strings (Iterator[str]): Iterator of strings to write\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            for string in strings:\n",
        "                file.write(string + '\\n')\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_lexicographic_strings_problem(input_file_path: str) -> Iterator[str]:\n",
        "    \"\"\"\n",
        "    Solve the Lexicographic Strings problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Iterator[str]: Iterator of generated strings\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        alphabet, n = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize generator\n",
        "        generator = LexicographicStringGenerator()\n",
        "\n",
        "        # Generate strings\n",
        "        return generator.generate_strings_iterator(alphabet, n)\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Lexicographic Strings problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_lexf.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_lexf.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Lexicographic Strings Problem...\")\n",
        "\n",
        "        # Parse input first to get info\n",
        "        alphabet, n = parse_input_file(input_file)\n",
        "\n",
        "        # Get generation info\n",
        "        generator = LexicographicStringGenerator()\n",
        "        info = generator.get_generation_info(alphabet, n)\n",
        "\n",
        "        print(f\"\\nInput Information:\")\n",
        "        print(f\"Alphabet: {info['alphabet']}\")\n",
        "        print(f\"Alphabet size: {info['alphabet_size']}\")\n",
        "        print(f\"String length: {info['string_length']}\")\n",
        "        print(f\"Total strings to generate: {info['total_strings']:,}\")\n",
        "        print(f\"Estimated memory usage: {info['memory_estimate_mb']:.2f} MB\")\n",
        "\n",
        "        if not info['is_manageable']:\n",
        "            print(\"\\nWarning: Large number of strings. Consider using iterator approach.\")\n",
        "\n",
        "        # Generate strings using iterator for memory efficiency\n",
        "        strings_iterator = generator.generate_strings_iterator(alphabet, n)\n",
        "\n",
        "        # Display first few results\n",
        "        print(f\"\\nFirst few generated strings:\")\n",
        "        preview_count = min(10, info['total_strings'])\n",
        "        preview_strings = []\n",
        "\n",
        "        for i, string in enumerate(strings_iterator):\n",
        "            if i < preview_count:\n",
        "                preview_strings.append(string)\n",
        "                print(string)\n",
        "            else:\n",
        "                break\n",
        "\n",
        "        if info['total_strings'] > preview_count:\n",
        "            print(\"...\")\n",
        "            print(f\"(and {info['total_strings'] - preview_count:,} more)\")\n",
        "\n",
        "        # Write all strings to output file (regenerate iterator)\n",
        "        strings_iterator = generator.generate_strings_iterator(alphabet, n)\n",
        "        write_output_file(output_file, strings_iterator)\n",
        "        print(f\"\\nAll strings written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input\n",
        "    sample_alphabet = ['A', 'C', 'G', 'T']\n",
        "    sample_n = 2\n",
        "\n",
        "    print(f\"Input:\")\n",
        "    print(f\"Alphabet: {sample_alphabet}\")\n",
        "    print(f\"String length: {sample_n}\")\n",
        "\n",
        "    generator = LexicographicStringGenerator()\n",
        "\n",
        "    # Get info\n",
        "    info = generator.get_generation_info(sample_alphabet, sample_n)\n",
        "    print(f\"\\nGeneration Info:\")\n",
        "    print(f\"Total strings: {info['total_strings']}\")\n",
        "    print(f\"Expected: {len(sample_alphabet)**sample_n}\")\n",
        "\n",
        "    # Generate and display all strings\n",
        "    print(f\"\\nGenerated strings:\")\n",
        "    strings = generator.generate_all_strings(sample_alphabet, sample_n)\n",
        "\n",
        "    for string in strings:\n",
        "        print(string)\n",
        "\n",
        "    # Verify count\n",
        "    print(f\"\\nVerification:\")\n",
        "    print(f\"Generated {len(strings)} strings\")\n",
        "    print(f\"Expected {info['total_strings']} strings\")\n",
        "    print(f\"Match: {'✓' if len(strings) == info['total_strings'] else '✗'}\")\n",
        "\n",
        "    # Check lexicographic ordering\n",
        "    is_sorted = strings == sorted(strings)\n",
        "    print(f\"Lexicographically ordered: {'✓' if is_sorted else '✗'}\")\n",
        "\n",
        "\n",
        "def test_edge_cases():\n",
        "    \"\"\"\n",
        "    Test various edge cases and constraints.\n",
        "    \"\"\"\n",
        "    print(\"=== Testing Edge Cases ===\")\n",
        "\n",
        "    generator = LexicographicStringGenerator()\n",
        "\n",
        "    test_cases = [\n",
        "        (['A'], 1, \"Single symbol, length 1\"),\n",
        "        (['A'], 3, \"Single symbol, length 3\"),\n",
        "        (['A', 'B'], 1, \"Two symbols, length 1\"),\n",
        "        (['X', 'Y', 'Z'], 2, \"Three symbols, length 2\"),\n",
        "        (['A', 'C', 'G', 'T'], 1, \"DNA alphabet, length 1\"),\n",
        "    ]\n",
        "\n",
        "    for alphabet, n, description in test_cases:\n",
        "        try:\n",
        "            info = generator.get_generation_info(alphabet, n)\n",
        "            strings = generator.generate_all_strings(alphabet, n)\n",
        "            print(f\"\\n{description}:\")\n",
        "            print(f\"  Input: alphabet={alphabet}, n={n}\")\n",
        "            print(f\"  Output: {len(strings)} strings\")\n",
        "            print(f\"  First few: {strings[:min(5, len(strings))]}\")\n",
        "            if len(strings) > 5:\n",
        "                print(f\"  Last few: {strings[-min(3, len(strings)):]}\")\n",
        "        except Exception as e:\n",
        "            print(f\"\\n{description}: Error - {e}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test edge cases\n",
        "    test_edge_cases()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test algorithm efficiency\n",
        "    print(\"=== Algorithm Efficiency Test ===\")\n",
        "    generator = LexicographicStringGenerator()\n",
        "\n",
        "    # Test with larger alphabets\n",
        "    efficiency_tests = [\n",
        "        (['A', 'B', 'C'], 4),\n",
        "        (['A', 'C', 'G', 'T'], 3),\n",
        "        (['A', 'B', 'C', 'D', 'E'], 3),\n",
        "    ]\n",
        "\n",
        "    for alphabet, n in efficiency_tests:\n",
        "        info = generator.get_generation_info(alphabet, n)\n",
        "        print(f\"Alphabet size {len(alphabet)}, length {n}: {info['total_strings']:,} strings\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xa2YM90Gdbnv",
        "outputId": "381e124f-453f-4432-9732-0e1486a3862d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input:\n",
            "Alphabet: ['A', 'C', 'G', 'T']\n",
            "String length: 2\n",
            "\n",
            "Generation Info:\n",
            "Total strings: 16\n",
            "Expected: 16\n",
            "\n",
            "Generated strings:\n",
            "AA\n",
            "AC\n",
            "AG\n",
            "AT\n",
            "CA\n",
            "CC\n",
            "CG\n",
            "CT\n",
            "GA\n",
            "GC\n",
            "GG\n",
            "GT\n",
            "TA\n",
            "TC\n",
            "TG\n",
            "TT\n",
            "\n",
            "Verification:\n",
            "Generated 16 strings\n",
            "Expected 16 strings\n",
            "Match: ✓\n",
            "Lexicographically ordered: ✓\n",
            "\n",
            "============================================================\n",
            "=== Testing Edge Cases ===\n",
            "\n",
            "Single symbol, length 1:\n",
            "  Input: alphabet=['A'], n=1\n",
            "  Output: 1 strings\n",
            "  First few: ['A']\n",
            "\n",
            "Single symbol, length 3:\n",
            "  Input: alphabet=['A'], n=3\n",
            "  Output: 1 strings\n",
            "  First few: ['AAA']\n",
            "\n",
            "Two symbols, length 1:\n",
            "  Input: alphabet=['A', 'B'], n=1\n",
            "  Output: 2 strings\n",
            "  First few: ['A', 'B']\n",
            "\n",
            "Three symbols, length 2:\n",
            "  Input: alphabet=['X', 'Y', 'Z'], n=2\n",
            "  Output: 9 strings\n",
            "  First few: ['XX', 'XY', 'XZ', 'YX', 'YY']\n",
            "  Last few: ['ZX', 'ZY', 'ZZ']\n",
            "\n",
            "DNA alphabet, length 1:\n",
            "  Input: alphabet=['A', 'C', 'G', 'T'], n=1\n",
            "  Output: 4 strings\n",
            "  First few: ['A', 'C', 'G', 'T']\n",
            "\n",
            "============================================================\n",
            "=== Algorithm Efficiency Test ===\n",
            "Alphabet size 3, length 4: 81 strings\n",
            "Alphabet size 4, length 3: 64 strings\n",
            "Alphabet size 5, length 3: 125 strings\n",
            "\n",
            "============================================================\n",
            "Solving Lexicographic Strings Problem...\n",
            "\n",
            "Input Information:\n",
            "Alphabet: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']\n",
            "Alphabet size: 8\n",
            "String length: 3\n",
            "Total strings to generate: 512\n",
            "Estimated memory usage: 0.01 MB\n",
            "\n",
            "First few generated strings:\n",
            "AAA\n",
            "AAB\n",
            "AAC\n",
            "AAD\n",
            "AAE\n",
            "AAF\n",
            "AAG\n",
            "AAH\n",
            "ABA\n",
            "ABB\n",
            "...\n",
            "(and 502 more)\n",
            "\n",
            "All strings written to: output_lexf.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "id": "l0uFvO9Jee_d"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## RNA Splicing Problem Solution"
      ],
      "metadata": {
        "id": "eQkXr_USjZpT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind RNA Splicing Problem Solution\n",
        "\n",
        "This module removes introns from DNA sequences, then transcribes and translates\n",
        "the remaining exons to produce a protein string. This simulates the process of\n",
        "RNA splicing where introns are removed and exons are concatenated before translation.\n",
        "\n",
        "The problem involves:\n",
        "1. Removing intron sequences from the main DNA string\n",
        "2. Transcribing the resulting DNA to RNA (T -> U)\n",
        "3. Translating the RNA to protein using the genetic code\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Tuple, Optional\n",
        "import re\n",
        "\n",
        "\n",
        "class RNASplicer:\n",
        "    \"\"\"\n",
        "    A class to handle RNA splicing operations including intron removal,\n",
        "    transcription, and translation.\n",
        "    \"\"\"\n",
        "\n",
        "    # Standard RNA codon table\n",
        "    RNA_CODON_TABLE = {\n",
        "        'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
        "        'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
        "        'UAU': 'Y', 'UAC': 'Y', 'UAA': '*', 'UAG': '*',\n",
        "        'UGU': 'C', 'UGC': 'C', 'UGA': '*', 'UGG': 'W',\n",
        "        'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
        "        'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
        "        'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
        "        'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
        "        'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
        "        'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
        "        'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
        "        'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
        "        'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
        "        'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
        "        'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
        "        'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
        "    }\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the RNA splicer.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_dna_sequence(sequence: str) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that a sequence contains only valid DNA bases.\n",
        "\n",
        "        Args:\n",
        "            sequence (str): DNA sequence to validate\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid, False otherwise\n",
        "        \"\"\"\n",
        "        return bool(re.match(r'^[ATGC]*$', sequence.upper()))\n",
        "\n",
        "    def remove_introns(self, dna_sequence: str, introns: List[str]) -> str:\n",
        "        \"\"\"\n",
        "        Remove all intron sequences from the main DNA sequence.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Main DNA sequence\n",
        "            introns (List[str]): List of intron sequences to remove\n",
        "\n",
        "        Returns:\n",
        "            str: DNA sequence with introns removed (exons only)\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If DNA sequence or introns contain invalid bases\n",
        "        \"\"\"\n",
        "        # Validate input sequences\n",
        "        if not self.validate_dna_sequence(dna_sequence):\n",
        "            raise ValueError(\"Main DNA sequence contains invalid bases\")\n",
        "\n",
        "        for intron in introns:\n",
        "            if not self.validate_dna_sequence(intron):\n",
        "                raise ValueError(f\"Intron sequence contains invalid bases: {intron}\")\n",
        "\n",
        "        # Start with the original sequence\n",
        "        result = dna_sequence.upper()\n",
        "\n",
        "        # Remove each intron from the sequence\n",
        "        for intron in introns:\n",
        "            intron_upper = intron.upper()\n",
        "            # Remove all occurrences of this intron\n",
        "            while intron_upper in result:\n",
        "                result = result.replace(intron_upper, '', 1)  # Remove one occurrence at a time\n",
        "\n",
        "        return result\n",
        "\n",
        "    @staticmethod\n",
        "    def transcribe_dna_to_rna(dna_sequence: str) -> str:\n",
        "        \"\"\"\n",
        "        Transcribe DNA sequence to RNA by replacing T with U.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): DNA sequence\n",
        "\n",
        "        Returns:\n",
        "            str: RNA sequence\n",
        "        \"\"\"\n",
        "        return dna_sequence.upper().replace('T', 'U')\n",
        "\n",
        "    def translate_rna_to_protein(self, rna_sequence: str) -> str:\n",
        "        \"\"\"\n",
        "        Translate RNA sequence to protein using the genetic code.\n",
        "\n",
        "        Args:\n",
        "            rna_sequence (str): RNA sequence\n",
        "\n",
        "        Returns:\n",
        "            str: Protein sequence\n",
        "        \"\"\"\n",
        "        protein = []\n",
        "\n",
        "        # Translate codons one by one\n",
        "        for i in range(0, len(rna_sequence) - 2, 3):\n",
        "            codon = rna_sequence[i:i+3]\n",
        "\n",
        "            # Skip incomplete codons\n",
        "            if len(codon) != 3:\n",
        "                break\n",
        "\n",
        "            # Get amino acid for this codon\n",
        "            amino_acid = self.RNA_CODON_TABLE.get(codon.upper())\n",
        "\n",
        "            if amino_acid is None:\n",
        "                # Invalid codon - skip it\n",
        "                continue\n",
        "            elif amino_acid == '*':\n",
        "                # Stop codon - end translation\n",
        "                break\n",
        "            else:\n",
        "                protein.append(amino_acid)\n",
        "\n",
        "        return ''.join(protein)\n",
        "\n",
        "    def process_rna_splicing(self, dna_sequence: str, introns: List[str]) -> str:\n",
        "        \"\"\"\n",
        "        Complete RNA splicing process: remove introns, transcribe, and translate.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Main DNA sequence\n",
        "            introns (List[str]): List of intron sequences\n",
        "\n",
        "        Returns:\n",
        "            str: Final protein sequence\n",
        "        \"\"\"\n",
        "        # Step 1: Remove introns to get exons\n",
        "        exons = self.remove_introns(dna_sequence, introns)\n",
        "\n",
        "        # Step 2: Transcribe DNA to RNA\n",
        "        rna = self.transcribe_dna_to_rna(exons)\n",
        "\n",
        "        # Step 3: Translate RNA to protein\n",
        "        protein = self.translate_rna_to_protein(rna)\n",
        "\n",
        "        return protein\n",
        "\n",
        "    def get_detailed_analysis(self, dna_sequence: str, introns: List[str]) -> dict:\n",
        "        \"\"\"\n",
        "        Get detailed step-by-step analysis of the RNA splicing process.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Main DNA sequence\n",
        "            introns (List[str]): List of intron sequences\n",
        "\n",
        "        Returns:\n",
        "            dict: Detailed analysis of each step\n",
        "        \"\"\"\n",
        "        analysis = {}\n",
        "\n",
        "        # Original DNA\n",
        "        analysis['original_dna'] = dna_sequence.upper()\n",
        "        analysis['original_length'] = len(dna_sequence)\n",
        "\n",
        "        # Introns\n",
        "        analysis['introns'] = [intron.upper() for intron in introns]\n",
        "        analysis['num_introns'] = len(introns)\n",
        "\n",
        "        # After intron removal\n",
        "        exons = self.remove_introns(dna_sequence, introns)\n",
        "        analysis['exons'] = exons\n",
        "        analysis['exons_length'] = len(exons)\n",
        "        analysis['removed_bases'] = len(dna_sequence) - len(exons)\n",
        "\n",
        "        # RNA transcription\n",
        "        rna = self.transcribe_dna_to_rna(exons)\n",
        "        analysis['rna'] = rna\n",
        "\n",
        "        # Protein translation\n",
        "        protein = self.translate_rna_to_protein(rna)\n",
        "        analysis['protein'] = protein\n",
        "        analysis['protein_length'] = len(protein)\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_fasta_file(file_path: str) -> List[Tuple[str, str]]:\n",
        "    \"\"\"\n",
        "    Parse FASTA file to extract all sequences.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to FASTA file\n",
        "\n",
        "    Returns:\n",
        "        List[Tuple[str, str]]: List of (header, sequence) tuples\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        return parse_fasta_string(content)\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def parse_fasta_string(fasta_content: str) -> List[Tuple[str, str]]:\n",
        "    \"\"\"\n",
        "    Parse FASTA format string to extract sequences.\n",
        "\n",
        "    Args:\n",
        "        fasta_content (str): FASTA format content\n",
        "\n",
        "    Returns:\n",
        "        List[Tuple[str, str]]: List of (header, sequence) tuples\n",
        "    \"\"\"\n",
        "    sequences = []\n",
        "    header = None\n",
        "    seq_lines = []\n",
        "\n",
        "    for line in fasta_content.strip().split('\\n'):\n",
        "        line = line.strip()\n",
        "        if line.startswith('>'):\n",
        "            # Save previous sequence if exists\n",
        "            if header is not None:\n",
        "                sequences.append((header, ''.join(seq_lines)))\n",
        "            # Start new sequence\n",
        "            header = line[1:]  # Remove '>' character\n",
        "            seq_lines = []\n",
        "        else:\n",
        "            seq_lines.append(line)\n",
        "\n",
        "    # Add the last sequence\n",
        "    if header is not None:\n",
        "        sequences.append((header, ''.join(seq_lines)))\n",
        "\n",
        "    if not sequences:\n",
        "        raise ValueError(\"No valid FASTA sequences found\")\n",
        "\n",
        "    return sequences\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, protein: str) -> None:\n",
        "    \"\"\"\n",
        "    Write protein sequence to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        protein (str): Protein sequence\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            file.write(protein + '\\n')\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_rna_splicing_problem(input_file_path: str) -> str:\n",
        "    \"\"\"\n",
        "    Solve the RNA Splicing problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input FASTA file\n",
        "\n",
        "    Returns:\n",
        "        str: Final protein sequence\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse FASTA file\n",
        "        sequences = parse_fasta_file(input_file_path)\n",
        "\n",
        "        if len(sequences) < 1:\n",
        "            raise ValueError(\"At least one DNA sequence is required\")\n",
        "\n",
        "        # First sequence is the main DNA string\n",
        "        main_dna = sequences[0][1]\n",
        "\n",
        "        # Remaining sequences are introns\n",
        "        introns = [seq for _, seq in sequences[1:]]\n",
        "\n",
        "        # Validate length constraint\n",
        "        if len(main_dna) > 1000:\n",
        "            raise ValueError(f\"DNA sequence length {len(main_dna)} exceeds maximum of 1000 bp\")\n",
        "\n",
        "        # Initialize splicer\n",
        "        splicer = RNASplicer()\n",
        "\n",
        "        # Process RNA splicing\n",
        "        protein = splicer.process_rna_splicing(main_dna, introns)\n",
        "\n",
        "        return protein\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the RNA Splicing problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_splc.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_splc.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving RNA Splicing Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        protein = solve_rna_splicing_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResult:\")\n",
        "        print(f\"Final protein sequence: {protein}\")\n",
        "        print(f\"Protein length: {len(protein)} amino acids\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, protein)\n",
        "        print(f\"\\nResult written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data from the search results.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample FASTA content from the problem\n",
        "    sample_fasta = \"\"\">Rosalind_10\n",
        "ATGGTCTACATAGCTGACAAACAGCACGTAGCAATCGGTCGAATCTCGAGAGGCATATGGTCACATGATCGGTCGAGCGTGTTTCAAAGTTTGCGCCTAG\n",
        ">Rosalind_12\n",
        "ATCGGTCGAA\n",
        ">Rosalind_15\n",
        "ATCGGTCGAGCGTGT\"\"\"\n",
        "\n",
        "    # Parse sequences\n",
        "    sequences = parse_fasta_string(sample_fasta)\n",
        "    print(f\"Parsed {len(sequences)} sequences:\")\n",
        "    for i, (header, seq) in enumerate(sequences):\n",
        "        print(f\"  {i+1}. {header}: {len(seq)} bp\")\n",
        "\n",
        "    # Extract main DNA and introns\n",
        "    main_dna = sequences[0][1]\n",
        "    introns = [seq for _, seq in sequences[1:]]\n",
        "\n",
        "    print(f\"\\nMain DNA sequence ({len(main_dna)} bp):\")\n",
        "    print(f\"{main_dna}\")\n",
        "    print(f\"\\nIntrons ({len(introns)} total):\")\n",
        "    for i, intron in enumerate(introns, 1):\n",
        "        print(f\"  {i}. {intron} ({len(intron)} bp)\")\n",
        "\n",
        "    # Process with detailed analysis\n",
        "    splicer = RNASplicer()\n",
        "    analysis = splicer.get_detailed_analysis(main_dna, introns)\n",
        "\n",
        "    print(f\"\\n=== Step-by-Step Analysis ===\")\n",
        "    print(f\"Original DNA: {analysis['original_dna']}\")\n",
        "    print(f\"After removing introns: {analysis['exons']}\")\n",
        "    print(f\"RNA: {analysis['rna']}\")\n",
        "    print(f\"Protein: {analysis['protein']}\")\n",
        "\n",
        "    print(f\"\\n=== Summary ===\")\n",
        "    print(f\"Original length: {analysis['original_length']} bp\")\n",
        "    print(f\"Removed bases: {analysis['removed_bases']} bp\")\n",
        "    print(f\"Final exons: {analysis['exons_length']} bp\")\n",
        "    print(f\"Final protein: {analysis['protein_length']} amino acids\")\n",
        "\n",
        "    # Expected output\n",
        "    print(f\"\\nExpected output: MVYIADKQHVASREAYGHMFKVCA\")\n",
        "    print(f\"Our result:      {analysis['protein']}\")\n",
        "    print(f\"Match: {'✓' if analysis['protein'] == 'MVYIADKQHVASREAYGHMFKVCA' else '✗'}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test edge cases\n",
        "    print(\"=== Testing Edge Cases ===\")\n",
        "    splicer = RNASplicer()\n",
        "\n",
        "    # Test with no introns\n",
        "    test_dna = \"ATGAAATTCTAG\"  # Simple gene: ATG AAA TTC TAG\n",
        "    test_protein = splicer.process_rna_splicing(test_dna, [])\n",
        "    print(f\"No introns: {test_dna} -> {test_protein}\")\n",
        "\n",
        "    # Test with overlapping removal\n",
        "    test_dna2 = \"ATGAAACCCTTTGGG\"\n",
        "    test_introns = [\"AAA\", \"TTT\"]\n",
        "    test_protein2 = splicer.process_rna_splicing(test_dna2, test_introns)\n",
        "    print(f\"With introns: {test_dna2} -> {test_protein2}\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6fJQUbfwjbKp",
        "outputId": "027b34df-66fa-4114-e75f-031c62be6659"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Parsed 3 sequences:\n",
            "  1. Rosalind_10: 100 bp\n",
            "  2. Rosalind_12: 10 bp\n",
            "  3. Rosalind_15: 15 bp\n",
            "\n",
            "Main DNA sequence (100 bp):\n",
            "ATGGTCTACATAGCTGACAAACAGCACGTAGCAATCGGTCGAATCTCGAGAGGCATATGGTCACATGATCGGTCGAGCGTGTTTCAAAGTTTGCGCCTAG\n",
            "\n",
            "Introns (2 total):\n",
            "  1. ATCGGTCGAA (10 bp)\n",
            "  2. ATCGGTCGAGCGTGT (15 bp)\n",
            "\n",
            "=== Step-by-Step Analysis ===\n",
            "Original DNA: ATGGTCTACATAGCTGACAAACAGCACGTAGCAATCGGTCGAATCTCGAGAGGCATATGGTCACATGATCGGTCGAGCGTGTTTCAAAGTTTGCGCCTAG\n",
            "After removing introns: ATGGTCTACATAGCTGACAAACAGCACGTAGCATCTCGAGAGGCATATGGTCACATGTTCAAAGTTTGCGCCTAG\n",
            "RNA: AUGGUCUACAUAGCUGACAAACAGCACGUAGCAUCUCGAGAGGCAUAUGGUCACAUGUUCAAAGUUUGCGCCUAG\n",
            "Protein: MVYIADKQHVASREAYGHMFKVCA\n",
            "\n",
            "=== Summary ===\n",
            "Original length: 100 bp\n",
            "Removed bases: 25 bp\n",
            "Final exons: 75 bp\n",
            "Final protein: 24 amino acids\n",
            "\n",
            "Expected output: MVYIADKQHVASREAYGHMFKVCA\n",
            "Our result:      MVYIADKQHVASREAYGHMFKVCA\n",
            "Match: ✓\n",
            "\n",
            "============================================================\n",
            "=== Testing Edge Cases ===\n",
            "No introns: ATGAAATTCTAG -> MKF\n",
            "With introns: ATGAAACCCTTTGGG -> MPG\n",
            "\n",
            "============================================================\n",
            "Solving RNA Splicing Problem...\n",
            "\n",
            "Result:\n",
            "Final protein sequence: MCQKSSLRRGKLQLLDSTGLGSSFSIGNTRQLGGGKSLFESDPTSVKPNFEKSFFRPRLLSLRFYRPSRSWLIHPSGRNAKDNELYLPGTLIFLIHWELRRHTALWLLYAGCHGWRLIRVRGVLERVLYMSWKAPGPSFTRLNRTVALSVGAGGRLATIFYLYKDQ\n",
            "Protein length: 166 amino acids\n",
            "\n",
            "Result written to: output_splc.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Locating Restriction Sites Problem Solution"
      ],
      "metadata": {
        "id": "kQltb3rFdP9x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Locating Restriction Sites Problem Solution\n",
        "\n",
        "This module finds reverse palindromes in DNA sequences. A reverse palindrome\n",
        "is a DNA string that equals its reverse complement, which are recognition sites\n",
        "for restriction enzymes used by bacteria to defend against phages.\n",
        "\n",
        "The problem asks for all reverse palindromes of length 4-12 with their positions.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Tuple, Optional\n",
        "import re\n",
        "\n",
        "\n",
        "class RestrictionSiteFinder:\n",
        "    \"\"\"\n",
        "    A class to find restriction sites (reverse palindromes) in DNA sequences.\n",
        "    \"\"\"\n",
        "\n",
        "    # DNA complement mapping\n",
        "    COMPLEMENT_MAP = {\n",
        "        'A': 'T',\n",
        "        'T': 'A',\n",
        "        'G': 'C',\n",
        "        'C': 'G'\n",
        "    }\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the restriction site finder.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def get_complement(dna_sequence: str) -> str:\n",
        "        \"\"\"\n",
        "        Get the complement of a DNA sequence.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Input DNA sequence\n",
        "\n",
        "        Returns:\n",
        "            str: Complement sequence\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence contains invalid bases\n",
        "        \"\"\"\n",
        "        try:\n",
        "            return ''.join(RestrictionSiteFinder.COMPLEMENT_MAP[base.upper()]\n",
        "                          for base in dna_sequence)\n",
        "        except KeyError as e:\n",
        "            raise ValueError(f\"Invalid DNA base found: {e}\")\n",
        "\n",
        "    @staticmethod\n",
        "    def get_reverse_complement(dna_sequence: str) -> str:\n",
        "        \"\"\"\n",
        "        Get the reverse complement of a DNA sequence.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Input DNA sequence\n",
        "\n",
        "        Returns:\n",
        "            str: Reverse complement sequence\n",
        "        \"\"\"\n",
        "        complement = RestrictionSiteFinder.get_complement(dna_sequence)\n",
        "        return complement[::-1]  # Reverse the complement\n",
        "\n",
        "    @staticmethod\n",
        "    def is_reverse_palindrome(dna_sequence: str) -> bool:\n",
        "        \"\"\"\n",
        "        Check if a DNA sequence is a reverse palindrome.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): DNA sequence to check\n",
        "\n",
        "        Returns:\n",
        "            bool: True if sequence equals its reverse complement\n",
        "        \"\"\"\n",
        "        try:\n",
        "            reverse_comp = RestrictionSiteFinder.get_reverse_complement(dna_sequence)\n",
        "            return dna_sequence.upper() == reverse_comp.upper()\n",
        "        except ValueError:\n",
        "            return False\n",
        "\n",
        "    def find_restriction_sites(self, dna_sequence: str,\n",
        "                             min_length: int = 4,\n",
        "                             max_length: int = 12) -> List[Tuple[int, int]]:\n",
        "        \"\"\"\n",
        "        Find all restriction sites (reverse palindromes) in a DNA sequence.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Input DNA sequence\n",
        "            min_length (int): Minimum palindrome length (default: 4)\n",
        "            max_length (int): Maximum palindrome length (default: 12)\n",
        "\n",
        "        Returns:\n",
        "            List[Tuple[int, int]]: List of (position, length) tuples (1-indexed positions)\n",
        "        \"\"\"\n",
        "        # Clean and validate sequence\n",
        "        sequence = dna_sequence.strip().upper()\n",
        "        if not all(base in 'ATGC' for base in sequence):\n",
        "            raise ValueError(\"DNA sequence contains invalid bases\")\n",
        "\n",
        "        restriction_sites = []\n",
        "        seq_length = len(sequence)\n",
        "\n",
        "        # Check all possible positions and lengths\n",
        "        for position in range(seq_length):\n",
        "            for length in range(min_length, min(max_length + 1, seq_length - position + 1)):\n",
        "                # Extract substring\n",
        "                substring = sequence[position:position + length]\n",
        "\n",
        "                # Check if it's a reverse palindrome\n",
        "                if self.is_reverse_palindrome(substring):\n",
        "                    # Add to results (convert to 1-indexed position)\n",
        "                    restriction_sites.append((position + 1, length))\n",
        "\n",
        "        return restriction_sites\n",
        "\n",
        "    def analyze_palindrome(self, dna_sequence: str) -> dict:\n",
        "        \"\"\"\n",
        "        Provide detailed analysis of a palindromic sequence.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): DNA sequence to analyze\n",
        "\n",
        "        Returns:\n",
        "            dict: Analysis results including complement, reverse complement, etc.\n",
        "        \"\"\"\n",
        "        sequence = dna_sequence.strip().upper()\n",
        "\n",
        "        analysis = {\n",
        "            'sequence': sequence,\n",
        "            'length': len(sequence),\n",
        "            'complement': self.get_complement(sequence),\n",
        "            'reverse': sequence[::-1],\n",
        "            'reverse_complement': self.get_reverse_complement(sequence),\n",
        "            'is_reverse_palindrome': self.is_reverse_palindrome(sequence)\n",
        "        }\n",
        "\n",
        "        return analysis\n",
        "\n",
        "\n",
        "def parse_fasta_file(file_path: str) -> Tuple[str, str]:\n",
        "    \"\"\"\n",
        "    Parse FASTA file to extract header and DNA sequence.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to FASTA file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[str, str]: (header, dna_sequence)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        lines = content.split('\\n')\n",
        "        if not lines or not lines[0].startswith('>'):\n",
        "            raise ValueError(\"Invalid FASTA format: missing header\")\n",
        "\n",
        "        header = lines[0][1:]  # Remove '>' character\n",
        "        sequence = ''.join(line.strip() for line in lines[1:] if not line.startswith('>'))\n",
        "\n",
        "        if not sequence:\n",
        "            raise ValueError(\"Invalid FASTA format: no sequence found\")\n",
        "\n",
        "        # Validate sequence length constraint\n",
        "        if len(sequence) > 1000:\n",
        "            raise ValueError(f\"Sequence length {len(sequence)} exceeds maximum of 1000 bp\")\n",
        "\n",
        "        return header, sequence\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, restriction_sites: List[Tuple[int, int]]) -> None:\n",
        "    \"\"\"\n",
        "    Write restriction sites to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        restriction_sites (List[Tuple[int, int]]): List of (position, length) pairs\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            for position, length in restriction_sites:\n",
        "                file.write(f\"{position} {length}\\n\")\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_restriction_sites_problem(input_file_path: str) -> List[Tuple[int, int]]:\n",
        "    \"\"\"\n",
        "    Solve the Restriction Sites problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input FASTA file\n",
        "\n",
        "    Returns:\n",
        "        List[Tuple[int, int]]: List of (position, length) for each restriction site\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse FASTA file\n",
        "        header, dna_sequence = parse_fasta_file(input_file_path)\n",
        "\n",
        "        # Initialize finder\n",
        "        finder = RestrictionSiteFinder()\n",
        "\n",
        "        # Find all restriction sites\n",
        "        restriction_sites = finder.find_restriction_sites(dna_sequence)\n",
        "\n",
        "        return restriction_sites\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Restriction Sites problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"/content/rosalind_revp.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_revp.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Restriction Sites Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        restriction_sites = solve_restriction_sites_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nFound {len(restriction_sites)} restriction sites:\")\n",
        "        print(\"Position Length\")\n",
        "        print(\"-\" * 15)\n",
        "        for position, length in restriction_sites:\n",
        "            print(f\"{position:>8} {length:>6}\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, restriction_sites)\n",
        "        print(f\"\\nResults written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample data from problem\n",
        "    sample_header = \"Rosalind_24\"\n",
        "    sample_dna = \"TCAATGCATGCGGGTCTATATGCAT\"\n",
        "\n",
        "    print(f\"Header: {sample_header}\")\n",
        "    print(f\"DNA Sequence: {sample_dna}\")\n",
        "    print(f\"Length: {len(sample_dna)} bp\")\n",
        "\n",
        "    # Find restriction sites\n",
        "    finder = RestrictionSiteFinder()\n",
        "    sites = finder.find_restriction_sites(sample_dna)\n",
        "\n",
        "    print(f\"\\nFound restriction sites:\")\n",
        "    print(\"Position Length Sequence\")\n",
        "    print(\"-\" * 25)\n",
        "\n",
        "    for position, length in sites:\n",
        "        # Extract the palindromic sequence (convert to 0-indexed for slicing)\n",
        "        sequence = sample_dna[position-1:position-1+length]\n",
        "        print(f\"{position:>8} {length:>6} {sequence}\")\n",
        "\n",
        "    # Verify a few examples\n",
        "    print(f\"\\n=== Verification ===\")\n",
        "    verification_cases = [\n",
        "        (4, 6),  # Position 4, length 6\n",
        "        (5, 4),  # Position 5, length 4\n",
        "    ]\n",
        "\n",
        "    for pos, length in verification_cases:\n",
        "        sequence = sample_dna[pos-1:pos-1+length]\n",
        "        analysis = finder.analyze_palindrome(sequence)\n",
        "\n",
        "        print(f\"\\nPosition {pos}, Length {length}:\")\n",
        "        print(f\"  Sequence: {analysis['sequence']}\")\n",
        "        print(f\"  Reverse complement: {analysis['reverse_complement']}\")\n",
        "        print(f\"  Is palindrome: {analysis['is_reverse_palindrome']}\")\n",
        "\n",
        "\n",
        "def test_individual_cases():\n",
        "    \"\"\"\n",
        "    Test individual palindrome cases for educational purposes.\n",
        "    \"\"\"\n",
        "    print(\"=== Testing Individual Cases ===\")\n",
        "\n",
        "    finder = RestrictionSiteFinder()\n",
        "\n",
        "    test_sequences = [\n",
        "        \"GCATGC\",    # Classic example from problem description\n",
        "        \"ATAT\",      # Simple palindrome\n",
        "        \"GAATTC\",    # EcoRI recognition site\n",
        "        \"GGATCC\",    # BamHI recognition site\n",
        "        \"ATCGAT\",    # Another common palindrome\n",
        "        \"ABCD\",      # Invalid sequence (should fail)\n",
        "    ]\n",
        "\n",
        "    for seq in test_sequences:\n",
        "        try:\n",
        "            analysis = finder.analyze_palindrome(seq)\n",
        "            print(f\"\\nSequence: {seq}\")\n",
        "            print(f\"  Complement: {analysis['complement']}\")\n",
        "            print(f\"  Reverse complement: {analysis['reverse_complement']}\")\n",
        "            print(f\"  Is reverse palindrome: {analysis['is_reverse_palindrome']}\")\n",
        "        except Exception as e:\n",
        "            print(f\"\\nSequence: {seq} -> Error: {e}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo with sample data\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test individual cases\n",
        "    test_individual_cases()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zDJDMnGAdYbA",
        "outputId": "7a1c5655-e762-420f-8de9-f2928dcf0211"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Header: Rosalind_24\n",
            "DNA Sequence: TCAATGCATGCGGGTCTATATGCAT\n",
            "Length: 25 bp\n",
            "\n",
            "Found restriction sites:\n",
            "Position Length Sequence\n",
            "-------------------------\n",
            "       4      6 ATGCAT\n",
            "       5      4 TGCA\n",
            "       6      6 GCATGC\n",
            "       7      4 CATG\n",
            "      17      4 TATA\n",
            "      18      4 ATAT\n",
            "      20      6 ATGCAT\n",
            "      21      4 TGCA\n",
            "\n",
            "=== Verification ===\n",
            "\n",
            "Position 4, Length 6:\n",
            "  Sequence: ATGCAT\n",
            "  Reverse complement: ATGCAT\n",
            "  Is palindrome: True\n",
            "\n",
            "Position 5, Length 4:\n",
            "  Sequence: TGCA\n",
            "  Reverse complement: TGCA\n",
            "  Is palindrome: True\n",
            "\n",
            "============================================================\n",
            "=== Testing Individual Cases ===\n",
            "\n",
            "Sequence: GCATGC\n",
            "  Complement: CGTACG\n",
            "  Reverse complement: GCATGC\n",
            "  Is reverse palindrome: True\n",
            "\n",
            "Sequence: ATAT\n",
            "  Complement: TATA\n",
            "  Reverse complement: ATAT\n",
            "  Is reverse palindrome: True\n",
            "\n",
            "Sequence: GAATTC\n",
            "  Complement: CTTAAG\n",
            "  Reverse complement: GAATTC\n",
            "  Is reverse palindrome: True\n",
            "\n",
            "Sequence: GGATCC\n",
            "  Complement: CCTAGG\n",
            "  Reverse complement: GGATCC\n",
            "  Is reverse palindrome: True\n",
            "\n",
            "Sequence: ATCGAT\n",
            "  Complement: TAGCTA\n",
            "  Reverse complement: ATCGAT\n",
            "  Is reverse palindrome: True\n",
            "\n",
            "Sequence: ABCD -> Error: Invalid DNA base found: 'B'\n",
            "\n",
            "============================================================\n",
            "Solving Restriction Sites Problem...\n",
            "\n",
            "Found 85 restriction sites:\n",
            "Position Length\n",
            "---------------\n",
            "       1      4\n",
            "      19      6\n",
            "      20      4\n",
            "      30      6\n",
            "      31      4\n",
            "      42      4\n",
            "      62      4\n",
            "      72      6\n",
            "      73      4\n",
            "      93      4\n",
            "      97      4\n",
            "     121      4\n",
            "     137      4\n",
            "     138      4\n",
            "     167      6\n",
            "     168      4\n",
            "     196      6\n",
            "     197      4\n",
            "     203      4\n",
            "     227      4\n",
            "     231      4\n",
            "     275      4\n",
            "     295      4\n",
            "     299      4\n",
            "     301      4\n",
            "     321      4\n",
            "     335      4\n",
            "     340      4\n",
            "     342      6\n",
            "     343      4\n",
            "     361      4\n",
            "     366      4\n",
            "     379      4\n",
            "     384      8\n",
            "     385      6\n",
            "     386      4\n",
            "     409     12\n",
            "     410     10\n",
            "     411      8\n",
            "     412      6\n",
            "     413      4\n",
            "     420      6\n",
            "     421      4\n",
            "     462      4\n",
            "     492      4\n",
            "     499      4\n",
            "     518      4\n",
            "     526      4\n",
            "     548      4\n",
            "     560      4\n",
            "     574      4\n",
            "     579      4\n",
            "     584      4\n",
            "     584      6\n",
            "     585      4\n",
            "     585      6\n",
            "     586      4\n",
            "     587      4\n",
            "     596      4\n",
            "     607      4\n",
            "     610      4\n",
            "     617      4\n",
            "     629      4\n",
            "     648      4\n",
            "     661      4\n",
            "     693      4\n",
            "     695      6\n",
            "     696      4\n",
            "     706      4\n",
            "     707      4\n",
            "     719     10\n",
            "     720      8\n",
            "     721      6\n",
            "     722      4\n",
            "     737      4\n",
            "     747      6\n",
            "     748      4\n",
            "     750      4\n",
            "     752      6\n",
            "     753      4\n",
            "     785      6\n",
            "     786      4\n",
            "     796      6\n",
            "     797      4\n",
            "     825      4\n",
            "\n",
            "Results written to: output_revp.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Rosalind Calculating Protein Mass Problem Solution"
      ],
      "metadata": {
        "id": "ryVWt8QaZp4C"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Calculating Protein Mass Problem Solution\n",
        "\n",
        "This module calculates the total monoisotopic mass of a protein string by summing\n",
        "the monoisotopic masses of its constituent amino acids.\n",
        "\n",
        "In mass spectrometry, the monoisotopic mass uses the principal (most abundant)\n",
        "isotope of each atom. For peptides excised from the middle of proteins, we sum\n",
        "the residue masses without adding water molecule mass.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import Dict, Optional\n",
        "import re\n",
        "\n",
        "\n",
        "class ProteinMassCalculator:\n",
        "    \"\"\"\n",
        "    A class to handle protein mass calculations using monoisotopic masses.\n",
        "    \"\"\"\n",
        "\n",
        "    # Monoisotopic mass table for amino acids (in Daltons)\n",
        "    # Source: Standard biochemistry references\n",
        "    MONOISOTOPIC_MASSES = {\n",
        "        'A': 71.03711,   # Alanine\n",
        "        'C': 103.00919,  # Cysteine\n",
        "        'D': 115.02694,  # Aspartic acid\n",
        "        'E': 129.04259,  # Glutamic acid\n",
        "        'F': 147.06841,  # Phenylalanine\n",
        "        'G': 57.02146,   # Glycine\n",
        "        'H': 137.05891,  # Histidine\n",
        "        'I': 113.08406,  # Isoleucine\n",
        "        'K': 128.09496,  # Lysine\n",
        "        'L': 113.08406,  # Leucine\n",
        "        'M': 131.04049,  # Methionine\n",
        "        'N': 114.04293,  # Asparagine\n",
        "        'P': 97.05276,   # Proline\n",
        "        'Q': 128.05858,  # Glutamine\n",
        "        'R': 156.10111,  # Arginine\n",
        "        'S': 87.03203,   # Serine\n",
        "        'T': 101.04768,  # Threonine\n",
        "        'V': 99.06841,   # Valine\n",
        "        'W': 186.07931,  # Tryptophan\n",
        "        'Y': 163.06333   # Tyrosine\n",
        "    }\n",
        "\n",
        "    # Water molecule monoisotopic mass (for reference, not used in this problem)\n",
        "    WATER_MASS = 18.01056\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the protein mass calculator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def validate_protein_string(protein: str) -> bool:\n",
        "        \"\"\"\n",
        "        Validate that the protein string contains only valid amino acid codes.\n",
        "\n",
        "        Args:\n",
        "            protein (str): Protein sequence string\n",
        "\n",
        "        Returns:\n",
        "            bool: True if valid, False otherwise\n",
        "        \"\"\"\n",
        "        if not protein:\n",
        "            return False\n",
        "\n",
        "        # Check if all characters are valid amino acids\n",
        "        valid_chars = set(ProteinMassCalculator.MONOISOTOPIC_MASSES.keys())\n",
        "        protein_chars = set(protein.upper())\n",
        "\n",
        "        return protein_chars.issubset(valid_chars)\n",
        "\n",
        "    @staticmethod\n",
        "    def get_amino_acid_mass(amino_acid: str) -> Optional[float]:\n",
        "        \"\"\"\n",
        "        Get the monoisotopic mass of a single amino acid.\n",
        "\n",
        "        Args:\n",
        "            amino_acid (str): Single amino acid code\n",
        "\n",
        "        Returns:\n",
        "            float or None: Monoisotopic mass in Daltons, or None if invalid\n",
        "        \"\"\"\n",
        "        return ProteinMassCalculator.MONOISOTOPIC_MASSES.get(amino_acid.upper())\n",
        "\n",
        "    def calculate_protein_mass(self, protein_string: str) -> float:\n",
        "        \"\"\"\n",
        "        Calculate the total monoisotopic mass of a protein string.\n",
        "\n",
        "        Args:\n",
        "            protein_string (str): Protein sequence\n",
        "\n",
        "        Returns:\n",
        "            float: Total monoisotopic mass in Daltons\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If protein string contains invalid amino acids\n",
        "        \"\"\"\n",
        "        # Clean the protein string\n",
        "        protein = protein_string.strip().upper()\n",
        "\n",
        "        # Validate the protein string\n",
        "        if not self.validate_protein_string(protein):\n",
        "            invalid_chars = set(protein) - set(self.MONOISOTOPIC_MASSES.keys())\n",
        "            raise ValueError(f\"Invalid amino acid codes found: {invalid_chars}\")\n",
        "\n",
        "        # Calculate total mass\n",
        "        total_mass = 0.0\n",
        "\n",
        "        for amino_acid in protein:\n",
        "            mass = self.get_amino_acid_mass(amino_acid)\n",
        "            if mass is not None:\n",
        "                total_mass += mass\n",
        "            else:\n",
        "                raise ValueError(f\"Unknown amino acid: {amino_acid}\")\n",
        "\n",
        "        return total_mass\n",
        "\n",
        "    def get_mass_breakdown(self, protein_string: str) -> Dict[str, float]:\n",
        "        \"\"\"\n",
        "        Get a detailed breakdown of mass contributions by amino acid.\n",
        "\n",
        "        Args:\n",
        "            protein_string (str): Protein sequence\n",
        "\n",
        "        Returns:\n",
        "            Dict[str, float]: Dictionary with amino acid counts and masses\n",
        "        \"\"\"\n",
        "        protein = protein_string.strip().upper()\n",
        "\n",
        "        if not self.validate_protein_string(protein):\n",
        "            raise ValueError(\"Invalid protein string\")\n",
        "\n",
        "        breakdown = {}\n",
        "        amino_acid_counts = {}\n",
        "\n",
        "        # Count amino acids\n",
        "        for aa in protein:\n",
        "            amino_acid_counts[aa] = amino_acid_counts.get(aa, 0) + 1\n",
        "\n",
        "        # Calculate mass contributions\n",
        "        for aa, count in amino_acid_counts.items():\n",
        "            mass_per_aa = self.get_amino_acid_mass(aa)\n",
        "            breakdown[aa] = {\n",
        "                'count': count,\n",
        "                'mass_per_residue': mass_per_aa,\n",
        "                'total_mass': mass_per_aa * count\n",
        "            }\n",
        "\n",
        "        return breakdown\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> str:\n",
        "    \"\"\"\n",
        "    Parse input file to extract the protein string.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        str: Protein string\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        # Remove any whitespace and newlines\n",
        "        protein_string = re.sub(r'\\s+', '', content)\n",
        "\n",
        "        if not protein_string:\n",
        "            raise ValueError(\"Input file is empty or contains no valid protein sequence\")\n",
        "\n",
        "        # Validate length constraint\n",
        "        if len(protein_string) > 1000:\n",
        "            raise ValueError(f\"Protein string length {len(protein_string)} exceeds maximum of 1000\")\n",
        "\n",
        "        return protein_string\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, mass: float) -> None:\n",
        "    \"\"\"\n",
        "    Write result to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        mass (float): Calculated protein mass\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            # Format to 3 decimal places to match expected output format\n",
        "            file.write(f\"{mass:.3f}\\n\")\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_protein_mass_problem(input_file_path: str) -> float:\n",
        "    \"\"\"\n",
        "    Solve the Protein Mass problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file containing protein string\n",
        "\n",
        "    Returns:\n",
        "        float: Total protein mass in Daltons\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        protein_string = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize calculator\n",
        "        calculator = ProteinMassCalculator()\n",
        "\n",
        "        # Calculate mass\n",
        "        total_mass = calculator.calculate_protein_mass(protein_string)\n",
        "\n",
        "        return total_mass\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Protein Mass problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_prtm.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_prtm.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Protein Mass Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        total_mass = solve_protein_mass_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResult:\")\n",
        "        print(f\"Total protein mass: {total_mass:.3f} Da\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, total_mass)\n",
        "        print(f\"\\nResult written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input\n",
        "    sample_protein = \"SKADYEK\"\n",
        "\n",
        "    print(f\"Input protein string: {sample_protein}\")\n",
        "\n",
        "    calculator = ProteinMassCalculator()\n",
        "\n",
        "    # Calculate mass\n",
        "    total_mass = calculator.calculate_protein_mass(sample_protein)\n",
        "    print(f\"Total mass: {total_mass:.3f} Da\")\n",
        "\n",
        "    # Expected output\n",
        "    print(f\"Expected output: 821.392 Da\")\n",
        "\n",
        "    # Show detailed breakdown\n",
        "    print(f\"\\n=== Detailed Breakdown ===\")\n",
        "    breakdown = calculator.get_mass_breakdown(sample_protein)\n",
        "\n",
        "    running_total = 0.0\n",
        "    for aa in sample_protein:\n",
        "        mass = calculator.get_amino_acid_mass(aa)\n",
        "        running_total += mass\n",
        "        print(f\"{aa}: {mass:.5f} Da (running total: {running_total:.5f})\")\n",
        "\n",
        "    print(f\"\\nFinal total: {total_mass:.5f} Da\")\n",
        "\n",
        "    # Show amino acid frequency analysis\n",
        "    print(f\"\\n=== Amino Acid Composition ===\")\n",
        "    for aa, info in sorted(breakdown.items()):\n",
        "        print(f\"{aa}: {info['count']} × {info['mass_per_residue']:.5f} = {info['total_mass']:.5f} Da\")\n",
        "\n",
        "\n",
        "def show_mass_table():\n",
        "    \"\"\"\n",
        "    Display the complete monoisotopic mass table.\n",
        "    \"\"\"\n",
        "    print(\"=== Monoisotopic Mass Table ===\")\n",
        "    print(\"Amino Acid | Code | Mass (Da)\")\n",
        "    print(\"-\" * 35)\n",
        "\n",
        "    # Full names for amino acids\n",
        "    aa_names = {\n",
        "        'A': 'Alanine', 'C': 'Cysteine', 'D': 'Aspartic acid',\n",
        "        'E': 'Glutamic acid', 'F': 'Phenylalanine', 'G': 'Glycine',\n",
        "        'H': 'Histidine', 'I': 'Isoleucine', 'K': 'Lysine',\n",
        "        'L': 'Leucine', 'M': 'Methionine', 'N': 'Asparagine',\n",
        "        'P': 'Proline', 'Q': 'Glutamine', 'R': 'Arginine',\n",
        "        'S': 'Serine', 'T': 'Threonine', 'V': 'Valine',\n",
        "        'W': 'Tryptophan', 'Y': 'Tyrosine'\n",
        "    }\n",
        "\n",
        "    for code, mass in sorted(ProteinMassCalculator.MONOISOTOPIC_MASSES.items()):\n",
        "        name = aa_names.get(code, 'Unknown')\n",
        "        print(f\"{name:<12} | {code:>4} | {mass:>9.5f}\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Show mass table\n",
        "    show_mass_table()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Run demo with sample\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test with additional cases\n",
        "    print(\"=== Additional Test Cases ===\")\n",
        "    calculator = ProteinMassCalculator()\n",
        "\n",
        "    test_cases = [\n",
        "        \"A\",           # Single amino acid\n",
        "        \"AA\",          # Repeated amino acid\n",
        "        \"GAVL\",        # Small peptide\n",
        "        \"MSKADYEK\",    # Extended sample\n",
        "    ]\n",
        "\n",
        "    for test_protein in test_cases:\n",
        "        try:\n",
        "            mass = calculator.calculate_protein_mass(test_protein)\n",
        "            print(f\"{test_protein}: {mass:.3f} Da\")\n",
        "        except Exception as e:\n",
        "            print(f\"{test_protein}: Error - {e}\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EOa2KhzRZsLo",
        "outputId": "98d590e3-e492-4331-e08e-3f972703e472"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Monoisotopic Mass Table ===\n",
            "Amino Acid | Code | Mass (Da)\n",
            "-----------------------------------\n",
            "Alanine      |    A |  71.03711\n",
            "Cysteine     |    C | 103.00919\n",
            "Aspartic acid |    D | 115.02694\n",
            "Glutamic acid |    E | 129.04259\n",
            "Phenylalanine |    F | 147.06841\n",
            "Glycine      |    G |  57.02146\n",
            "Histidine    |    H | 137.05891\n",
            "Isoleucine   |    I | 113.08406\n",
            "Lysine       |    K | 128.09496\n",
            "Leucine      |    L | 113.08406\n",
            "Methionine   |    M | 131.04049\n",
            "Asparagine   |    N | 114.04293\n",
            "Proline      |    P |  97.05276\n",
            "Glutamine    |    Q | 128.05858\n",
            "Arginine     |    R | 156.10111\n",
            "Serine       |    S |  87.03203\n",
            "Threonine    |    T | 101.04768\n",
            "Valine       |    V |  99.06841\n",
            "Tryptophan   |    W | 186.07931\n",
            "Tyrosine     |    Y | 163.06333\n",
            "\n",
            "============================================================\n",
            "=== Demo with Sample Data ===\n",
            "Input protein string: SKADYEK\n",
            "Total mass: 821.392 Da\n",
            "Expected output: 821.392 Da\n",
            "\n",
            "=== Detailed Breakdown ===\n",
            "S: 87.03203 Da (running total: 87.03203)\n",
            "K: 128.09496 Da (running total: 215.12699)\n",
            "A: 71.03711 Da (running total: 286.16410)\n",
            "D: 115.02694 Da (running total: 401.19104)\n",
            "Y: 163.06333 Da (running total: 564.25437)\n",
            "E: 129.04259 Da (running total: 693.29696)\n",
            "K: 128.09496 Da (running total: 821.39192)\n",
            "\n",
            "Final total: 821.39192 Da\n",
            "\n",
            "=== Amino Acid Composition ===\n",
            "A: 1 × 71.03711 = 71.03711 Da\n",
            "D: 1 × 115.02694 = 115.02694 Da\n",
            "E: 1 × 129.04259 = 129.04259 Da\n",
            "K: 2 × 128.09496 = 256.18992 Da\n",
            "S: 1 × 87.03203 = 87.03203 Da\n",
            "Y: 1 × 163.06333 = 163.06333 Da\n",
            "\n",
            "============================================================\n",
            "=== Additional Test Cases ===\n",
            "A: 71.037 Da\n",
            "AA: 142.074 Da\n",
            "GAVL: 340.211 Da\n",
            "MSKADYEK: 952.432 Da\n",
            "\n",
            "============================================================\n",
            "Solving Protein Mass Problem...\n",
            "\n",
            "Result:\n",
            "Total protein mass: 116728.427 Da\n",
            "\n",
            "Result written to: output_prtm.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Rosalind Open Reading Frames (ORF) Problem Solution"
      ],
      "metadata": {
        "id": "Urlk_ARiXbZI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Open Reading Frames (ORF) Problem Solution\n",
        "\n",
        "This module finds all distinct candidate protein strings that can be translated\n",
        "from Open Reading Frames (ORFs) in a DNA sequence. It considers all 6 reading\n",
        "frames: 3 from the original sequence and 3 from the reverse complement.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Set, Optional\n",
        "import re\n",
        "\n",
        "\n",
        "class DNATranslator:\n",
        "    \"\"\"\n",
        "    A class to handle DNA sequence translation and ORF finding operations.\n",
        "    \"\"\"\n",
        "\n",
        "    # Standard genetic code table (DNA codons to amino acids)\n",
        "    CODON_TABLE = {\n",
        "        'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
        "        'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
        "        'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
        "        'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
        "        'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
        "        'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
        "        'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
        "        'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
        "        'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
        "        'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
        "        'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
        "        'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
        "        'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
        "        'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
        "        'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
        "        'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
        "    }\n",
        "\n",
        "    # Complement mapping for DNA bases\n",
        "    COMPLEMENT_MAP = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
        "\n",
        "    START_CODON = 'ATG'\n",
        "    STOP_CODONS = {'TAA', 'TAG', 'TGA'}\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the DNA translator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def reverse_complement(dna_sequence: str) -> str:\n",
        "        \"\"\"\n",
        "        Generate the reverse complement of a DNA sequence.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Input DNA sequence\n",
        "\n",
        "        Returns:\n",
        "            str: Reverse complement of the input sequence\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If sequence contains invalid DNA bases\n",
        "        \"\"\"\n",
        "        try:\n",
        "            complement = ''.join(DNATranslator.COMPLEMENT_MAP[base]\n",
        "                               for base in dna_sequence.upper())\n",
        "            return complement[::-1]\n",
        "        except KeyError as e:\n",
        "            raise ValueError(f\"Invalid DNA base found: {e}\")\n",
        "\n",
        "    @staticmethod\n",
        "    def translate_codon(codon: str) -> Optional[str]:\n",
        "        \"\"\"\n",
        "        Translate a single codon to its corresponding amino acid.\n",
        "\n",
        "        Args:\n",
        "            codon (str): 3-letter DNA codon\n",
        "\n",
        "        Returns:\n",
        "            str or None: Corresponding amino acid or None if invalid codon\n",
        "        \"\"\"\n",
        "        if len(codon) != 3:\n",
        "            return None\n",
        "        return DNATranslator.CODON_TABLE.get(codon.upper())\n",
        "\n",
        "    def find_orfs_in_frame(self, sequence: str, frame: int = 0) -> List[str]:\n",
        "        \"\"\"\n",
        "        Find all ORFs in a specific reading frame.\n",
        "\n",
        "        Args:\n",
        "            sequence (str): DNA sequence\n",
        "            frame (int): Reading frame offset (0, 1, or 2)\n",
        "\n",
        "        Returns:\n",
        "            List[str]: List of protein sequences found in this frame\n",
        "        \"\"\"\n",
        "        proteins = []\n",
        "        sequence = sequence.upper()\n",
        "\n",
        "        # Start from the specified frame\n",
        "        for i in range(frame, len(sequence) - 2, 3):\n",
        "            codon = sequence[i:i+3]\n",
        "\n",
        "            # Check if this is a start codon\n",
        "            if codon == self.START_CODON:\n",
        "                protein = 'M'  # Start with Methionine\n",
        "\n",
        "                # Continue translating until stop codon or end of sequence\n",
        "                for j in range(i + 3, len(sequence) - 2, 3):\n",
        "                    next_codon = sequence[j:j+3]\n",
        "                    amino_acid = self.translate_codon(next_codon)\n",
        "\n",
        "                    if amino_acid is None:  # Invalid codon\n",
        "                        break\n",
        "                    elif amino_acid == '*':  # Stop codon\n",
        "                        proteins.append(protein)\n",
        "                        break\n",
        "                    else:\n",
        "                        protein += amino_acid\n",
        "\n",
        "        return proteins\n",
        "\n",
        "    def find_all_orfs(self, dna_sequence: str) -> Set[str]:\n",
        "        \"\"\"\n",
        "        Find all possible ORFs in a DNA sequence across all 6 reading frames.\n",
        "\n",
        "        Args:\n",
        "            dna_sequence (str): Input DNA sequence\n",
        "\n",
        "        Returns:\n",
        "            Set[str]: Set of unique protein sequences\n",
        "        \"\"\"\n",
        "        all_proteins = set()\n",
        "\n",
        "        # Clean the sequence (remove whitespace and convert to uppercase)\n",
        "        clean_sequence = re.sub(r'\\s+', '', dna_sequence.upper())\n",
        "\n",
        "        # Get reverse complement\n",
        "        rev_comp = self.reverse_complement(clean_sequence)\n",
        "\n",
        "        # Find ORFs in all 6 reading frames\n",
        "        sequences = [clean_sequence, rev_comp]\n",
        "\n",
        "        for seq in sequences:\n",
        "            for frame in range(3):\n",
        "                proteins = self.find_orfs_in_frame(seq, frame)\n",
        "                all_proteins.update(proteins)\n",
        "\n",
        "        return all_proteins\n",
        "\n",
        "\n",
        "def parse_fasta(file_content: str) -> List[tuple]:\n",
        "    \"\"\"\n",
        "    Parse FASTA format content and return sequences.\n",
        "\n",
        "    Args:\n",
        "        file_content (str): Content of FASTA file\n",
        "\n",
        "    Returns:\n",
        "        List[tuple]: List of (header, sequence) tuples\n",
        "    \"\"\"\n",
        "    sequences = []\n",
        "    current_header = \"\"\n",
        "    current_sequence = \"\"\n",
        "\n",
        "    for line in file_content.strip().split('\\n'):\n",
        "        line = line.strip()\n",
        "        if line.startswith('>'):\n",
        "            if current_header and current_sequence:\n",
        "                sequences.append((current_header, current_sequence))\n",
        "            current_header = line[1:]  # Remove '>' character\n",
        "            current_sequence = \"\"\n",
        "        else:\n",
        "            current_sequence += line\n",
        "\n",
        "    # Add the last sequence\n",
        "    if current_header and current_sequence:\n",
        "        sequences.append((current_header, current_sequence))\n",
        "\n",
        "    return sequences\n",
        "\n",
        "\n",
        "def solve_orf_problem(input_file_path: str) -> List[str]:\n",
        "    \"\"\"\n",
        "    Solve the Open Reading Frame problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file containing DNA sequence in FASTA format\n",
        "\n",
        "    Returns:\n",
        "        List[str]: List of unique protein sequences\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Read input file\n",
        "        with open(input_file_path, 'r') as file:\n",
        "            file_content = file.read()\n",
        "\n",
        "        # Parse FASTA format\n",
        "        sequences = parse_fasta(file_content)\n",
        "\n",
        "        if not sequences:\n",
        "            raise ValueError(\"No valid FASTA sequences found in input file\")\n",
        "\n",
        "        # Initialize translator\n",
        "        translator = DNATranslator()\n",
        "\n",
        "        # Process all sequences (usually just one for this problem)\n",
        "        all_proteins = set()\n",
        "        for header, sequence in sequences:\n",
        "            proteins = translator.find_all_orfs(sequence)\n",
        "            all_proteins.update(proteins)\n",
        "\n",
        "        return sorted(list(all_proteins))\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{input_file_path}' not found\")\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error processing file: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the ORF problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # For Google Colab usage\n",
        "    input_file = \"rosalind_orf.txt\"  # Change this to your input file name\n",
        "\n",
        "    try:\n",
        "        # Solve the problem\n",
        "        protein_sequences = solve_orf_problem(input_file)\n",
        "\n",
        "        # Print results\n",
        "        print(\"Found protein sequences:\")\n",
        "        for protein in protein_sequences:\n",
        "            print(protein)\n",
        "\n",
        "        # Optionally write to output file\n",
        "        with open(\"/content/rosalind_orf.txt\", \"w\") as f:\n",
        "            for protein in protein_sequences:\n",
        "                f.write(protein + \"\\n\")\n",
        "\n",
        "        print(f\"\\nTotal unique protein sequences found: {len(protein_sequences)}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Error: {e}\")\n",
        "        print(\"Make sure your input file is in the correct FASTA format\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Test with sample data\n",
        "    sample_data = \"\"\">Rosalind_99\n",
        "AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG\"\"\"\n",
        "\n",
        "    # Test the parsing\n",
        "    sequences = parse_fasta(sample_data)\n",
        "    print(\"Parsed sequences:\")\n",
        "    for header, seq in sequences:\n",
        "        print(f\"Header: {header}\")\n",
        "        print(f\"Sequence length: {len(seq)}\")\n",
        "\n",
        "    # Test the translator\n",
        "    translator = DNATranslator()\n",
        "    if sequences:\n",
        "        proteins = translator.find_all_orfs(sequences[0][1])\n",
        "        print(f\"\\nFound proteins: {sorted(proteins)}\")\n",
        "\n",
        "    # Run main function\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ps4TOdPdKUHf",
        "outputId": "d27ac710-76e6-4032-dbec-e3c5e7ce442b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Parsed sequences:\n",
            "Header: Rosalind_99\n",
            "Sequence length: 96\n",
            "\n",
            "Found proteins: ['M', 'MGMTPRLGLESLLE', 'MLLGSFRLIPKETLIQVAGSSPCNLS', 'MTPRLGLESLLE']\n",
            "Found protein sequences:\n",
            "M\n",
            "MAGNSYTESRRSSPIPKLK\n",
            "MAGPS\n",
            "MCPARSDISEI\n",
            "MDLRKCVFSCSGHW\n",
            "MGMAGNSYTESRRSSPIPKLK\n",
            "MGRPC\n",
            "MHCSQFLGLRTQGDLVHSNCVSRISCSK\n",
            "MHFCFTGVCVTSGCNGDNSSDDIVV\n",
            "MISG\n",
            "MISWCRDTSLI\n",
            "MKLSCLRGLLLMDLRKCVFSCSGHW\n",
            "MKYLYTTISSDELSPLHPEVTHTPVKQKCIVVSS\n",
            "MLEVARLLERC\n",
            "MLISRAR\n",
            "MPIRY\n",
            "MPRTAENTLTQVH\n",
            "MQRLTV\n",
            "MRDLWV\n",
            "MRKCENRQLPHKGFAPITRCS\n",
            "MSERAGHIKNKIYRVI\n",
            "MSLPSPEVQ\n",
            "MSVRCHGAGIGSHNTRPDDFLGLDFTYVRASWAH\n",
            "MTSLG\n",
            "\n",
            "Total unique protein sequences found: 24\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Solution for Enumerating Gene Orders (Permutations) Problem"
      ],
      "metadata": {
        "id": "1Jx1Du0oVcxh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Enumerating Gene Orders Problem Solution\n",
        "\n",
        "This module generates all possible permutations of a given length n, representing\n",
        "different arrangements of synteny blocks in genomic rearrangements.\n",
        "\n",
        "The problem asks for:\n",
        "1. Total number of permutations of length n\n",
        "2. All permutations listed in any order\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from itertools import permutations\n",
        "from typing import List, Tuple\n",
        "import math\n",
        "\n",
        "\n",
        "class PermutationGenerator:\n",
        "    \"\"\"\n",
        "    A class to handle permutation generation and counting operations.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the permutation generator.\"\"\"\n",
        "        pass\n",
        "\n",
        "    @staticmethod\n",
        "    def calculate_factorial(n: int) -> int:\n",
        "        \"\"\"\n",
        "        Calculate factorial of n (n!).\n",
        "\n",
        "        Args:\n",
        "            n (int): Non-negative integer\n",
        "\n",
        "        Returns:\n",
        "            int: Factorial of n\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If n is negative\n",
        "        \"\"\"\n",
        "        if n < 0:\n",
        "            raise ValueError(\"Factorial is not defined for negative numbers\")\n",
        "        return math.factorial(n)\n",
        "\n",
        "    @staticmethod\n",
        "    def count_permutations(n: int) -> int:\n",
        "        \"\"\"\n",
        "        Count the total number of permutations of length n.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            int: Total number of permutations (n!)\n",
        "        \"\"\"\n",
        "        return PermutationGenerator.calculate_factorial(n)\n",
        "\n",
        "    @staticmethod\n",
        "    def generate_all_permutations(n: int) -> List[Tuple[int, ...]]:\n",
        "        \"\"\"\n",
        "        Generate all permutations of integers from 1 to n.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            List[Tuple[int, ...]]: List of all permutations\n",
        "\n",
        "        Raises:\n",
        "            ValueError: If n is not a positive integer or exceeds reasonable limits\n",
        "        \"\"\"\n",
        "        if not isinstance(n, int) or n <= 0:\n",
        "            raise ValueError(\"n must be a positive integer\")\n",
        "\n",
        "        if n > 10:  # Safety check to prevent memory issues\n",
        "            raise ValueError(\"n is too large. Maximum supported value is 10\")\n",
        "\n",
        "        # Generate all permutations of {1, 2, ..., n}\n",
        "        numbers = list(range(1, n + 1))\n",
        "        all_perms = list(permutations(numbers))\n",
        "\n",
        "        return all_perms\n",
        "\n",
        "    def solve_permutation_problem(self, n: int) -> Tuple[int, List[str]]:\n",
        "        \"\"\"\n",
        "        Solve the complete permutation problem.\n",
        "\n",
        "        Args:\n",
        "            n (int): Length of permutation\n",
        "\n",
        "        Returns:\n",
        "            Tuple[int, List[str]]: (count_of_permutations, formatted_permutations)\n",
        "        \"\"\"\n",
        "        # Validate input\n",
        "        if not isinstance(n, int) or n <= 0:\n",
        "            raise ValueError(\"Input must be a positive integer\")\n",
        "\n",
        "        if n > 7:  # Based on problem constraint\n",
        "            raise ValueError(\"n must be ≤ 7 according to problem constraints\")\n",
        "\n",
        "        # Generate all permutations\n",
        "        all_perms = self.generate_all_permutations(n)\n",
        "\n",
        "        # Count permutations\n",
        "        count = len(all_perms)\n",
        "\n",
        "        # Format permutations as strings\n",
        "        formatted_perms = [' '.join(map(str, perm)) for perm in all_perms]\n",
        "\n",
        "        return count, formatted_perms\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> int:\n",
        "    \"\"\"\n",
        "    Parse input file to extract the integer n.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        int: The integer n from the file\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file content is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            content = file.read().strip()\n",
        "\n",
        "        # Parse the integer\n",
        "        try:\n",
        "            n = int(content)\n",
        "            return n\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Invalid input: '{content}' is not a valid integer\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, count: int, permutations: List[str]) -> None:\n",
        "    \"\"\"\n",
        "    Write results to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        count (int): Number of permutations\n",
        "        permutations (List[str]): List of formatted permutations\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            # Write count first\n",
        "            file.write(f\"{count}\\n\")\n",
        "\n",
        "            # Write all permutations\n",
        "            for perm in permutations:\n",
        "                file.write(f\"{perm}\\n\")\n",
        "\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_gene_orders_problem(input_file_path: str) -> Tuple[int, List[str]]:\n",
        "    \"\"\"\n",
        "    Solve the Gene Orders problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file containing integer n\n",
        "\n",
        "    Returns:\n",
        "        Tuple[int, List[str]]: (count, list_of_permutations)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        n = parse_input_file(input_file_path)\n",
        "\n",
        "        # Validate constraints\n",
        "        if n > 7:\n",
        "            raise ValueError(\"n must be ≤ 7 according to problem constraints\")\n",
        "\n",
        "        # Initialize solver\n",
        "        generator = PermutationGenerator()\n",
        "\n",
        "        # Solve the problem\n",
        "        count, permutations = generator.solve_permutation_problem(n)\n",
        "\n",
        "        return count, permutations\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Gene Orders problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"/content/rosalind_perm.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_perm.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Gene Orders (Permutations) Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        count, permutations = solve_gene_orders_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResults:\")\n",
        "        print(f\"Total number of permutations: {count}\")\n",
        "        print(f\"All permutations:\")\n",
        "\n",
        "        for perm in permutations:\n",
        "            print(perm)\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, count, permutations)\n",
        "        print(f\"\\nResults written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input\n",
        "    n = 3\n",
        "\n",
        "    generator = PermutationGenerator()\n",
        "    count, permutations = generator.solve_permutation_problem(n)\n",
        "\n",
        "    print(f\"Input: {n}\")\n",
        "    print(f\"Output:\")\n",
        "    print(count)\n",
        "    for perm in permutations:\n",
        "        print(perm)\n",
        "\n",
        "    # Verify the math\n",
        "    expected_count = math.factorial(n)\n",
        "    print(f\"\\nVerification: {n}! = {expected_count} ✓\" if count == expected_count else \"✗\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo first\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*50)\n",
        "\n",
        "    # Test with different values\n",
        "    print(\"=== Testing with different values ===\")\n",
        "    generator = PermutationGenerator()\n",
        "\n",
        "    for test_n in [1, 2, 3, 4]:\n",
        "        try:\n",
        "            count, perms = generator.solve_permutation_problem(test_n)\n",
        "            print(f\"n={test_n}: {count} permutations\")\n",
        "        except Exception as e:\n",
        "            print(f\"n={test_n}: Error - {e}\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*50)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "Rjf2syGeKT5v",
        "outputId": "cad325a0-3869-48c7-f9be-478e61ddd6a0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input: 3\n",
            "Output:\n",
            "6\n",
            "1 2 3\n",
            "1 3 2\n",
            "2 1 3\n",
            "2 3 1\n",
            "3 1 2\n",
            "3 2 1\n",
            "\n",
            "Verification: 3! = 6 ✓\n",
            "\n",
            "==================================================\n",
            "=== Testing with different values ===\n",
            "n=1: 1 permutations\n",
            "n=2: 2 permutations\n",
            "n=3: 6 permutations\n",
            "n=4: 24 permutations\n",
            "\n",
            "==================================================\n",
            "Solving Gene Orders (Permutations) Problem...\n",
            "\n",
            "Results:\n",
            "Total number of permutations: 720\n",
            "All permutations:\n",
            "1 2 3 4 5 6\n",
            "1 2 3 4 6 5\n",
            "1 2 3 5 4 6\n",
            "1 2 3 5 6 4\n",
            "1 2 3 6 4 5\n",
            "1 2 3 6 5 4\n",
            "1 2 4 3 5 6\n",
            "1 2 4 3 6 5\n",
            "1 2 4 5 3 6\n",
            "1 2 4 5 6 3\n",
            "1 2 4 6 3 5\n",
            "1 2 4 6 5 3\n",
            "1 2 5 3 4 6\n",
            "1 2 5 3 6 4\n",
            "1 2 5 4 3 6\n",
            "1 2 5 4 6 3\n",
            "1 2 5 6 3 4\n",
            "1 2 5 6 4 3\n",
            "1 2 6 3 4 5\n",
            "1 2 6 3 5 4\n",
            "1 2 6 4 3 5\n",
            "1 2 6 4 5 3\n",
            "1 2 6 5 3 4\n",
            "1 2 6 5 4 3\n",
            "1 3 2 4 5 6\n",
            "1 3 2 4 6 5\n",
            "1 3 2 5 4 6\n",
            "1 3 2 5 6 4\n",
            "1 3 2 6 4 5\n",
            "1 3 2 6 5 4\n",
            "1 3 4 2 5 6\n",
            "1 3 4 2 6 5\n",
            "1 3 4 5 2 6\n",
            "1 3 4 5 6 2\n",
            "1 3 4 6 2 5\n",
            "1 3 4 6 5 2\n",
            "1 3 5 2 4 6\n",
            "1 3 5 2 6 4\n",
            "1 3 5 4 2 6\n",
            "1 3 5 4 6 2\n",
            "1 3 5 6 2 4\n",
            "1 3 5 6 4 2\n",
            "1 3 6 2 4 5\n",
            "1 3 6 2 5 4\n",
            "1 3 6 4 2 5\n",
            "1 3 6 4 5 2\n",
            "1 3 6 5 2 4\n",
            "1 3 6 5 4 2\n",
            "1 4 2 3 5 6\n",
            "1 4 2 3 6 5\n",
            "1 4 2 5 3 6\n",
            "1 4 2 5 6 3\n",
            "1 4 2 6 3 5\n",
            "1 4 2 6 5 3\n",
            "1 4 3 2 5 6\n",
            "1 4 3 2 6 5\n",
            "1 4 3 5 2 6\n",
            "1 4 3 5 6 2\n",
            "1 4 3 6 2 5\n",
            "1 4 3 6 5 2\n",
            "1 4 5 2 3 6\n",
            "1 4 5 2 6 3\n",
            "1 4 5 3 2 6\n",
            "1 4 5 3 6 2\n",
            "1 4 5 6 2 3\n",
            "1 4 5 6 3 2\n",
            "1 4 6 2 3 5\n",
            "1 4 6 2 5 3\n",
            "1 4 6 3 2 5\n",
            "1 4 6 3 5 2\n",
            "1 4 6 5 2 3\n",
            "1 4 6 5 3 2\n",
            "1 5 2 3 4 6\n",
            "1 5 2 3 6 4\n",
            "1 5 2 4 3 6\n",
            "1 5 2 4 6 3\n",
            "1 5 2 6 3 4\n",
            "1 5 2 6 4 3\n",
            "1 5 3 2 4 6\n",
            "1 5 3 2 6 4\n",
            "1 5 3 4 2 6\n",
            "1 5 3 4 6 2\n",
            "1 5 3 6 2 4\n",
            "1 5 3 6 4 2\n",
            "1 5 4 2 3 6\n",
            "1 5 4 2 6 3\n",
            "1 5 4 3 2 6\n",
            "1 5 4 3 6 2\n",
            "1 5 4 6 2 3\n",
            "1 5 4 6 3 2\n",
            "1 5 6 2 3 4\n",
            "1 5 6 2 4 3\n",
            "1 5 6 3 2 4\n",
            "1 5 6 3 4 2\n",
            "1 5 6 4 2 3\n",
            "1 5 6 4 3 2\n",
            "1 6 2 3 4 5\n",
            "1 6 2 3 5 4\n",
            "1 6 2 4 3 5\n",
            "1 6 2 4 5 3\n",
            "1 6 2 5 3 4\n",
            "1 6 2 5 4 3\n",
            "1 6 3 2 4 5\n",
            "1 6 3 2 5 4\n",
            "1 6 3 4 2 5\n",
            "1 6 3 4 5 2\n",
            "1 6 3 5 2 4\n",
            "1 6 3 5 4 2\n",
            "1 6 4 2 3 5\n",
            "1 6 4 2 5 3\n",
            "1 6 4 3 2 5\n",
            "1 6 4 3 5 2\n",
            "1 6 4 5 2 3\n",
            "1 6 4 5 3 2\n",
            "1 6 5 2 3 4\n",
            "1 6 5 2 4 3\n",
            "1 6 5 3 2 4\n",
            "1 6 5 3 4 2\n",
            "1 6 5 4 2 3\n",
            "1 6 5 4 3 2\n",
            "2 1 3 4 5 6\n",
            "2 1 3 4 6 5\n",
            "2 1 3 5 4 6\n",
            "2 1 3 5 6 4\n",
            "2 1 3 6 4 5\n",
            "2 1 3 6 5 4\n",
            "2 1 4 3 5 6\n",
            "2 1 4 3 6 5\n",
            "2 1 4 5 3 6\n",
            "2 1 4 5 6 3\n",
            "2 1 4 6 3 5\n",
            "2 1 4 6 5 3\n",
            "2 1 5 3 4 6\n",
            "2 1 5 3 6 4\n",
            "2 1 5 4 3 6\n",
            "2 1 5 4 6 3\n",
            "2 1 5 6 3 4\n",
            "2 1 5 6 4 3\n",
            "2 1 6 3 4 5\n",
            "2 1 6 3 5 4\n",
            "2 1 6 4 3 5\n",
            "2 1 6 4 5 3\n",
            "2 1 6 5 3 4\n",
            "2 1 6 5 4 3\n",
            "2 3 1 4 5 6\n",
            "2 3 1 4 6 5\n",
            "2 3 1 5 4 6\n",
            "2 3 1 5 6 4\n",
            "2 3 1 6 4 5\n",
            "2 3 1 6 5 4\n",
            "2 3 4 1 5 6\n",
            "2 3 4 1 6 5\n",
            "2 3 4 5 1 6\n",
            "2 3 4 5 6 1\n",
            "2 3 4 6 1 5\n",
            "2 3 4 6 5 1\n",
            "2 3 5 1 4 6\n",
            "2 3 5 1 6 4\n",
            "2 3 5 4 1 6\n",
            "2 3 5 4 6 1\n",
            "2 3 5 6 1 4\n",
            "2 3 5 6 4 1\n",
            "2 3 6 1 4 5\n",
            "2 3 6 1 5 4\n",
            "2 3 6 4 1 5\n",
            "2 3 6 4 5 1\n",
            "2 3 6 5 1 4\n",
            "2 3 6 5 4 1\n",
            "2 4 1 3 5 6\n",
            "2 4 1 3 6 5\n",
            "2 4 1 5 3 6\n",
            "2 4 1 5 6 3\n",
            "2 4 1 6 3 5\n",
            "2 4 1 6 5 3\n",
            "2 4 3 1 5 6\n",
            "2 4 3 1 6 5\n",
            "2 4 3 5 1 6\n",
            "2 4 3 5 6 1\n",
            "2 4 3 6 1 5\n",
            "2 4 3 6 5 1\n",
            "2 4 5 1 3 6\n",
            "2 4 5 1 6 3\n",
            "2 4 5 3 1 6\n",
            "2 4 5 3 6 1\n",
            "2 4 5 6 1 3\n",
            "2 4 5 6 3 1\n",
            "2 4 6 1 3 5\n",
            "2 4 6 1 5 3\n",
            "2 4 6 3 1 5\n",
            "2 4 6 3 5 1\n",
            "2 4 6 5 1 3\n",
            "2 4 6 5 3 1\n",
            "2 5 1 3 4 6\n",
            "2 5 1 3 6 4\n",
            "2 5 1 4 3 6\n",
            "2 5 1 4 6 3\n",
            "2 5 1 6 3 4\n",
            "2 5 1 6 4 3\n",
            "2 5 3 1 4 6\n",
            "2 5 3 1 6 4\n",
            "2 5 3 4 1 6\n",
            "2 5 3 4 6 1\n",
            "2 5 3 6 1 4\n",
            "2 5 3 6 4 1\n",
            "2 5 4 1 3 6\n",
            "2 5 4 1 6 3\n",
            "2 5 4 3 1 6\n",
            "2 5 4 3 6 1\n",
            "2 5 4 6 1 3\n",
            "2 5 4 6 3 1\n",
            "2 5 6 1 3 4\n",
            "2 5 6 1 4 3\n",
            "2 5 6 3 1 4\n",
            "2 5 6 3 4 1\n",
            "2 5 6 4 1 3\n",
            "2 5 6 4 3 1\n",
            "2 6 1 3 4 5\n",
            "2 6 1 3 5 4\n",
            "2 6 1 4 3 5\n",
            "2 6 1 4 5 3\n",
            "2 6 1 5 3 4\n",
            "2 6 1 5 4 3\n",
            "2 6 3 1 4 5\n",
            "2 6 3 1 5 4\n",
            "2 6 3 4 1 5\n",
            "2 6 3 4 5 1\n",
            "2 6 3 5 1 4\n",
            "2 6 3 5 4 1\n",
            "2 6 4 1 3 5\n",
            "2 6 4 1 5 3\n",
            "2 6 4 3 1 5\n",
            "2 6 4 3 5 1\n",
            "2 6 4 5 1 3\n",
            "2 6 4 5 3 1\n",
            "2 6 5 1 3 4\n",
            "2 6 5 1 4 3\n",
            "2 6 5 3 1 4\n",
            "2 6 5 3 4 1\n",
            "2 6 5 4 1 3\n",
            "2 6 5 4 3 1\n",
            "3 1 2 4 5 6\n",
            "3 1 2 4 6 5\n",
            "3 1 2 5 4 6\n",
            "3 1 2 5 6 4\n",
            "3 1 2 6 4 5\n",
            "3 1 2 6 5 4\n",
            "3 1 4 2 5 6\n",
            "3 1 4 2 6 5\n",
            "3 1 4 5 2 6\n",
            "3 1 4 5 6 2\n",
            "3 1 4 6 2 5\n",
            "3 1 4 6 5 2\n",
            "3 1 5 2 4 6\n",
            "3 1 5 2 6 4\n",
            "3 1 5 4 2 6\n",
            "3 1 5 4 6 2\n",
            "3 1 5 6 2 4\n",
            "3 1 5 6 4 2\n",
            "3 1 6 2 4 5\n",
            "3 1 6 2 5 4\n",
            "3 1 6 4 2 5\n",
            "3 1 6 4 5 2\n",
            "3 1 6 5 2 4\n",
            "3 1 6 5 4 2\n",
            "3 2 1 4 5 6\n",
            "3 2 1 4 6 5\n",
            "3 2 1 5 4 6\n",
            "3 2 1 5 6 4\n",
            "3 2 1 6 4 5\n",
            "3 2 1 6 5 4\n",
            "3 2 4 1 5 6\n",
            "3 2 4 1 6 5\n",
            "3 2 4 5 1 6\n",
            "3 2 4 5 6 1\n",
            "3 2 4 6 1 5\n",
            "3 2 4 6 5 1\n",
            "3 2 5 1 4 6\n",
            "3 2 5 1 6 4\n",
            "3 2 5 4 1 6\n",
            "3 2 5 4 6 1\n",
            "3 2 5 6 1 4\n",
            "3 2 5 6 4 1\n",
            "3 2 6 1 4 5\n",
            "3 2 6 1 5 4\n",
            "3 2 6 4 1 5\n",
            "3 2 6 4 5 1\n",
            "3 2 6 5 1 4\n",
            "3 2 6 5 4 1\n",
            "3 4 1 2 5 6\n",
            "3 4 1 2 6 5\n",
            "3 4 1 5 2 6\n",
            "3 4 1 5 6 2\n",
            "3 4 1 6 2 5\n",
            "3 4 1 6 5 2\n",
            "3 4 2 1 5 6\n",
            "3 4 2 1 6 5\n",
            "3 4 2 5 1 6\n",
            "3 4 2 5 6 1\n",
            "3 4 2 6 1 5\n",
            "3 4 2 6 5 1\n",
            "3 4 5 1 2 6\n",
            "3 4 5 1 6 2\n",
            "3 4 5 2 1 6\n",
            "3 4 5 2 6 1\n",
            "3 4 5 6 1 2\n",
            "3 4 5 6 2 1\n",
            "3 4 6 1 2 5\n",
            "3 4 6 1 5 2\n",
            "3 4 6 2 1 5\n",
            "3 4 6 2 5 1\n",
            "3 4 6 5 1 2\n",
            "3 4 6 5 2 1\n",
            "3 5 1 2 4 6\n",
            "3 5 1 2 6 4\n",
            "3 5 1 4 2 6\n",
            "3 5 1 4 6 2\n",
            "3 5 1 6 2 4\n",
            "3 5 1 6 4 2\n",
            "3 5 2 1 4 6\n",
            "3 5 2 1 6 4\n",
            "3 5 2 4 1 6\n",
            "3 5 2 4 6 1\n",
            "3 5 2 6 1 4\n",
            "3 5 2 6 4 1\n",
            "3 5 4 1 2 6\n",
            "3 5 4 1 6 2\n",
            "3 5 4 2 1 6\n",
            "3 5 4 2 6 1\n",
            "3 5 4 6 1 2\n",
            "3 5 4 6 2 1\n",
            "3 5 6 1 2 4\n",
            "3 5 6 1 4 2\n",
            "3 5 6 2 1 4\n",
            "3 5 6 2 4 1\n",
            "3 5 6 4 1 2\n",
            "3 5 6 4 2 1\n",
            "3 6 1 2 4 5\n",
            "3 6 1 2 5 4\n",
            "3 6 1 4 2 5\n",
            "3 6 1 4 5 2\n",
            "3 6 1 5 2 4\n",
            "3 6 1 5 4 2\n",
            "3 6 2 1 4 5\n",
            "3 6 2 1 5 4\n",
            "3 6 2 4 1 5\n",
            "3 6 2 4 5 1\n",
            "3 6 2 5 1 4\n",
            "3 6 2 5 4 1\n",
            "3 6 4 1 2 5\n",
            "3 6 4 1 5 2\n",
            "3 6 4 2 1 5\n",
            "3 6 4 2 5 1\n",
            "3 6 4 5 1 2\n",
            "3 6 4 5 2 1\n",
            "3 6 5 1 2 4\n",
            "3 6 5 1 4 2\n",
            "3 6 5 2 1 4\n",
            "3 6 5 2 4 1\n",
            "3 6 5 4 1 2\n",
            "3 6 5 4 2 1\n",
            "4 1 2 3 5 6\n",
            "4 1 2 3 6 5\n",
            "4 1 2 5 3 6\n",
            "4 1 2 5 6 3\n",
            "4 1 2 6 3 5\n",
            "4 1 2 6 5 3\n",
            "4 1 3 2 5 6\n",
            "4 1 3 2 6 5\n",
            "4 1 3 5 2 6\n",
            "4 1 3 5 6 2\n",
            "4 1 3 6 2 5\n",
            "4 1 3 6 5 2\n",
            "4 1 5 2 3 6\n",
            "4 1 5 2 6 3\n",
            "4 1 5 3 2 6\n",
            "4 1 5 3 6 2\n",
            "4 1 5 6 2 3\n",
            "4 1 5 6 3 2\n",
            "4 1 6 2 3 5\n",
            "4 1 6 2 5 3\n",
            "4 1 6 3 2 5\n",
            "4 1 6 3 5 2\n",
            "4 1 6 5 2 3\n",
            "4 1 6 5 3 2\n",
            "4 2 1 3 5 6\n",
            "4 2 1 3 6 5\n",
            "4 2 1 5 3 6\n",
            "4 2 1 5 6 3\n",
            "4 2 1 6 3 5\n",
            "4 2 1 6 5 3\n",
            "4 2 3 1 5 6\n",
            "4 2 3 1 6 5\n",
            "4 2 3 5 1 6\n",
            "4 2 3 5 6 1\n",
            "4 2 3 6 1 5\n",
            "4 2 3 6 5 1\n",
            "4 2 5 1 3 6\n",
            "4 2 5 1 6 3\n",
            "4 2 5 3 1 6\n",
            "4 2 5 3 6 1\n",
            "4 2 5 6 1 3\n",
            "4 2 5 6 3 1\n",
            "4 2 6 1 3 5\n",
            "4 2 6 1 5 3\n",
            "4 2 6 3 1 5\n",
            "4 2 6 3 5 1\n",
            "4 2 6 5 1 3\n",
            "4 2 6 5 3 1\n",
            "4 3 1 2 5 6\n",
            "4 3 1 2 6 5\n",
            "4 3 1 5 2 6\n",
            "4 3 1 5 6 2\n",
            "4 3 1 6 2 5\n",
            "4 3 1 6 5 2\n",
            "4 3 2 1 5 6\n",
            "4 3 2 1 6 5\n",
            "4 3 2 5 1 6\n",
            "4 3 2 5 6 1\n",
            "4 3 2 6 1 5\n",
            "4 3 2 6 5 1\n",
            "4 3 5 1 2 6\n",
            "4 3 5 1 6 2\n",
            "4 3 5 2 1 6\n",
            "4 3 5 2 6 1\n",
            "4 3 5 6 1 2\n",
            "4 3 5 6 2 1\n",
            "4 3 6 1 2 5\n",
            "4 3 6 1 5 2\n",
            "4 3 6 2 1 5\n",
            "4 3 6 2 5 1\n",
            "4 3 6 5 1 2\n",
            "4 3 6 5 2 1\n",
            "4 5 1 2 3 6\n",
            "4 5 1 2 6 3\n",
            "4 5 1 3 2 6\n",
            "4 5 1 3 6 2\n",
            "4 5 1 6 2 3\n",
            "4 5 1 6 3 2\n",
            "4 5 2 1 3 6\n",
            "4 5 2 1 6 3\n",
            "4 5 2 3 1 6\n",
            "4 5 2 3 6 1\n",
            "4 5 2 6 1 3\n",
            "4 5 2 6 3 1\n",
            "4 5 3 1 2 6\n",
            "4 5 3 1 6 2\n",
            "4 5 3 2 1 6\n",
            "4 5 3 2 6 1\n",
            "4 5 3 6 1 2\n",
            "4 5 3 6 2 1\n",
            "4 5 6 1 2 3\n",
            "4 5 6 1 3 2\n",
            "4 5 6 2 1 3\n",
            "4 5 6 2 3 1\n",
            "4 5 6 3 1 2\n",
            "4 5 6 3 2 1\n",
            "4 6 1 2 3 5\n",
            "4 6 1 2 5 3\n",
            "4 6 1 3 2 5\n",
            "4 6 1 3 5 2\n",
            "4 6 1 5 2 3\n",
            "4 6 1 5 3 2\n",
            "4 6 2 1 3 5\n",
            "4 6 2 1 5 3\n",
            "4 6 2 3 1 5\n",
            "4 6 2 3 5 1\n",
            "4 6 2 5 1 3\n",
            "4 6 2 5 3 1\n",
            "4 6 3 1 2 5\n",
            "4 6 3 1 5 2\n",
            "4 6 3 2 1 5\n",
            "4 6 3 2 5 1\n",
            "4 6 3 5 1 2\n",
            "4 6 3 5 2 1\n",
            "4 6 5 1 2 3\n",
            "4 6 5 1 3 2\n",
            "4 6 5 2 1 3\n",
            "4 6 5 2 3 1\n",
            "4 6 5 3 1 2\n",
            "4 6 5 3 2 1\n",
            "5 1 2 3 4 6\n",
            "5 1 2 3 6 4\n",
            "5 1 2 4 3 6\n",
            "5 1 2 4 6 3\n",
            "5 1 2 6 3 4\n",
            "5 1 2 6 4 3\n",
            "5 1 3 2 4 6\n",
            "5 1 3 2 6 4\n",
            "5 1 3 4 2 6\n",
            "5 1 3 4 6 2\n",
            "5 1 3 6 2 4\n",
            "5 1 3 6 4 2\n",
            "5 1 4 2 3 6\n",
            "5 1 4 2 6 3\n",
            "5 1 4 3 2 6\n",
            "5 1 4 3 6 2\n",
            "5 1 4 6 2 3\n",
            "5 1 4 6 3 2\n",
            "5 1 6 2 3 4\n",
            "5 1 6 2 4 3\n",
            "5 1 6 3 2 4\n",
            "5 1 6 3 4 2\n",
            "5 1 6 4 2 3\n",
            "5 1 6 4 3 2\n",
            "5 2 1 3 4 6\n",
            "5 2 1 3 6 4\n",
            "5 2 1 4 3 6\n",
            "5 2 1 4 6 3\n",
            "5 2 1 6 3 4\n",
            "5 2 1 6 4 3\n",
            "5 2 3 1 4 6\n",
            "5 2 3 1 6 4\n",
            "5 2 3 4 1 6\n",
            "5 2 3 4 6 1\n",
            "5 2 3 6 1 4\n",
            "5 2 3 6 4 1\n",
            "5 2 4 1 3 6\n",
            "5 2 4 1 6 3\n",
            "5 2 4 3 1 6\n",
            "5 2 4 3 6 1\n",
            "5 2 4 6 1 3\n",
            "5 2 4 6 3 1\n",
            "5 2 6 1 3 4\n",
            "5 2 6 1 4 3\n",
            "5 2 6 3 1 4\n",
            "5 2 6 3 4 1\n",
            "5 2 6 4 1 3\n",
            "5 2 6 4 3 1\n",
            "5 3 1 2 4 6\n",
            "5 3 1 2 6 4\n",
            "5 3 1 4 2 6\n",
            "5 3 1 4 6 2\n",
            "5 3 1 6 2 4\n",
            "5 3 1 6 4 2\n",
            "5 3 2 1 4 6\n",
            "5 3 2 1 6 4\n",
            "5 3 2 4 1 6\n",
            "5 3 2 4 6 1\n",
            "5 3 2 6 1 4\n",
            "5 3 2 6 4 1\n",
            "5 3 4 1 2 6\n",
            "5 3 4 1 6 2\n",
            "5 3 4 2 1 6\n",
            "5 3 4 2 6 1\n",
            "5 3 4 6 1 2\n",
            "5 3 4 6 2 1\n",
            "5 3 6 1 2 4\n",
            "5 3 6 1 4 2\n",
            "5 3 6 2 1 4\n",
            "5 3 6 2 4 1\n",
            "5 3 6 4 1 2\n",
            "5 3 6 4 2 1\n",
            "5 4 1 2 3 6\n",
            "5 4 1 2 6 3\n",
            "5 4 1 3 2 6\n",
            "5 4 1 3 6 2\n",
            "5 4 1 6 2 3\n",
            "5 4 1 6 3 2\n",
            "5 4 2 1 3 6\n",
            "5 4 2 1 6 3\n",
            "5 4 2 3 1 6\n",
            "5 4 2 3 6 1\n",
            "5 4 2 6 1 3\n",
            "5 4 2 6 3 1\n",
            "5 4 3 1 2 6\n",
            "5 4 3 1 6 2\n",
            "5 4 3 2 1 6\n",
            "5 4 3 2 6 1\n",
            "5 4 3 6 1 2\n",
            "5 4 3 6 2 1\n",
            "5 4 6 1 2 3\n",
            "5 4 6 1 3 2\n",
            "5 4 6 2 1 3\n",
            "5 4 6 2 3 1\n",
            "5 4 6 3 1 2\n",
            "5 4 6 3 2 1\n",
            "5 6 1 2 3 4\n",
            "5 6 1 2 4 3\n",
            "5 6 1 3 2 4\n",
            "5 6 1 3 4 2\n",
            "5 6 1 4 2 3\n",
            "5 6 1 4 3 2\n",
            "5 6 2 1 3 4\n",
            "5 6 2 1 4 3\n",
            "5 6 2 3 1 4\n",
            "5 6 2 3 4 1\n",
            "5 6 2 4 1 3\n",
            "5 6 2 4 3 1\n",
            "5 6 3 1 2 4\n",
            "5 6 3 1 4 2\n",
            "5 6 3 2 1 4\n",
            "5 6 3 2 4 1\n",
            "5 6 3 4 1 2\n",
            "5 6 3 4 2 1\n",
            "5 6 4 1 2 3\n",
            "5 6 4 1 3 2\n",
            "5 6 4 2 1 3\n",
            "5 6 4 2 3 1\n",
            "5 6 4 3 1 2\n",
            "5 6 4 3 2 1\n",
            "6 1 2 3 4 5\n",
            "6 1 2 3 5 4\n",
            "6 1 2 4 3 5\n",
            "6 1 2 4 5 3\n",
            "6 1 2 5 3 4\n",
            "6 1 2 5 4 3\n",
            "6 1 3 2 4 5\n",
            "6 1 3 2 5 4\n",
            "6 1 3 4 2 5\n",
            "6 1 3 4 5 2\n",
            "6 1 3 5 2 4\n",
            "6 1 3 5 4 2\n",
            "6 1 4 2 3 5\n",
            "6 1 4 2 5 3\n",
            "6 1 4 3 2 5\n",
            "6 1 4 3 5 2\n",
            "6 1 4 5 2 3\n",
            "6 1 4 5 3 2\n",
            "6 1 5 2 3 4\n",
            "6 1 5 2 4 3\n",
            "6 1 5 3 2 4\n",
            "6 1 5 3 4 2\n",
            "6 1 5 4 2 3\n",
            "6 1 5 4 3 2\n",
            "6 2 1 3 4 5\n",
            "6 2 1 3 5 4\n",
            "6 2 1 4 3 5\n",
            "6 2 1 4 5 3\n",
            "6 2 1 5 3 4\n",
            "6 2 1 5 4 3\n",
            "6 2 3 1 4 5\n",
            "6 2 3 1 5 4\n",
            "6 2 3 4 1 5\n",
            "6 2 3 4 5 1\n",
            "6 2 3 5 1 4\n",
            "6 2 3 5 4 1\n",
            "6 2 4 1 3 5\n",
            "6 2 4 1 5 3\n",
            "6 2 4 3 1 5\n",
            "6 2 4 3 5 1\n",
            "6 2 4 5 1 3\n",
            "6 2 4 5 3 1\n",
            "6 2 5 1 3 4\n",
            "6 2 5 1 4 3\n",
            "6 2 5 3 1 4\n",
            "6 2 5 3 4 1\n",
            "6 2 5 4 1 3\n",
            "6 2 5 4 3 1\n",
            "6 3 1 2 4 5\n",
            "6 3 1 2 5 4\n",
            "6 3 1 4 2 5\n",
            "6 3 1 4 5 2\n",
            "6 3 1 5 2 4\n",
            "6 3 1 5 4 2\n",
            "6 3 2 1 4 5\n",
            "6 3 2 1 5 4\n",
            "6 3 2 4 1 5\n",
            "6 3 2 4 5 1\n",
            "6 3 2 5 1 4\n",
            "6 3 2 5 4 1\n",
            "6 3 4 1 2 5\n",
            "6 3 4 1 5 2\n",
            "6 3 4 2 1 5\n",
            "6 3 4 2 5 1\n",
            "6 3 4 5 1 2\n",
            "6 3 4 5 2 1\n",
            "6 3 5 1 2 4\n",
            "6 3 5 1 4 2\n",
            "6 3 5 2 1 4\n",
            "6 3 5 2 4 1\n",
            "6 3 5 4 1 2\n",
            "6 3 5 4 2 1\n",
            "6 4 1 2 3 5\n",
            "6 4 1 2 5 3\n",
            "6 4 1 3 2 5\n",
            "6 4 1 3 5 2\n",
            "6 4 1 5 2 3\n",
            "6 4 1 5 3 2\n",
            "6 4 2 1 3 5\n",
            "6 4 2 1 5 3\n",
            "6 4 2 3 1 5\n",
            "6 4 2 3 5 1\n",
            "6 4 2 5 1 3\n",
            "6 4 2 5 3 1\n",
            "6 4 3 1 2 5\n",
            "6 4 3 1 5 2\n",
            "6 4 3 2 1 5\n",
            "6 4 3 2 5 1\n",
            "6 4 3 5 1 2\n",
            "6 4 3 5 2 1\n",
            "6 4 5 1 2 3\n",
            "6 4 5 1 3 2\n",
            "6 4 5 2 1 3\n",
            "6 4 5 2 3 1\n",
            "6 4 5 3 1 2\n",
            "6 4 5 3 2 1\n",
            "6 5 1 2 3 4\n",
            "6 5 1 2 4 3\n",
            "6 5 1 3 2 4\n",
            "6 5 1 3 4 2\n",
            "6 5 1 4 2 3\n",
            "6 5 1 4 3 2\n",
            "6 5 2 1 3 4\n",
            "6 5 2 1 4 3\n",
            "6 5 2 3 1 4\n",
            "6 5 2 3 4 1\n",
            "6 5 2 4 1 3\n",
            "6 5 2 4 3 1\n",
            "6 5 3 1 2 4\n",
            "6 5 3 1 4 2\n",
            "6 5 3 2 1 4\n",
            "6 5 3 2 4 1\n",
            "6 5 3 4 1 2\n",
            "6 5 3 4 2 1\n",
            "6 5 4 1 2 3\n",
            "6 5 4 1 3 2\n",
            "6 5 4 2 1 3\n",
            "6 5 4 2 3 1\n",
            "6 5 4 3 1 2\n",
            "6 5 4 3 2 1\n",
            "\n",
            "Results written to: output_perm.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Insertion Sort Swaps Problem"
      ],
      "metadata": {
        "id": "wPf0XdYnVklH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Rosalind Insertion Sort Swaps Problem Solution\n",
        "\n",
        "This module counts the number of swaps (shifts) performed by the insertion sort\n",
        "algorithm when sorting an array of integers.\n",
        "\n",
        "The problem asks for the total number of swaps needed to sort an array using\n",
        "insertion sort, which is equivalent to counting the number of inversions in the array.\n",
        "\n",
        "Author: Bioinformatics Solution\n",
        "Compatible with: Python 3.6+\n",
        "Platform: Google Colab\n",
        "\"\"\"\n",
        "\n",
        "from typing import List, Tuple\n",
        "import copy\n",
        "\n",
        "\n",
        "class InsertionSortAnalyzer:\n",
        "    \"\"\"\n",
        "    A class to analyze insertion sort performance and count swaps/inversions.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"Initialize the insertion sort analyzer.\"\"\"\n",
        "        self.swap_count = 0\n",
        "\n",
        "    def insertion_sort_with_count(self, arr: List[int]) -> int:\n",
        "        \"\"\"\n",
        "        Perform insertion sort while counting the number of swaps.\n",
        "\n",
        "        Args:\n",
        "            arr (List[int]): Array to sort (will be modified)\n",
        "\n",
        "        Returns:\n",
        "            int: Number of swaps performed\n",
        "        \"\"\"\n",
        "        swap_count = 0\n",
        "        n = len(arr)\n",
        "\n",
        "        # Start from second element (index 1)\n",
        "        for i in range(1, n):\n",
        "            current_value = arr[i]\n",
        "            position = i\n",
        "\n",
        "            # Shift elements to the right while they are greater than current_value\n",
        "            while position > 0 and arr[position - 1] > current_value:\n",
        "                arr[position] = arr[position - 1]  # This is a swap/shift\n",
        "                position -= 1\n",
        "                swap_count += 1\n",
        "\n",
        "            # Place current_value at its correct position\n",
        "            arr[position] = current_value\n",
        "\n",
        "        return swap_count\n",
        "\n",
        "    def count_swaps_without_sorting(self, arr: List[int]) -> int:\n",
        "        \"\"\"\n",
        "        Count swaps needed for insertion sort without actually sorting.\n",
        "        This counts inversions in the array.\n",
        "\n",
        "        Args:\n",
        "            arr (List[int]): Original array (won't be modified)\n",
        "\n",
        "        Returns:\n",
        "            int: Number of swaps needed\n",
        "        \"\"\"\n",
        "        swap_count = 0\n",
        "        n = len(arr)\n",
        "\n",
        "        # For each element starting from second\n",
        "        for i in range(1, n):\n",
        "            current_value = arr[i]\n",
        "\n",
        "            # Count how many elements to the left are greater\n",
        "            for j in range(i - 1, -1, -1):\n",
        "                if arr[j] > current_value:\n",
        "                    swap_count += 1\n",
        "                else:\n",
        "                    break  # Elements to the left are sorted, so we can break\n",
        "\n",
        "        return swap_count\n",
        "\n",
        "    def count_inversions_merge_sort(self, arr: List[int]) -> int:\n",
        "        \"\"\"\n",
        "        Count inversions using modified merge sort (O(n log n) approach).\n",
        "\n",
        "        Args:\n",
        "            arr (List[int]): Array to analyze\n",
        "\n",
        "        Returns:\n",
        "            int: Number of inversions (swaps needed)\n",
        "        \"\"\"\n",
        "        def merge_and_count(arr: List[int], temp: List[int], left: int, mid: int, right: int) -> int:\n",
        "            \"\"\"Helper function to merge and count inversions.\"\"\"\n",
        "            i, j, k = left, mid + 1, left\n",
        "            inversion_count = 0\n",
        "\n",
        "            # Merge the two halves while counting inversions\n",
        "            while i <= mid and j <= right:\n",
        "                if arr[i] <= arr[j]:\n",
        "                    temp[k] = arr[i]\n",
        "                    i += 1\n",
        "                else:\n",
        "                    temp[k] = arr[j]\n",
        "                    # All elements from i to mid are greater than arr[j]\n",
        "                    inversion_count += (mid - i + 1)\n",
        "                    j += 1\n",
        "                k += 1\n",
        "\n",
        "            # Copy remaining elements\n",
        "            while i <= mid:\n",
        "                temp[k] = arr[i]\n",
        "                i += 1\n",
        "                k += 1\n",
        "\n",
        "            while j <= right:\n",
        "                temp[k] = arr[j]\n",
        "                j += 1\n",
        "                k += 1\n",
        "\n",
        "            # Copy back to original array\n",
        "            for i in range(left, right + 1):\n",
        "                arr[i] = temp[i]\n",
        "\n",
        "            return inversion_count\n",
        "\n",
        "        def merge_sort_and_count(arr: List[int], temp: List[int], left: int, right: int) -> int:\n",
        "            \"\"\"Recursive function to perform merge sort and count inversions.\"\"\"\n",
        "            inversion_count = 0\n",
        "            if left < right:\n",
        "                mid = (left + right) // 2\n",
        "\n",
        "                inversion_count += merge_sort_and_count(arr, temp, left, mid)\n",
        "                inversion_count += merge_sort_and_count(arr, temp, mid + 1, right)\n",
        "                inversion_count += merge_and_count(arr, temp, left, mid, right)\n",
        "\n",
        "            return inversion_count\n",
        "\n",
        "        # Create a copy to avoid modifying original array\n",
        "        arr_copy = arr.copy()\n",
        "        temp = [0] * len(arr_copy)\n",
        "        return merge_sort_and_count(arr_copy, temp, 0, len(arr_copy) - 1)\n",
        "\n",
        "    def solve_insertion_sort_problem(self, arr: List[int], method: str = \"direct\") -> int:\n",
        "        \"\"\"\n",
        "        Solve the insertion sort swap counting problem.\n",
        "\n",
        "        Args:\n",
        "            arr (List[int]): Input array\n",
        "            method (str): Method to use - \"direct\", \"count_only\", or \"merge_sort\"\n",
        "\n",
        "        Returns:\n",
        "            int: Number of swaps needed\n",
        "        \"\"\"\n",
        "        if not arr:\n",
        "            return 0\n",
        "\n",
        "        if method == \"direct\":\n",
        "            # Actually perform insertion sort and count swaps\n",
        "            arr_copy = copy.deepcopy(arr)\n",
        "            return self.insertion_sort_with_count(arr_copy)\n",
        "\n",
        "        elif method == \"count_only\":\n",
        "            # Count swaps without sorting (O(n²) but doesn't modify array)\n",
        "            return self.count_swaps_without_sorting(arr)\n",
        "\n",
        "        elif method == \"merge_sort\":\n",
        "            # Use merge sort approach (O(n log n))\n",
        "            return self.count_inversions_merge_sort(arr)\n",
        "\n",
        "        else:\n",
        "            raise ValueError(\"Method must be 'direct', 'count_only', or 'merge_sort'\")\n",
        "\n",
        "\n",
        "def parse_input_file(file_path: str) -> Tuple[int, List[int]]:\n",
        "    \"\"\"\n",
        "    Parse input file to extract n and the array.\n",
        "\n",
        "    Args:\n",
        "        file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        Tuple[int, List[int]]: (n, array)\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If file format is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(file_path, 'r') as file:\n",
        "            lines = file.read().strip().split('\\n')\n",
        "\n",
        "        if len(lines) < 2:\n",
        "            raise ValueError(\"Input file must contain at least 2 lines\")\n",
        "\n",
        "        # Parse n\n",
        "        try:\n",
        "            n = int(lines[0].strip())\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"First line must be an integer, got: '{lines[0]}'\")\n",
        "\n",
        "        # Parse array\n",
        "        try:\n",
        "            array = list(map(int, lines[1].strip().split()))\n",
        "        except ValueError:\n",
        "            raise ValueError(f\"Second line must contain integers, got: '{lines[1]}'\")\n",
        "\n",
        "        # Validate\n",
        "        if len(array) != n:\n",
        "            raise ValueError(f\"Array length {len(array)} doesn't match specified n={n}\")\n",
        "\n",
        "        if n > 1000:\n",
        "            raise ValueError(f\"n must be ≤ 1000, got {n}\")\n",
        "\n",
        "        return n, array\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        raise FileNotFoundError(f\"Input file '{file_path}' not found\")\n",
        "\n",
        "\n",
        "def write_output_file(output_path: str, swap_count: int) -> None:\n",
        "    \"\"\"\n",
        "    Write result to output file.\n",
        "\n",
        "    Args:\n",
        "        output_path (str): Path to output file\n",
        "        swap_count (int): Number of swaps\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(output_path, 'w') as file:\n",
        "            file.write(f\"{swap_count}\\n\")\n",
        "    except Exception as e:\n",
        "        raise IOError(f\"Error writing to output file: {e}\")\n",
        "\n",
        "\n",
        "def solve_insertion_sort_swaps_problem(input_file_path: str) -> int:\n",
        "    \"\"\"\n",
        "    Solve the Insertion Sort Swaps problem for a given input file.\n",
        "\n",
        "    Args:\n",
        "        input_file_path (str): Path to input file\n",
        "\n",
        "    Returns:\n",
        "        int: Number of swaps needed\n",
        "\n",
        "    Raises:\n",
        "        FileNotFoundError: If input file doesn't exist\n",
        "        ValueError: If input is invalid\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Parse input\n",
        "        n, array = parse_input_file(input_file_path)\n",
        "\n",
        "        # Initialize analyzer\n",
        "        analyzer = InsertionSortAnalyzer()\n",
        "\n",
        "        # Solve using the most appropriate method\n",
        "        # For n ≤ 1000, direct method is fine\n",
        "        if n <= 1000:\n",
        "            swap_count = analyzer.solve_insertion_sort_problem(array, method=\"direct\")\n",
        "        else:\n",
        "            # Use merge sort method for larger arrays\n",
        "            swap_count = analyzer.solve_insertion_sort_problem(array, method=\"merge_sort\")\n",
        "\n",
        "        return swap_count\n",
        "\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Error solving problem: {str(e)}\")\n",
        "\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function to run the Insertion Sort Swaps problem solver.\n",
        "    Designed to work in Google Colab environment.\n",
        "    \"\"\"\n",
        "    # Configuration\n",
        "    input_file = \"rosalind_ins.txt\"  # Change this to your input file name\n",
        "    output_file = \"output_ins.txt\"\n",
        "\n",
        "    try:\n",
        "        print(\"Solving Insertion Sort Swaps Problem...\")\n",
        "\n",
        "        # Solve the problem\n",
        "        swap_count = solve_insertion_sort_swaps_problem(input_file)\n",
        "\n",
        "        # Display results\n",
        "        print(f\"\\nResult:\")\n",
        "        print(f\"Number of swaps needed: {swap_count}\")\n",
        "\n",
        "        # Write to output file\n",
        "        write_output_file(output_file, swap_count)\n",
        "        print(f\"\\nResult written to: {output_file}\")\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: Input file '{input_file}' not found.\")\n",
        "        print(\"Please make sure the file exists in the current directory.\")\n",
        "\n",
        "    except ValueError as e:\n",
        "        print(f\"Error: {e}\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Unexpected error: {e}\")\n",
        "\n",
        "\n",
        "def demo_with_sample():\n",
        "    \"\"\"\n",
        "    Demonstrate the solution with the sample data.\n",
        "    \"\"\"\n",
        "    print(\"=== Demo with Sample Data ===\")\n",
        "\n",
        "    # Sample input\n",
        "    n = 6\n",
        "    array = [6, 10, 4, 5, 1, 2]\n",
        "\n",
        "    print(f\"Input:\")\n",
        "    print(f\"n = {n}\")\n",
        "    print(f\"Array = {array}\")\n",
        "\n",
        "    analyzer = InsertionSortAnalyzer()\n",
        "\n",
        "    # Test all methods\n",
        "    methods = [\"direct\", \"count_only\", \"merge_sort\"]\n",
        "\n",
        "    for method in methods:\n",
        "        swap_count = analyzer.solve_insertion_sort_problem(array.copy(), method)\n",
        "        print(f\"\\nMethod '{method}': {swap_count} swaps\")\n",
        "\n",
        "    # Expected output is 12\n",
        "    print(f\"\\nExpected output: 12\")\n",
        "\n",
        "    # Show step-by-step insertion sort\n",
        "    print(f\"\\n=== Step-by-step insertion sort ===\")\n",
        "    arr_copy = array.copy()\n",
        "    print(f\"Initial: {arr_copy}\")\n",
        "\n",
        "    total_swaps = 0\n",
        "    for i in range(1, len(arr_copy)):\n",
        "        current = arr_copy[i]\n",
        "        pos = i\n",
        "        swaps_this_round = 0\n",
        "\n",
        "        while pos > 0 and arr_copy[pos - 1] > current:\n",
        "            arr_copy[pos] = arr_copy[pos - 1]\n",
        "            pos -= 1\n",
        "            swaps_this_round += 1\n",
        "\n",
        "        arr_copy[pos] = current\n",
        "        total_swaps += swaps_this_round\n",
        "        print(f\"After inserting {current}: {arr_copy} (swaps: {swaps_this_round}, total: {total_swaps})\")\n",
        "\n",
        "\n",
        "# Example usage and testing\n",
        "if __name__ == \"__main__\":\n",
        "    # Run demo first\n",
        "    demo_with_sample()\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Test with additional cases\n",
        "    print(\"=== Additional Test Cases ===\")\n",
        "    analyzer = InsertionSortAnalyzer()\n",
        "\n",
        "    test_cases = [\n",
        "        [1, 2, 3, 4, 5],  # Already sorted - 0 swaps\n",
        "        [5, 4, 3, 2, 1],  # Reverse sorted - maximum swaps\n",
        "        [3, 1, 4, 1, 5],  # Random case\n",
        "    ]\n",
        "\n",
        "    for i, test_array in enumerate(test_cases):\n",
        "        swaps = analyzer.solve_insertion_sort_problem(test_array.copy(), \"direct\")\n",
        "        print(f\"Test {i+1}: {test_array} -> {swaps} swaps\")\n",
        "\n",
        "    print(\"\\n\" + \"=\"*60)\n",
        "\n",
        "    # Uncomment to run with actual file input\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "doB4a9TvKTu_",
        "outputId": "59a80c3c-bec2-4d5a-8726-ffb31d73001b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "=== Demo with Sample Data ===\n",
            "Input:\n",
            "n = 6\n",
            "Array = [6, 10, 4, 5, 1, 2]\n",
            "\n",
            "Method 'direct': 12 swaps\n",
            "\n",
            "Method 'count_only': 6 swaps\n",
            "\n",
            "Method 'merge_sort': 12 swaps\n",
            "\n",
            "Expected output: 12\n",
            "\n",
            "=== Step-by-step insertion sort ===\n",
            "Initial: [6, 10, 4, 5, 1, 2]\n",
            "After inserting 10: [6, 10, 4, 5, 1, 2] (swaps: 0, total: 0)\n",
            "After inserting 4: [4, 6, 10, 5, 1, 2] (swaps: 2, total: 2)\n",
            "After inserting 5: [4, 5, 6, 10, 1, 2] (swaps: 2, total: 4)\n",
            "After inserting 1: [1, 4, 5, 6, 10, 2] (swaps: 4, total: 8)\n",
            "After inserting 2: [1, 2, 4, 5, 6, 10] (swaps: 4, total: 12)\n",
            "\n",
            "============================================================\n",
            "=== Additional Test Cases ===\n",
            "Test 1: [1, 2, 3, 4, 5] -> 0 swaps\n",
            "Test 2: [5, 4, 3, 2, 1] -> 10 swaps\n",
            "Test 3: [3, 1, 4, 1, 5] -> 3 swaps\n",
            "\n",
            "============================================================\n",
            "Solving Insertion Sort Swaps Problem...\n",
            "Error: Error solving problem: Input file 'rosalind_ins.txt' not found\n"
          ]
        }
      ]
    }
  ]
}